// @grafana/runtime/dist/cjs/internal/openFeature/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ofrepWebProvider = require('@openfeature/ofrep-web-provider');
var reactSdk = require('@openfeature/react-sdk');
var config = require('../../config.cjs');

"use strict";
const GRAFANA_CORE_OPEN_FEATURE_DOMAIN = "internal-grafana-core";
async function initOpenFeature() {
  const subPath = config.config.appSubUrl || "";
  const baseUrl = `${subPath}/apis/features.grafana.app/v0alpha1/namespaces/${config.config.namespace}`;
  const ofProvider = new ofrepWebProvider.OFREPWebProvider({
    baseUrl,
    pollInterval: -1,
    // disable polling
    timeoutMs: 5e3
  });
  await reactSdk.OpenFeature.setProviderAndWait(GRAFANA_CORE_OPEN_FEATURE_DOMAIN, ofProvider, {
    targetingKey: config.config.namespace,
    ...config.config.openFeatureContext
  });
}
function getFeatureFlagClient() {
  return reactSdk.OpenFeature.getClient(GRAFANA_CORE_OPEN_FEATURE_DOMAIN);
}

exports.GRAFANA_CORE_OPEN_FEATURE_DOMAIN = GRAFANA_CORE_OPEN_FEATURE_DOMAIN;
exports.getFeatureFlagClient = getFeatureFlagClient;
exports.initOpenFeature = initOpenFeature;
