// @grafana/ui/dist/cjs/components/Modal/Modal.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var react = require('@floating-ui/react');
var overlays = require('@react-aria/overlays');
var React = require('react');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var IconButton = require('../IconButton/IconButton.cjs');
var Stack = require('../Layout/Stack/Stack.cjs');
var Portal = require('../Portal/Portal.cjs');
var ModalHeader = require('./ModalHeader.cjs');
var getModalStyles = require('./getModalStyles.cjs');

"use strict";
function Modal(props) {
  const {
    title,
    ariaLabel,
    children,
    isOpen = false,
    closeOnEscape = true,
    closeOnBackdropClick = true,
    className,
    contentClassName,
    onDismiss,
    onClickBackdrop,
    trapFocus = true
  } = props;
  const styles = ThemeContext.useStyles2(getModalStyles.getModalStyles);
  const titleId = React.useId();
  const { context, refs } = react.useFloating({
    open: isOpen,
    onOpenChange: (open) => {
      if (!open) {
        onDismiss == null ? void 0 : onDismiss();
      }
    }
  });
  const dismiss = react.useDismiss(context, {
    enabled: closeOnEscape
  });
  const role = react.useRole(context, {
    role: "dialog"
  });
  const { getFloatingProps } = react.useInteractions([dismiss, role]);
  if (!isOpen) {
    return null;
  }
  const headerClass = css.cx(styles.modalHeader, typeof title !== "string" && styles.modalHeaderWithTabs);
  return /* @__PURE__ */ jsxRuntime.jsxs(overlays.OverlayContainer, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        role: "presentation",
        className: styles.modalBackdrop,
        onClick: onClickBackdrop || (closeOnBackdropClick ? onDismiss : void 0)
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(react.FloatingFocusManager, { context, modal: trapFocus, getInsideElements: () => [Portal.getPortalContainer()], children: /* @__PURE__ */ jsxRuntime.jsxs(
      "div",
      {
        className: css.cx(styles.modal, className),
        ref: refs.setFloating,
        "aria-label": ariaLabel,
        "aria-labelledby": typeof title === "string" ? titleId : void 0,
        ...getFloatingProps(),
        children: [
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: headerClass, children: [
            typeof title === "string" && /* @__PURE__ */ jsxRuntime.jsx(DefaultModalHeader, { ...props, title, id: titleId }),
            // FIXME: custom title components won't get an accessible title.
            // Do we really want to support them or shall we just limit this ModalTabsHeader?
            typeof title !== "string" && title,
            /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles.modalHeaderClose, children: /* @__PURE__ */ jsxRuntime.jsx(
              IconButton.IconButton,
              {
                name: "times",
                size: "xl",
                onClick: onDismiss,
                "aria-label": i18n.t("grafana-ui.modal.close-tooltip", "Close")
              }
            ) })
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx("div", { className: css.cx(styles.modalContent, contentClassName), children })
        ]
      }
    ) })
  ] });
}
function ModalButtonRow({ leftItems, children }) {
  const styles = ThemeContext.useStyles2(getModalStyles.getModalStyles);
  if (leftItems) {
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles.modalButtonRow, children: /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { justifyContent: "space-between", children: [
      /* @__PURE__ */ jsxRuntime.jsx(Stack.Stack, { justifyContent: "flex-start", gap: 2, children: leftItems }),
      /* @__PURE__ */ jsxRuntime.jsx(Stack.Stack, { justifyContent: "flex-end", gap: 2, children })
    ] }) });
  }
  return /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles.modalButtonRow, children: /* @__PURE__ */ jsxRuntime.jsx(Stack.Stack, { justifyContent: "flex-end", gap: 2, wrap: "wrap", children }) });
}
Modal.ButtonRow = ModalButtonRow;
function DefaultModalHeader({ icon, iconTooltip, title, id }) {
  return /* @__PURE__ */ jsxRuntime.jsx(ModalHeader.ModalHeader, { icon, iconTooltip, title, id });
}

exports.Modal = Modal;
