// @grafana/ui/dist/esm/components/DataSourceSettings/SecureSocksProxySettings.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { Trans, t } from '@grafana/i18n';
import { InlineSwitch } from '../Switch/Switch.mjs';
import { InlineField } from '../Forms/InlineField.mjs';
import { Box } from '../Layout/Box/Box.mjs';
import { Stack } from '../Layout/Stack/Stack.mjs';

"use strict";
function SecureSocksProxySettings({
  options,
  onOptionsChange
}) {
  var _a;
  return /* @__PURE__ */ jsxs("div", { children: [
    /* @__PURE__ */ jsx("h3", { className: "page-heading", children: /* @__PURE__ */ jsx(Trans, { i18nKey: "grafana-ui.data-source-settings.secure-socks-heading", children: "Secure Socks Proxy" }) }),
    /* @__PURE__ */ jsx(Box, { marginBottom: 5, children: /* @__PURE__ */ jsx(Stack, { direction: "row", alignItems: "flex-start", wrap: true, children: /* @__PURE__ */ jsx(Box, { position: "relative", marginBottom: 0.5, children: /* @__PURE__ */ jsx(Stack, { direction: "row", alignItems: "flex-start", children: /* @__PURE__ */ jsx(
      InlineField,
      {
        labelWidth: 26,
        label: t("grafana-ui.data-source-settings.secure-socks-label", "Enabled"),
        tooltip: t(
          "grafana-ui.data-source-settings.secure-socks-tooltip",
          "Connect to this datasource via the secure socks proxy."
        ),
        children: /* @__PURE__ */ jsx(
          InlineSwitch,
          {
            value: (_a = options.jsonData.enableSecureSocksProxy) != null ? _a : false,
            onChange: (event) => onOptionsChange({
              ...options,
              jsonData: { ...options.jsonData, enableSecureSocksProxy: event.currentTarget.checked }
            })
          }
        )
      }
    ) }) }) }) })
  ] });
}

export { SecureSocksProxySettings };
