// @grafana/ui/dist/esm/components/StatsPicker/StatsPicker.mjs
import { jsx } from 'react/jsx-runtime';
import { difference } from 'lodash';
import { memo, useEffect } from 'react';
import { fieldReducers } from '@grafana/data';
import { Select } from '../Select/Select.mjs';

"use strict";
const StatsPicker = memo(
  ({
    placeholder,
    onChange,
    stats,
    allowMultiple = false,
    defaultStat,
    className,
    width,
    menuPlacement,
    inputId,
    filterOptions
  }) => {
    useEffect(() => {
      const current = fieldReducers.list(stats);
      if (current.length !== stats.length) {
        const found = current.map((v) => v.id);
        const notFound = difference(stats, found);
        console.warn("Unknown stats", notFound, stats);
        onChange(current.map((stat) => stat.id));
      }
      if (!allowMultiple && stats.length > 1) {
        console.warn("Removing extra stat", stats);
        onChange([stats[0]]);
      }
      if (defaultStat && stats.length < 1) {
        onChange([defaultStat]);
      }
    }, [stats, allowMultiple, defaultStat, onChange]);
    const onSelectionChange = (item) => {
      if (Array.isArray(item)) {
        onChange(item.map((v) => v.value));
      } else {
        onChange(item && item.value ? [item.value] : []);
      }
    };
    const select = fieldReducers.selectOptions(stats, filterOptions);
    return /* @__PURE__ */ jsx(
      Select,
      {
        value: select.current,
        className,
        isClearable: !defaultStat,
        isMulti: allowMultiple,
        width,
        isSearchable: true,
        options: select.options,
        placeholder,
        onChange: onSelectionChange,
        menuPlacement,
        inputId
      }
    );
  }
);
StatsPicker.displayName = "StatsPicker";

export { StatsPicker };
