// @grafana/ui/dist/cjs/components/Combobox/useComboboxFloat.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var react = require('@floating-ui/react');
var React = require('react');
var floating = require('../../utils/floating.cjs');
var measureText = require('../../utils/measureText.cjs');
var getComboboxStyles = require('./getComboboxStyles.cjs');

"use strict";
const WIDTH_CALCULATION_LIMIT_ITEMS = 1e5;
const POPOVER_PADDING = 16;
const SCROLL_CONTAINER_PADDING = 8;
const useComboboxFloat = (items, isOpen) => {
  var _a, _b;
  const inputRef = React.useRef(null);
  const floatingRef = React.useRef(null);
  const scrollRef = React.useRef(null);
  const [popoverMaxSize, setPopoverMaxSize] = React.useState({
    width: 0,
    height: 0
  });
  const scrollbarWidth = React.useMemo(() => getScrollbarWidth(), []);
  const middleware = [
    react.autoPlacement({
      // see https://floating-ui.com/docs/autoplacement
      allowedPlacements: ["bottom-start", "bottom-end", "top-start", "top-end"],
      boundary: (_a = document.getElementById(floating.BOUNDARY_ELEMENT_ID)) != null ? _a : void 0,
      crossAxis: true
    }),
    react.size({
      apply({ availableWidth, availableHeight }) {
        const preferredMaxWidth = availableWidth - POPOVER_PADDING;
        const preferredMaxHeight = availableHeight - POPOVER_PADDING;
        const width = Math.max(preferredMaxWidth, 0);
        const height = Math.min(Math.max(preferredMaxHeight, getComboboxStyles.MENU_OPTION_HEIGHT * 6), getComboboxStyles.POPOVER_MAX_HEIGHT);
        setPopoverMaxSize({ width, height });
      }
    })
  ];
  const elements = { reference: inputRef.current, floating: floatingRef.current };
  const { floatingStyles } = react.useFloating({
    strategy: "fixed",
    open: isOpen,
    placement: "bottom-start",
    middleware,
    elements,
    whileElementsMounted: react.autoUpdate
  });
  const longestItemWidth = React.useMemo(() => {
    var _a2;
    let longestItem = "";
    const itemsToLookAt = Math.min(items.length, WIDTH_CALCULATION_LIMIT_ITEMS);
    for (let i = 0; i < itemsToLookAt; i++) {
      const itemLabel = (_a2 = items[i].label) != null ? _a2 : items[i].value.toString();
      longestItem = itemLabel.length > longestItem.length ? itemLabel : longestItem;
    }
    const size2 = measureText.measureText(longestItem, getComboboxStyles.MENU_ITEM_FONT_SIZE, getComboboxStyles.MENU_ITEM_FONT_WEIGHT).width;
    return size2 + SCROLL_CONTAINER_PADDING + getComboboxStyles.MENU_ITEM_PADDING * 2 + scrollbarWidth;
  }, [items, scrollbarWidth]);
  const floatStyles = {
    ...floatingStyles,
    width: longestItemWidth,
    maxWidth: popoverMaxSize.width,
    minWidth: (_b = inputRef.current) == null ? void 0 : _b.offsetWidth,
    maxHeight: popoverMaxSize.height
  };
  return { inputRef, floatingRef, scrollRef, floatStyles };
};
function getScrollbarWidth() {
  var _a;
  const outer = document.createElement("div");
  outer.style.visibility = "hidden";
  outer.style.overflow = "scroll";
  document.body.appendChild(outer);
  const inner = document.createElement("div");
  outer.appendChild(inner);
  const scrollbarWidth = outer.offsetWidth - inner.offsetWidth;
  (_a = outer.parentNode) == null ? void 0 : _a.removeChild(outer);
  return scrollbarWidth;
}

exports.useComboboxFloat = useComboboxFloat;
