// @grafana/ui/dist/cjs/components/Layout/utils/styles.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var css = require('@emotion/css');
var responsiveness = require('./responsiveness.cjs');

"use strict";
const getSizeStyles = (theme, width, minWidth, maxWidth, height, minHeight, maxHeight) => {
  return css.css([
    responsiveness.getResponsiveStyle(theme, width, (val) => ({
      width: theme.spacing(val)
    })),
    responsiveness.getResponsiveStyle(theme, minWidth, (val) => ({
      minWidth: theme.spacing(val)
    })),
    responsiveness.getResponsiveStyle(theme, maxWidth, (val) => ({
      maxWidth: theme.spacing(val)
    })),
    responsiveness.getResponsiveStyle(theme, height, (val) => ({
      height: theme.spacing(val)
    })),
    responsiveness.getResponsiveStyle(theme, minHeight, (val) => ({
      minHeight: theme.spacing(val)
    })),
    responsiveness.getResponsiveStyle(theme, maxHeight, (val) => ({
      maxHeight: theme.spacing(val)
    }))
  ]);
};

exports.getSizeStyles = getSizeStyles;
