// @grafana/ui/dist/esm/components/RefreshPicker/RefreshPicker.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import { formatDuration } from 'date-fns';
import { memo } from 'react';
import { parseDuration } from '@grafana/data';
import { selectors } from '@grafana/e2e-selectors';
import { t } from '@grafana/i18n';
import { ButtonGroup } from '../Button/ButtonGroup.mjs';
import { ButtonSelect } from '../Dropdown/ButtonSelect.mjs';
import { ToolbarButton } from '../ToolbarButton/ToolbarButton.mjs';

"use strict";
const defaultIntervals = ["5s", "10s", "30s", "1m", "5m", "15m", "30m", "1h", "2h", "1d"];
const offOption = {
  label: "Off",
  value: "",
  ariaLabel: "Turn off auto refresh"
};
const liveOption = {
  label: "Live",
  value: "LIVE",
  ariaLabel: "Turn on live streaming"
};
const autoOption = {
  label: "Auto",
  value: "auto",
  ariaLabel: "Select refresh from the query range"
};
const RefreshPickerComponent = memo((props) => {
  const {
    intervals,
    onRefresh,
    onIntervalChanged,
    value,
    tooltip,
    isLoading,
    isLive,
    text,
    noIntervalPicker,
    showAutoInterval,
    width,
    primary,
    isOnCanvas
  } = props;
  const currentValue = value || "";
  const options = intervalsToOptions({ intervals, showAutoInterval });
  const option = options.find(({ value: value2 }) => value2 === currentValue);
  const translatedOffOption = translateOption(offOption.value);
  let selectedValue = option || translatedOffOption;
  const handleChangeSelect = (item) => {
    if (onIntervalChanged && item.value != null) {
      onIntervalChanged(item.value);
    }
  };
  const getVariant = () => {
    if (isLive) {
      return "primary";
    }
    if (primary) {
      return "primary";
    }
    return isOnCanvas ? "canvas" : "default";
  };
  const variant = getVariant();
  if (selectedValue.label === translatedOffOption.label) {
    selectedValue = { value: "" };
  }
  const durationAriaLabel = selectedValue.ariaLabel;
  const ariaLabelDurationSelectedMessage = t(
    "refresh-picker.aria-label.duration-selected",
    "Choose refresh time interval with current interval {{durationAriaLabel}} selected",
    { durationAriaLabel }
  );
  const ariaLabelChooseIntervalMessage = t(
    "refresh-picker.aria-label.choose-interval",
    "Auto refresh turned off. Choose refresh time interval"
  );
  const ariaLabel = selectedValue.value === "" ? ariaLabelChooseIntervalMessage : ariaLabelDurationSelectedMessage;
  const tooltipIntervalSelected = t("refresh-picker.tooltip.interval-selected", "Set auto refresh interval");
  const tooltipAutoRefreshOff = t("refresh-picker.tooltip.turned-off", "Auto refresh off");
  const tooltipAutoRefresh = selectedValue.value === "" ? tooltipAutoRefreshOff : tooltipIntervalSelected;
  return /* @__PURE__ */ jsxs(ButtonGroup, { className: "refresh-picker", children: [
    /* @__PURE__ */ jsx(
      ToolbarButton,
      {
        "aria-label": text,
        tooltip,
        onClick: onRefresh,
        variant,
        icon: isLoading ? "spinner" : "sync",
        style: width ? { width } : void 0,
        "data-testid": selectors.components.RefreshPicker.runButtonV2,
        children: text
      }
    ),
    !noIntervalPicker && /* @__PURE__ */ jsx(
      ButtonSelect,
      {
        className: css({
          borderTopLeftRadius: "unset",
          borderBottomLeftRadius: "unset"
        }),
        value: selectedValue,
        options,
        onChange: handleChangeSelect,
        variant,
        "data-testid": selectors.components.RefreshPicker.intervalButtonV2,
        "aria-label": ariaLabel,
        tooltip: tooltipAutoRefresh
      }
    )
  ] });
});
RefreshPickerComponent.displayName = "RefreshPicker";
const RefreshPicker = Object.assign(RefreshPickerComponent, {
  isLive: (refreshInterval) => refreshInterval === liveOption.value,
  liveOption,
  offOption,
  autoOption
});
const translateOption = (option) => {
  switch (option) {
    case liveOption.value:
      return {
        label: t("refresh-picker.live-option.label", "Live"),
        value: option,
        ariaLabel: t("refresh-picker.live-option.aria-label", "Turn on live streaming")
      };
    case offOption.value:
      return {
        label: t("refresh-picker.off-option.label", "Off"),
        value: option,
        ariaLabel: t("refresh-picker.off-option.aria-label", "Turn off auto refresh")
      };
    case autoOption.value:
      return {
        label: t("refresh-picker.auto-option.label", autoOption.label),
        value: option,
        ariaLabel: t("refresh-picker.auto-option.aria-label", autoOption.ariaLabel)
      };
  }
  return {
    label: option,
    value: option
  };
};
const intervalsToOptions = ({
  intervals = defaultIntervals,
  showAutoInterval = false
} = {}) => {
  const options = intervals.map((interval) => {
    const duration = parseDuration(interval);
    const ariaLabel = formatDuration(duration);
    return {
      label: interval,
      value: interval,
      ariaLabel
    };
  });
  if (showAutoInterval) {
    options.unshift(translateOption(autoOption.value));
  }
  options.unshift(translateOption(offOption.value));
  return options;
};

export { RefreshPicker, defaultIntervals, intervalsToOptions, translateOption };
