// @grafana/runtime/dist/cjs/analytics/utils.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var config = require('../config.cjs');
require('../services/index.cjs');
var EchoSrv = require('../services/EchoSrv.cjs');
var LocationService = require('../services/LocationService.cjs');

"use strict";
const reportMetaAnalytics = (payload) => {
  EchoSrv.getEchoSrv().addEvent({
    type: EchoSrv.EchoEventType.MetaAnalytics,
    payload
  });
};
const reportPageview = () => {
  var _a;
  const location = LocationService.locationService.getLocation();
  const page = `${(_a = config.config.appSubUrl) != null ? _a : ""}${location.pathname}${location.search}${location.hash}`;
  EchoSrv.getEchoSrv().addEvent({
    type: EchoSrv.EchoEventType.Pageview,
    payload: {
      page
    }
  });
};
const reportInteraction = (interactionName, properties) => {
  if (config.config.reportingStaticContext && config.config.reportingStaticContext instanceof Object) {
    properties = { ...properties, ...config.config.reportingStaticContext };
  }
  EchoSrv.getEchoSrv().addEvent({
    type: EchoSrv.EchoEventType.Interaction,
    payload: {
      interactionName,
      properties
    }
  });
};
const reportExperimentView = (id, group, variant) => {
  EchoSrv.getEchoSrv().addEvent({
    type: EchoSrv.EchoEventType.ExperimentView,
    payload: {
      experimentId: id,
      experimentGroup: group,
      experimentVariant: variant
    }
  });
};

exports.reportExperimentView = reportExperimentView;
exports.reportInteraction = reportInteraction;
exports.reportMetaAnalytics = reportMetaAnalytics;
exports.reportPageview = reportPageview;
