// @grafana/ui/dist/cjs/components/TagsInput/TagItem.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var tags = require('../../utils/tags.cjs');
var IconButton = require('../IconButton/IconButton.cjs');

"use strict";
const TagItem = ({ name, disabled, onRemove, autoColors = true }) => {
  const styles = ThemeContext.useStyles2(getStyles);
  const tagStyle = React.useMemo(() => {
    if (autoColors) {
      const { color, borderColor } = tags.getTagColorsFromName(name);
      return { backgroundColor: color, borderColor };
    }
    return void 0;
  }, [name, autoColors]);
  return /* @__PURE__ */ jsxRuntime.jsxs("li", { className: css.cx(styles.itemStyle, !tagStyle && styles.defaultTagColor), style: tagStyle, children: [
    /* @__PURE__ */ jsxRuntime.jsx("span", { className: styles.nameStyle, children: name }),
    /* @__PURE__ */ jsxRuntime.jsx(
      IconButton.IconButton,
      {
        name: "times",
        size: "lg",
        disabled,
        tooltip: i18n.t("grafana-ui.tags-input.remove", "Remove tag: {{name}}", { name }),
        onClick: () => onRemove(name),
        className: styles.buttonStyles
      }
    )
  ] });
};
const getStyles = (theme) => {
  const height = theme.spacing.gridSize * 3;
  return {
    itemStyle: css.css({
      display: "flex",
      gap: "3px",
      alignItems: "center",
      height: `${height}px`,
      lineHeight: `${height - 2}px`,
      borderWidth: "1px",
      borderStyle: "solid",
      borderRadius: theme.shape.radius.default,
      padding: `0 ${theme.spacing(0.5)}`,
      whiteSpace: "nowrap",
      textShadow: "none",
      fontWeight: 500,
      fontSize: theme.typography.size.sm,
      color: "#fff"
    }),
    defaultTagColor: css.css({
      backgroundColor: theme.colors.background.secondary,
      borderColor: theme.components.input.borderColor,
      color: theme.colors.text.primary
    }),
    nameStyle: css.css({
      maxWidth: "25ch",
      textOverflow: "ellipsis",
      overflow: "hidden"
    }),
    buttonStyles: css.css({
      margin: 0,
      "&:hover::before": {
        display: "none"
      }
    })
  };
};

exports.TagItem = TagItem;
