// @grafana/ui/dist/cjs/components/TagsInput/TagsInput.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var Button = require('../Button/Button.cjs');
var Input = require('../Input/Input.cjs');
var TagItem = require('./TagItem.cjs');

"use strict";
const TagsInput = React.forwardRef(
  ({
    placeholder: placeholderProp,
    tags = [],
    onChange,
    width,
    className,
    disabled,
    addOnBlur,
    invalid,
    id,
    autoColors = true
  }, ref) => {
    const placeholder = placeholderProp != null ? placeholderProp : i18n.t("grafana-ui.tags-input.placeholder-new-tag", "New tag (enter key to add)");
    const [newTagName, setNewTagName] = React.useState("");
    const styles = ThemeContext.useStyles2(getStyles);
    const theme = ThemeContext.useTheme2();
    const isTagTooLong = newTagName.length > 50;
    const onNameChange = React.useCallback((event) => {
      setNewTagName(event.target.value);
    }, []);
    const onRemove = (tagToRemove) => {
      onChange(tags.filter((x) => x !== tagToRemove));
    };
    const onAdd = (event) => {
      event == null ? void 0 : event.preventDefault();
      if (newTagName.length > 50) {
        return;
      }
      if (!tags.includes(newTagName)) {
        onChange(tags.concat(newTagName));
      }
      setNewTagName("");
    };
    const onBlur = () => {
      if (addOnBlur && newTagName) {
        onAdd();
      }
    };
    const onKeyboardAdd = (event) => {
      if (event.key === "Enter" && newTagName !== "") {
        onAdd(event);
      }
    };
    return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: css.cx(styles.wrapper, className, width ? css.css({ width: theme.spacing(width) }) : ""), children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        Input.Input,
        {
          ref,
          id,
          disabled,
          placeholder,
          onChange: onNameChange,
          value: newTagName,
          onKeyDown: onKeyboardAdd,
          onBlur,
          invalid: invalid || isTagTooLong,
          suffix: /* @__PURE__ */ jsxRuntime.jsx(
            Button.Button,
            {
              fill: "text",
              className: styles.addButtonStyle,
              onClick: onAdd,
              size: "md",
              disabled: newTagName.length <= 0 || isTagTooLong,
              title: isTagTooLong ? i18n.t("grafana-ui.tags-input.tag-too-long", "Tag too long, max 50 characters") : void 0,
              children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.tags-input.add", children: "Add" })
            }
          )
        }
      ),
      (tags == null ? void 0 : tags.length) > 0 && /* @__PURE__ */ jsxRuntime.jsx("ul", { className: styles.tags, children: tags.map((tag) => /* @__PURE__ */ jsxRuntime.jsx(TagItem.TagItem, { name: tag, onRemove, disabled, autoColors }, tag)) })
    ] });
  }
);
TagsInput.displayName = "TagsInput";
const getStyles = (theme) => ({
  wrapper: css.css({
    minHeight: theme.spacing(4),
    display: "flex",
    flexDirection: "column",
    gap: theme.spacing(1),
    flexWrap: "wrap"
  }),
  tags: css.css({
    display: "flex",
    justifyContent: "flex-start",
    flexWrap: "wrap",
    gap: theme.spacing(0.5)
  }),
  addButtonStyle: css.css({
    margin: `0 -${theme.spacing(1)}`
  })
});

exports.TagsInput = TagsInput;
