// @grafana/ui/dist/cjs/components/TextArea/TextArea.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var commonStyles = require('../Forms/commonStyles.cjs');

"use strict";
const TextArea = React.forwardRef(({ invalid, className, ...props }, ref) => {
  const styles = ThemeContext.useStyles2(getTextAreaStyle, invalid);
  return /* @__PURE__ */ jsxRuntime.jsx("textarea", { ...props, className: css.cx(styles.textarea, className), ref });
});
const getTextAreaStyle = (theme, invalid = false) => ({
  textarea: css.cx(
    commonStyles.sharedInputStyle(theme),
    commonStyles.getFocusStyle(theme),
    css.css({
      display: "block",
      borderRadius: theme.shape.radius.default,
      padding: `${theme.spacing.gridSize / 4}px ${theme.spacing.gridSize}px`,
      width: "100%",
      borderColor: invalid ? theme.colors.error.border : theme.components.input.borderColor
    })
  )
});
TextArea.displayName = "TextArea";

exports.TextArea = TextArea;
