// @grafana/ui/dist/cjs/components/Table/TableRT/FilterPopup.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../../themes/ThemeContext.cjs');
var Button = require('../../Button/Button.cjs');
var ClickOutsideWrapper = require('../../ClickOutsideWrapper/ClickOutsideWrapper.cjs');
var Label = require('../../Forms/Label.cjs');
var IconButton = require('../../IconButton/IconButton.cjs');
var Stack = require('../../Layout/Stack/Stack.cjs');
var utils = require('../utils.cjs');
var FilterList = require('./FilterList.cjs');

"use strict";
const FilterPopup = ({
  column: { preFilteredRows, filterValue, setFilter },
  onClose,
  field,
  searchFilter,
  setSearchFilter,
  operator,
  setOperator
}) => {
  const theme = ThemeContext.useTheme2();
  const uniqueValues = React.useMemo(() => utils.calculateUniqueFieldValues(preFilteredRows, field), [preFilteredRows, field]);
  const options = React.useMemo(() => utils.valuesToOptions(uniqueValues), [uniqueValues]);
  const filteredOptions = React.useMemo(() => utils.getFilteredOptions(options, filterValue), [options, filterValue]);
  const [values, setValues] = React.useState(filteredOptions);
  const [matchCase, setMatchCase] = React.useState(false);
  const ref = React.useRef(null);
  const onCancel = React.useCallback(() => onClose(), [onClose]);
  const onFilter = React.useCallback(
    (event) => {
      const filtered = values.length ? values : void 0;
      setFilter(filtered);
      onClose();
    },
    [setFilter, values, onClose]
  );
  const onClearFilter = React.useCallback(
    (event) => {
      setFilter(void 0);
      onClose();
    },
    [setFilter, onClose]
  );
  const clearFilterVisible = React.useMemo(() => filterValue !== void 0, [filterValue]);
  const styles = ThemeContext.useStyles2(getStyles);
  return /* @__PURE__ */ jsxRuntime.jsx(ClickOutsideWrapper.ClickOutsideWrapper, { onClick: onCancel, useCapture: true, children: /* @__PURE__ */ jsxRuntime.jsx("div", { ref, className: css.cx(styles.filterContainer), onClick: stopPropagation, children: /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { direction: "column", gap: 3, children: [
    /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { direction: "column", gap: 0.5, children: [
      /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { justifyContent: "space-between", alignItems: "center", children: [
        /* @__PURE__ */ jsxRuntime.jsx(Label.Label, { className: styles.label, children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.table.filter-popup-heading", children: "Filter by values:" }) }),
        /* @__PURE__ */ jsxRuntime.jsx(
          IconButton.IconButton,
          {
            name: "text-fields",
            tooltip: i18n.t("grafana-ui.table.filter-popup-match-case", "Match case"),
            style: { color: matchCase ? theme.colors.text.link : theme.colors.text.disabled },
            onClick: () => {
              setMatchCase((s) => !s);
            }
          }
        )
      ] }),
      /* @__PURE__ */ jsxRuntime.jsx("div", { className: css.cx(styles.listDivider) }),
      ref.current && /* @__PURE__ */ jsxRuntime.jsx(
        FilterList.FilterList,
        {
          referenceElement: ref.current,
          onChange: setValues,
          values,
          options,
          caseSensitive: matchCase,
          showOperators: true,
          searchFilter,
          setSearchFilter,
          operator,
          setOperator
        }
      )
    ] }),
    /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { gap: 3, children: [
      /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { children: [
        /* @__PURE__ */ jsxRuntime.jsx(Button.Button, { size: "sm", onClick: onFilter, children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.table.filter-popup-apply", children: "Ok" }) }),
        /* @__PURE__ */ jsxRuntime.jsx(Button.Button, { size: "sm", variant: "secondary", onClick: onCancel, children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.table.filter-popup-cancel", children: "Cancel" }) })
      ] }),
      clearFilterVisible && /* @__PURE__ */ jsxRuntime.jsx(Stack.Stack, { children: /* @__PURE__ */ jsxRuntime.jsx(Button.Button, { fill: "text", size: "sm", onClick: onClearFilter, children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.table.filter-popup-clear", children: "Clear filter" }) }) })
    ] })
  ] }) }) });
};
const getStyles = (theme) => ({
  filterContainer: css.css({
    label: "filterContainer",
    width: "100%",
    minWidth: "250px",
    height: "100%",
    maxHeight: "400px",
    backgroundColor: theme.colors.background.primary,
    border: `1px solid ${theme.colors.border.weak}`,
    padding: theme.spacing(2),
    boxShadow: theme.shadows.z3,
    borderRadius: theme.shape.radius.default
  }),
  listDivider: css.css({
    label: "listDivider",
    width: "100%",
    borderTop: `1px solid ${theme.colors.border.medium}`,
    padding: theme.spacing(0.5, 2)
  }),
  label: css.css({
    marginBottom: 0
  })
});
const stopPropagation = (event) => {
  event.stopPropagation();
};

exports.FilterPopup = FilterPopup;
