// @grafana/ui/dist/cjs/components/List/AbstractList.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var stylesFactory = require('../../themes/stylesFactory.cjs');

"use strict";
const getStyles = stylesFactory.stylesFactory((inlineList = false) => ({
  list: css.css({
    listStyleType: "none",
    margin: 0,
    padding: 0
  }),
  item: css.css({
    display: inlineList && "inline-block" || "block"
  })
}));
class AbstractList extends React.PureComponent {
  constructor(props) {
    super(props);
  }
  render() {
    const { items, renderItem, getItemKey, className, inline } = this.props;
    const styles = getStyles(inline);
    return /* @__PURE__ */ jsxRuntime.jsx("ul", { className: css.cx(styles.list, className), children: items.map((item, i) => {
      return /* @__PURE__ */ jsxRuntime.jsx("li", { className: styles.item, children: renderItem(item, i) }, getItemKey ? getItemKey(item) : i);
    }) });
  }
}

exports.AbstractList = AbstractList;
