// @grafana/ui/dist/cjs/components/Forms/InlineField.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var reactUtils = require('../../utils/reactUtils.cjs');
var FieldValidationMessage = require('./FieldValidationMessage.cjs');
var InlineLabel = require('./InlineLabel.cjs');

"use strict";
const InlineField = ({
  children,
  label,
  tooltip,
  labelWidth = "auto",
  invalid,
  loading,
  disabled,
  required,
  className,
  htmlFor,
  grow,
  shrink,
  error,
  transparent,
  interactive,
  validationMessageHorizontalOverflow,
  ...htmlProps
}) => {
  const theme = ThemeContext.useTheme2();
  const styles = getStyles(theme, grow, shrink);
  const inputId = htmlFor != null ? htmlFor : reactUtils.getChildId(children);
  const labelElement = typeof label === "string" ? /* @__PURE__ */ jsxRuntime.jsx(
    InlineLabel.InlineLabel,
    {
      interactive,
      width: labelWidth,
      tooltip,
      htmlFor: inputId,
      transparent,
      children: `${label}${required ? " *" : ""}`
    }
  ) : label;
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: css.cx(styles.container, className), ...htmlProps, children: [
    labelElement,
    /* @__PURE__ */ jsxRuntime.jsxs("div", { className: styles.childContainer, children: [
      React.cloneElement(children, { invalid, disabled, loading }),
      invalid && error && /* @__PURE__ */ jsxRuntime.jsx(
        "div",
        {
          className: css.cx(styles.fieldValidationWrapper, {
            [styles.validationMessageHorizontalOverflow]: !!validationMessageHorizontalOverflow
          }),
          children: /* @__PURE__ */ jsxRuntime.jsx(FieldValidationMessage.FieldValidationMessage, { children: error })
        }
      )
    ] })
  ] });
};
InlineField.displayName = "InlineField";
const getStyles = (theme, grow, shrink) => {
  return {
    container: css.css({
      display: "flex",
      flexDirection: "row",
      alignItems: "flex-start",
      textAlign: "left",
      position: "relative",
      flex: `${grow ? 1 : 0} ${shrink ? 1 : 0} auto`,
      margin: `0 ${theme.spacing(0.5)} ${theme.spacing(0.5)} 0`
    }),
    childContainer: css.css({
      flex: `${grow ? 1 : 0} ${shrink ? 1 : 0} auto`
    }),
    fieldValidationWrapper: css.css({
      marginTop: theme.spacing(0.5)
    }),
    validationMessageHorizontalOverflow: css.css({
      width: 0,
      overflowX: "visible",
      "& > *": {
        whiteSpace: "nowrap"
      }
    })
  };
};

exports.InlineField = InlineField;
