// @grafana/ui/dist/cjs/components/DataSourceSettings/BasicAuthSettings.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var i18n = require('@grafana/i18n');
var InlineField = require('../Forms/InlineField.cjs');
var FormField = require('../FormField/FormField.cjs');
var SecretFormField = require('../SecretFormField/SecretFormField.cjs');

"use strict";
const BasicAuthSettings = ({ dataSourceConfig, onChange }) => {
  const password = dataSourceConfig.secureJsonData ? dataSourceConfig.secureJsonData.basicAuthPassword : "";
  const onPasswordReset = () => {
    onChange({
      ...dataSourceConfig,
      secureJsonData: {
        ...dataSourceConfig.secureJsonData,
        basicAuthPassword: ""
      },
      secureJsonFields: {
        ...dataSourceConfig.secureJsonFields,
        basicAuthPassword: false
      }
    });
  };
  const onPasswordChange = (event) => {
    onChange({
      ...dataSourceConfig,
      secureJsonData: {
        ...dataSourceConfig.secureJsonData,
        basicAuthPassword: event.currentTarget.value
      }
    });
  };
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(InlineField.InlineField, { disabled: dataSourceConfig.readOnly, children: /* @__PURE__ */ jsxRuntime.jsx(
      FormField.FormField,
      {
        label: i18n.t("grafana-ui.data-source-basic-auth-settings.user-label", "User"),
        labelWidth: 10,
        inputWidth: 18,
        placeholder: i18n.t("grafana-ui.data-source-basic-auth-settings.user-placeholder", "user"),
        value: dataSourceConfig.basicAuthUser,
        onChange: (event) => onChange({ ...dataSourceConfig, basicAuthUser: event.currentTarget.value })
      }
    ) }),
    /* @__PURE__ */ jsxRuntime.jsx(InlineField.InlineField, { disabled: dataSourceConfig.readOnly, children: /* @__PURE__ */ jsxRuntime.jsx(
      SecretFormField.SecretFormField,
      {
        isConfigured: !!(dataSourceConfig.secureJsonFields && dataSourceConfig.secureJsonFields.basicAuthPassword),
        value: password || "",
        inputWidth: 18,
        labelWidth: 10,
        onReset: onPasswordReset,
        onChange: onPasswordChange
      }
    ) })
  ] });
};

exports.BasicAuthSettings = BasicAuthSettings;
