// @grafana/ui/dist/esm/components/MatchersUI/FieldTypeMatcherEditor.mjs
import { jsx } from 'react/jsx-runtime';
import { memo, useCallback, useMemo } from 'react';
import { FieldType, fieldMatchers, FieldMatcherID } from '@grafana/data';
import { t } from '@grafana/i18n';
import { getFieldTypeIconName } from '../../types/icon.mjs';
import { Select } from '../Select/Select.mjs';

"use strict";
const FieldTypeMatcherEditor = memo((props) => {
  const { data, options, onChange: onChangeFromProps, id } = props;
  const counts = useFieldCounts(data);
  const selectOptions = useSelectOptions(counts, options);
  const onChange = useCallback(
    (selection) => {
      return onChangeFromProps(selection.value);
    },
    [onChangeFromProps]
  );
  const selectedOption = selectOptions.find((v) => v.value === options);
  return /* @__PURE__ */ jsx(Select, { inputId: id, value: selectedOption, options: selectOptions, onChange });
});
FieldTypeMatcherEditor.displayName = "FieldTypeMatcherEditor";
const getAllFieldTypeIconOptions = () => [
  {
    value: FieldType.number,
    label: t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-number", "Number"),
    icon: getFieldTypeIconName(FieldType.number)
  },
  {
    value: FieldType.string,
    label: t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-string", "String"),
    icon: getFieldTypeIconName(FieldType.string)
  },
  {
    value: FieldType.time,
    label: t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-time", "Time"),
    icon: getFieldTypeIconName(FieldType.time)
  },
  {
    value: FieldType.boolean,
    label: t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-boolean", "Boolean"),
    icon: getFieldTypeIconName(FieldType.boolean)
  },
  {
    value: FieldType.trace,
    label: t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-traces", "Traces"),
    icon: getFieldTypeIconName(FieldType.trace)
  },
  {
    value: FieldType.enum,
    label: t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-enum", "Enum"),
    icon: getFieldTypeIconName(FieldType.enum)
  },
  {
    value: FieldType.other,
    label: t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-other", "Other"),
    icon: getFieldTypeIconName(FieldType.other)
  }
];
const useFieldCounts = (data) => {
  return useMemo(() => {
    const counts = /* @__PURE__ */ new Map();
    for (const t2 of getAllFieldTypeIconOptions()) {
      counts.set(t2.value, 0);
    }
    for (const frame of data) {
      for (const field of frame.fields) {
        const key = field.type || FieldType.other;
        let v = counts.get(key);
        if (!v) {
          v = 0;
        }
        counts.set(key, v + 1);
      }
    }
    return counts;
  }, [data]);
};
const useSelectOptions = (counts, opt) => {
  return useMemo(() => {
    let found = false;
    const options = [];
    for (const t2 of getAllFieldTypeIconOptions()) {
      const count = counts.get(t2.value);
      const match = opt === t2.value;
      if (count || match) {
        options.push({
          ...t2,
          label: `${t2.label} (${counts.get(t2.value)})`
        });
      }
      if (match) {
        found = true;
      }
    }
    if (opt && !found) {
      options.push({
        value: opt,
        label: `${opt} (No matches)`
      });
    }
    return options;
  }, [counts, opt]);
};
const getFieldTypeMatcherItem = () => ({
  id: FieldMatcherID.byType,
  component: FieldTypeMatcherEditor,
  matcher: fieldMatchers.get(FieldMatcherID.byType),
  name: t("grafana-ui.matchers-ui.name-fields-with-type", "Fields with type"),
  description: t(
    "grafana-ui.matchers-ui.description-fields-with-type",
    "Set properties for fields of a specific type (number, string, boolean)"
  ),
  optionsToLabel: (options) => options
});

export { FieldTypeMatcherEditor, getAllFieldTypeIconOptions, getFieldTypeMatcherItem };
