// @grafana/runtime/dist/esm/services/pluginMeta/apps.mjs
import { config } from '../../config.mjs';
import { getFeatureFlagClient } from '../../internal/openFeature/index.mjs';
import { getAppPluginMapper } from './mappers/mappers.mjs';
import { initPluginMetas } from './plugins.mjs';

"use strict";
let apps = {};
function initialized() {
  return Boolean(Object.keys(apps).length);
}
async function initAppPluginMetas() {
  if (!getFeatureFlagClient().getBooleanValue("useMTPlugins", false)) {
    apps = config.apps;
    return;
  }
  const metas = await initPluginMetas();
  const mapper = getAppPluginMapper();
  apps = mapper(metas);
}
async function getAppPluginMetas() {
  if (!initialized()) {
    await initAppPluginMetas();
  }
  return Object.values(structuredClone(apps));
}
async function getAppPluginMeta(pluginId) {
  if (!initialized()) {
    await initAppPluginMetas();
  }
  const app = apps[pluginId];
  return app ? structuredClone(app) : null;
}
async function isAppPluginInstalled(pluginId) {
  const app = await getAppPluginMeta(pluginId);
  return Boolean(app);
}
async function getAppPluginVersion(pluginId) {
  var _a;
  const app = await getAppPluginMeta(pluginId);
  return (_a = app == null ? void 0 : app.version) != null ? _a : null;
}
function setAppPluginMetas(override) {
  if (process.env.NODE_ENV !== "test") {
    throw new Error("setAppPluginMetas() function can only be called from tests.");
  }
  apps = structuredClone(override);
}

export { getAppPluginMeta, getAppPluginMetas, getAppPluginVersion, isAppPluginInstalled, setAppPluginMetas };
