// @grafana/runtime/dist/esm/services/pluginMeta/hooks.mjs
import { useAsync } from 'react-use';
import { getAppPluginMetas, getAppPluginMeta, isAppPluginInstalled, getAppPluginVersion } from './apps.mjs';
import { getPanelPluginMetas, getPanelPluginMeta, isPanelPluginInstalled, getPanelPluginVersion, getListedPanelPluginIds } from './panels.mjs';

"use strict";
function useAppPluginMetas() {
  const { loading, error, value } = useAsync(async () => getAppPluginMetas());
  return { loading, error, value };
}
function useAppPluginMeta(pluginId) {
  const { loading, error, value } = useAsync(async () => getAppPluginMeta(pluginId));
  return { loading, error, value };
}
function usePanelPluginMetas() {
  const { loading, error, value } = useAsync(async () => getPanelPluginMetas());
  return { loading, error, value };
}
function usePanelPluginMeta(pluginId) {
  const { loading, error, value } = useAsync(async () => getPanelPluginMeta(pluginId));
  return { loading, error, value };
}
function useAppPluginInstalled(pluginId) {
  const { loading, error, value } = useAsync(async () => isAppPluginInstalled(pluginId));
  return { loading, error, value };
}
function useAppPluginVersion(pluginId) {
  const { loading, error, value } = useAsync(async () => getAppPluginVersion(pluginId));
  return { loading, error, value };
}
function usePanelPluginInstalled(pluginId) {
  const { loading, error, value } = useAsync(async () => isPanelPluginInstalled(pluginId));
  return { loading, error, value };
}
function usePanelPluginVersion(pluginId) {
  const { loading, error, value } = useAsync(async () => getPanelPluginVersion(pluginId));
  return { loading, error, value };
}
function useListedPanelPluginIds() {
  const { loading, error, value } = useAsync(async () => getListedPanelPluginIds());
  return { loading, error, value };
}

export { useAppPluginInstalled, useAppPluginMeta, useAppPluginMetas, useAppPluginVersion, useListedPanelPluginIds, usePanelPluginInstalled, usePanelPluginMeta, usePanelPluginMetas, usePanelPluginVersion };
