// @grafana/runtime/dist/esm/services/pluginMeta/panels.mjs
import { config } from '../../config.mjs';
import { getFeatureFlagClient } from '../../internal/openFeature/index.mjs';
import { getPanelPluginMapper } from './mappers/mappers.mjs';
import { initPluginMetas } from './plugins.mjs';

"use strict";
let panels = {};
function initialized() {
  return Boolean(Object.keys(panels).length);
}
async function initPanelPluginMetas() {
  if (!getFeatureFlagClient().getBooleanValue("useMTPlugins", false)) {
    panels = config.panels;
    return;
  }
  const metas = await initPluginMetas();
  const mapper = getPanelPluginMapper();
  panels = mapper(metas);
}
async function getPanelPluginMetas() {
  if (!initialized()) {
    await initPanelPluginMetas();
  }
  return Object.values(structuredClone(panels));
}
async function getPanelPluginMeta(pluginId) {
  if (!initialized()) {
    await initPanelPluginMetas();
  }
  const panel = panels[pluginId];
  return panel ? structuredClone(panel) : null;
}
async function isPanelPluginInstalled(pluginId) {
  const panel = await getPanelPluginMeta(pluginId);
  return Boolean(panel);
}
async function getPanelPluginVersion(pluginId) {
  var _a;
  const panel = await getPanelPluginMeta(pluginId);
  return (_a = panel == null ? void 0 : panel.info.version) != null ? _a : null;
}
async function getListedPanelPluginIds() {
  const panels2 = await getPanelPluginMetas();
  return panels2.filter((p) => p.hideFromList === false).map((p) => p.id);
}
function setPanelPluginMetas(override) {
  if (process.env.NODE_ENV !== "test") {
    throw new Error("setPanelPluginMetas() function can only be called from tests.");
  }
  panels = structuredClone(override);
}

export { getListedPanelPluginIds, getPanelPluginMeta, getPanelPluginMetas, getPanelPluginVersion, isPanelPluginInstalled, setPanelPluginMetas };
