// @grafana/ui/dist/cjs/components/MatchersUI/FieldNameByRegexMatcherEditor.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var data = require('@grafana/data');
var i18n = require('@grafana/i18n');
var Input = require('../Input/Input.cjs');

"use strict";
const FieldNameByRegexMatcherEditor = React.memo((props) => {
  const { options, onChange } = props;
  const onBlur = React.useCallback(
    (e) => {
      return onChange(e.target.value);
    },
    [onChange]
  );
  return /* @__PURE__ */ jsxRuntime.jsx(
    Input.Input,
    {
      placeholder: i18n.t("grafana-ui.field-name-by-regex-matcher.input-placeholder", "Enter regular expression"),
      defaultValue: options,
      onBlur
    }
  );
});
FieldNameByRegexMatcherEditor.displayName = "FieldNameByRegexMatcherEditor";
const getFieldNameByRegexMatcherItem = () => ({
  id: data.FieldMatcherID.byRegexp,
  component: FieldNameByRegexMatcherEditor,
  matcher: data.fieldMatchers.get(data.FieldMatcherID.byRegexp),
  name: i18n.t("grafana-ui.matchers-ui.name-field-name-by-regex-matcher", "Fields with name matching regex"),
  description: i18n.t(
    "grafana-ui.matchers-ui.description-field-name-by-regex-matcher",
    "Set properties for fields with names matching a regex"
  ),
  optionsToLabel: (options) => options
});

exports.FieldNameByRegexMatcherEditor = FieldNameByRegexMatcherEditor;
exports.getFieldNameByRegexMatcherItem = getFieldNameByRegexMatcherItem;
