// @grafana/ui/dist/cjs/components/MatchersUI/FieldNameMatcherEditor.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var data = require('@grafana/data');
var i18n = require('@grafana/i18n');
var Select = require('../Select/Select.cjs');
var utils = require('./utils.cjs');

"use strict";
const FieldNameMatcherEditor = React.memo((props) => {
  const { data, options, onChange: onChangeFromProps, id } = props;
  const names = utils.useFieldDisplayNames(data);
  const selectOptions = utils.useSelectOptions(names, options);
  const onChange = React.useCallback(
    (selection) => {
      if (!utils.frameHasName(selection.value, names)) {
        return;
      }
      return onChangeFromProps(selection.value);
    },
    [names, onChangeFromProps]
  );
  const selectedOption = selectOptions.find((v) => v.value === options);
  return /* @__PURE__ */ jsxRuntime.jsx(Select.Select, { value: selectedOption, options: selectOptions, onChange, inputId: id });
});
FieldNameMatcherEditor.displayName = "FieldNameMatcherEditor";
const getFieldNameMatcherItem = () => ({
  id: data.FieldMatcherID.byName,
  component: FieldNameMatcherEditor,
  matcher: data.fieldMatchers.get(data.FieldMatcherID.byName),
  name: i18n.t("grafana-ui.matchers-ui.name-fields-with-name", "Fields with name"),
  description: i18n.t("grafana-ui.matchers-ui.description-fields-with-name", "Set properties for a specific field"),
  optionsToLabel: (options) => options
});

exports.FieldNameMatcherEditor = FieldNameMatcherEditor;
exports.getFieldNameMatcherItem = getFieldNameMatcherItem;
