// @grafana/ui/dist/cjs/components/MatchersUI/FieldTypeMatcherEditor.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var data = require('@grafana/data');
var i18n = require('@grafana/i18n');
var icon = require('../../types/icon.cjs');
var Select = require('../Select/Select.cjs');

"use strict";
const FieldTypeMatcherEditor = React.memo((props) => {
  const { data, options, onChange: onChangeFromProps, id } = props;
  const counts = useFieldCounts(data);
  const selectOptions = useSelectOptions(counts, options);
  const onChange = React.useCallback(
    (selection) => {
      return onChangeFromProps(selection.value);
    },
    [onChangeFromProps]
  );
  const selectedOption = selectOptions.find((v) => v.value === options);
  return /* @__PURE__ */ jsxRuntime.jsx(Select.Select, { inputId: id, value: selectedOption, options: selectOptions, onChange });
});
FieldTypeMatcherEditor.displayName = "FieldTypeMatcherEditor";
const getAllFieldTypeIconOptions = () => [
  {
    value: data.FieldType.number,
    label: i18n.t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-number", "Number"),
    icon: icon.getFieldTypeIconName(data.FieldType.number)
  },
  {
    value: data.FieldType.string,
    label: i18n.t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-string", "String"),
    icon: icon.getFieldTypeIconName(data.FieldType.string)
  },
  {
    value: data.FieldType.time,
    label: i18n.t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-time", "Time"),
    icon: icon.getFieldTypeIconName(data.FieldType.time)
  },
  {
    value: data.FieldType.boolean,
    label: i18n.t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-boolean", "Boolean"),
    icon: icon.getFieldTypeIconName(data.FieldType.boolean)
  },
  {
    value: data.FieldType.trace,
    label: i18n.t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-traces", "Traces"),
    icon: icon.getFieldTypeIconName(data.FieldType.trace)
  },
  {
    value: data.FieldType.enum,
    label: i18n.t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-enum", "Enum"),
    icon: icon.getFieldTypeIconName(data.FieldType.enum)
  },
  {
    value: data.FieldType.other,
    label: i18n.t("grafana-ui.matchers-ui.get-all-field-type-icon-options.label-other", "Other"),
    icon: icon.getFieldTypeIconName(data.FieldType.other)
  }
];
const useFieldCounts = (data$1) => {
  return React.useMemo(() => {
    const counts = /* @__PURE__ */ new Map();
    for (const t2 of getAllFieldTypeIconOptions()) {
      counts.set(t2.value, 0);
    }
    for (const frame of data$1) {
      for (const field of frame.fields) {
        const key = field.type || data.FieldType.other;
        let v = counts.get(key);
        if (!v) {
          v = 0;
        }
        counts.set(key, v + 1);
      }
    }
    return counts;
  }, [data$1]);
};
const useSelectOptions = (counts, opt) => {
  return React.useMemo(() => {
    let found = false;
    const options = [];
    for (const t2 of getAllFieldTypeIconOptions()) {
      const count = counts.get(t2.value);
      const match = opt === t2.value;
      if (count || match) {
        options.push({
          ...t2,
          label: `${t2.label} (${counts.get(t2.value)})`
        });
      }
      if (match) {
        found = true;
      }
    }
    if (opt && !found) {
      options.push({
        value: opt,
        label: `${opt} (No matches)`
      });
    }
    return options;
  }, [counts, opt]);
};
const getFieldTypeMatcherItem = () => ({
  id: data.FieldMatcherID.byType,
  component: FieldTypeMatcherEditor,
  matcher: data.fieldMatchers.get(data.FieldMatcherID.byType),
  name: i18n.t("grafana-ui.matchers-ui.name-fields-with-type", "Fields with type"),
  description: i18n.t(
    "grafana-ui.matchers-ui.description-fields-with-type",
    "Set properties for fields of a specific type (number, string, boolean)"
  ),
  optionsToLabel: (options) => options
});

exports.FieldTypeMatcherEditor = FieldTypeMatcherEditor;
exports.getAllFieldTypeIconOptions = getAllFieldTypeIconOptions;
exports.getFieldTypeMatcherItem = getFieldTypeMatcherItem;
