// @grafana/runtime/dist/cjs/utils/getCachedPromise.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var logging = require('./logging.cjs');

"use strict";
const cache = /* @__PURE__ */ new Map();
const MAX_CACHE_SIZE = 500;
let logger;
function getLogger() {
  if (!logger) {
    logger = logging.createMonitoringLogger("get-cached-promise-logs");
  }
  return logger;
}
function setLogger(override) {
  if (process.env.NODE_ENV !== "test") {
    throw new Error("setLogger function can only be called from tests.");
  }
  logger = override;
}
function logError({ error, key }) {
  const err = error instanceof Error ? error : new Error(String(error));
  const context = { message: err.message, key };
  if (err.stack) {
    context.stack = err.stack;
  }
  getLogger().logError(new Error(`Something failed while resolving a cached promise`), context);
}
function checkCacheSize() {
  if (cache.size <= MAX_CACHE_SIZE) {
    return;
  }
  cache.clear();
}
function addToCache(key, cached) {
  checkCacheSize();
  cache.set(key, cached);
}
function cachePromiseWithoutCatch({ key, promise }) {
  const cached = promise();
  addToCache(key, cached);
  return cached;
}
function cachePromiseWithDefaultValue({ defaultValue, key, promise }) {
  const cached = promise().catch((error) => {
    logError({ error, key });
    cache.delete(key);
    return defaultValue;
  });
  addToCache(key, cached);
  return cached;
}
function cachePromiseWithCallback({ key, promise, onError }) {
  const invalidate = () => cache.delete(key);
  const cached = promise().catch((error) => onError({ error, invalidate }));
  addToCache(key, cached);
  return cached;
}
function getCachedPromise(promise, options) {
  const { cacheKey, defaultValue, onError } = options != null ? options : {};
  const key = cacheKey != null ? cacheKey : promise.name;
  if (!key) {
    return Promise.reject(new Error(`getCachedPromise function must be invoked with a named function or cacheKey`));
  }
  const cached = cache.get(key);
  if (cached) {
    return cached;
  }
  if (onError) {
    return cachePromiseWithCallback({ key, onError, promise });
  }
  if (defaultValue !== void 0) {
    return cachePromiseWithDefaultValue({ defaultValue, key, promise });
  }
  return cachePromiseWithoutCatch({ key, promise });
}
function invalidateCache() {
  if (process.env.NODE_ENV !== "test") {
    throw new Error("invalidateCache function can only be called from tests.");
  }
  cache.clear();
}

exports.MAX_CACHE_SIZE = MAX_CACHE_SIZE;
exports.getCachedPromise = getCachedPromise;
exports.invalidateCache = invalidateCache;
exports.setLogger = setLogger;
