// @grafana/runtime/dist/cjs/utils/migrationHandler.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var config = require('../config.cjs');
require('../services/index.cjs');
var DataSourceWithBackend = require('./DataSourceWithBackend.cjs');
var backendSrv = require('../services/backendSrv.cjs');

"use strict";
function isMigrationHandler(object) {
  return object instanceof DataSourceWithBackend.DataSourceWithBackend && "hasBackendMigration" in object && "shouldMigrate" in object;
}
async function postMigrateRequest(queries) {
  var _a, _b;
  if (!(config.config.featureToggles.grafanaAPIServerWithExperimentalAPIs || config.config.featureToggles.datasourceAPIServers)) {
    console.warn("migrateQuery is only available with the experimental API server");
    return queries;
  }
  const dsnameURL = (_b = (_a = queries[0].datasource) == null ? void 0 : _a.type) == null ? void 0 : _b.replace(/^(grafana-)?(.*?)(-datasource)?$/, "$2");
  const groupName = `${dsnameURL}.datasource.grafana.app`;
  const apiVersion = "v0alpha1";
  const url = `/apis/${groupName}/${apiVersion}/namespaces/${config.config.namespace}/queryconvert`;
  const request = {
    queries: queries.map((query) => {
      return {
        ...query,
        JSON: query
        // JSON is not part of the type but it should be what holds the query
      };
    })
  };
  const res = await backendSrv.getBackendSrv().post(url, request);
  return res.queries.map((query) => query.JSON);
}
async function migrateQuery(datasource, query) {
  if (!datasource.hasBackendMigration || !datasource.shouldMigrate(query)) {
    return query;
  }
  const res = await postMigrateRequest([query]);
  return res[0];
}
async function migrateRequest(datasource, request) {
  if (!datasource.hasBackendMigration || !request.targets.some((query) => datasource.shouldMigrate(query))) {
    return request;
  }
  const res = await postMigrateRequest(request.targets);
  return { ...request, targets: res };
}

exports.isMigrationHandler = isMigrationHandler;
exports.migrateQuery = migrateQuery;
exports.migrateRequest = migrateRequest;
