// @grafana/runtime/dist/cjs/utils/publicDashboardQueryHandler.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rxjs = require('rxjs');
var config = require('../config.cjs');
var backendSrv = require('../services/backendSrv.cjs');
var queryResponse = require('./queryResponse.cjs');

"use strict";
function publicDashboardQueryHandler(request) {
  const {
    intervalMs,
    maxDataPoints,
    requestId,
    panelId,
    queryCachingTTL,
    range: { from: fromRange, to: toRange }
  } = request;
  if (!request.targets.length) {
    return rxjs.of({ data: [] });
  }
  const body = {
    intervalMs,
    maxDataPoints,
    queryCachingTTL,
    timeRange: {
      from: fromRange.valueOf().toString(),
      to: toRange.valueOf().toString(),
      timezone: request.timezone
    }
  };
  return backendSrv.getBackendSrv().fetch({
    url: `/api/public/dashboards/${config.config.publicDashboardAccessToken}/panels/${panelId}/query`,
    method: "POST",
    data: body,
    requestId
  }).pipe(
    rxjs.switchMap((raw) => {
      return rxjs.of(queryResponse.toDataQueryResponse(raw, request.targets));
    }),
    rxjs.catchError((err) => {
      return rxjs.of(queryResponse.toDataQueryResponse(err));
    })
  );
}

exports.publicDashboardQueryHandler = publicDashboardQueryHandler;
