// @grafana/runtime/dist/cjs/utils/useFavoriteDatasources.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var config = require('../config.cjs');
var userStorage$1 = require('./userStorage.cjs');

"use strict";
const FAVORITE_DATASOURCES_KEY = "favoriteDatasources";
const userStorage = new userStorage$1.UserStorage("grafana-runtime");
function useFavoriteDatasources() {
  const [favoriteDatasources, setFavoriteDatasources] = React.useState([]);
  const [initialFavoriteDataSources, setInitialFavoriteDataSources] = React.useState([]);
  const [isLoading, setIsLoading] = React.useState(false);
  React.useEffect(() => {
    if (!config.config.featureToggles.favoriteDatasources) {
      return;
    }
    const loadFavorites = async () => {
      setIsLoading(true);
      const stored = await userStorage.getItem(FAVORITE_DATASOURCES_KEY);
      if (stored) {
        const parsed = JSON.parse(stored);
        setFavoriteDatasources(parsed);
        setInitialFavoriteDataSources(parsed);
      }
      setIsLoading(false);
    };
    loadFavorites();
  }, []);
  const saveFavorites = React.useCallback(async (newFavorites) => {
    setIsLoading(true);
    await userStorage.setItem(FAVORITE_DATASOURCES_KEY, JSON.stringify(newFavorites));
    setFavoriteDatasources(newFavorites);
    setIsLoading(false);
  }, []);
  const addFavoriteDatasource = React.useCallback(
    (ds) => {
      if (ds.meta.builtIn) {
        return;
      }
      if (!favoriteDatasources.includes(ds.uid)) {
        const newFavorites = [...favoriteDatasources, ds.uid];
        saveFavorites(newFavorites);
      }
    },
    [favoriteDatasources, saveFavorites]
  );
  const removeFavoriteDatasource = React.useCallback(
    (ds) => {
      const newFavorites = favoriteDatasources.filter((uid) => uid !== ds.uid);
      if (newFavorites.length !== favoriteDatasources.length) {
        saveFavorites(newFavorites);
      }
    },
    [favoriteDatasources, saveFavorites]
  );
  const isFavoriteDatasource = React.useCallback(
    (dsUid) => {
      return favoriteDatasources.includes(dsUid);
    },
    [favoriteDatasources]
  );
  if (!config.config.featureToggles.favoriteDatasources) {
    return {
      enabled: false,
      isLoading: false,
      favoriteDatasources: [],
      initialFavoriteDataSources: [],
      addFavoriteDatasource: () => {
      },
      removeFavoriteDatasource: () => {
      },
      isFavoriteDatasource: () => false
    };
  }
  return {
    enabled: true,
    isLoading,
    favoriteDatasources,
    addFavoriteDatasource,
    removeFavoriteDatasource,
    isFavoriteDatasource,
    initialFavoriteDataSources
  };
}

exports.useFavoriteDatasources = useFavoriteDatasources;
