// @grafana/runtime/dist/cjs/utils/userStorage.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var lodash = require('lodash');
var rxjs = require('rxjs');
var data = require('@grafana/data');
var config = require('../config.cjs');
require('../services/index.cjs');
var backendSrv = require('../services/backendSrv.cjs');

"use strict";
const baseURL = `/apis/userstorage.grafana.app/v0alpha1/namespaces/${config.config.namespace}/user-storage`;
const storageCache = /* @__PURE__ */ new Map();
const operationLocks = /* @__PURE__ */ new Map();
function _clearStorageCache() {
  if (process.env.NODE_ENV !== "test") {
    throw new Error("clearStorageCache() function can only be called from tests.");
  }
  storageCache.clear();
  operationLocks.clear();
}
async function apiRequest(requestOptions) {
  try {
    const { data: responseData, ...meta } = await rxjs.lastValueFrom(
      backendSrv.getBackendSrv().fetch({
        ...requestOptions,
        url: baseURL + requestOptions.url,
        data: requestOptions.body,
        showErrorAlert: false
      })
    );
    return { data: responseData, meta };
  } catch (error) {
    return requestOptions.manageError ? requestOptions.manageError(error) : { error };
  }
}
class UserStorage {
  constructor(service) {
    this.service = service;
    this.userUID = config.config.bootData.user.uid === "" ? config.config.bootData.user.id.toString() : config.config.bootData.user.uid;
    this.resourceName = `${service}:${this.userUID}`;
    this.canUseUserStorage = config.config.bootData.user.isSignedIn;
  }
  /**
   * Acquires a lock for this resourceName to serialize operations.
   * Returns a function to release the lock when done.
   */
  async acquireLock() {
    let lockPromise = operationLocks.get(this.resourceName);
    if (lockPromise) {
      await lockPromise;
    }
    let resolveLock;
    const newLockPromise = new Promise((resolve) => {
      resolveLock = resolve;
    });
    operationLocks.set(this.resourceName, newLockPromise);
    return () => {
      if (resolveLock) {
        resolveLock();
      }
      if (operationLocks.get(this.resourceName) === newLockPromise) {
        operationLocks.delete(this.resourceName);
      }
    };
  }
  async init() {
    const cached = storageCache.get(this.resourceName);
    if (cached !== void 0) {
      if (cached instanceof Promise) {
        try {
          await cached;
          return;
        } catch (error) {
          return error;
        }
      } else {
        return;
      }
    }
    let requestPromise = storageCache.get(this.resourceName);
    if (requestPromise instanceof Promise) {
      try {
        await requestPromise;
        return;
      } catch (error) {
        return error;
      }
    }
    requestPromise = (async () => {
      const userStorage = await apiRequest({
        url: `/${this.resourceName}`,
        method: "GET",
        manageError: (error) => {
          if (lodash.get(error, "status") === 404) {
            return { error: null };
          }
          return { error };
        }
      });
      if ("error" in userStorage) {
        if (userStorage.error === null) {
          return null;
        }
        throw userStorage.error;
      }
      return userStorage.data.spec;
    })();
    const existing = storageCache.get(this.resourceName);
    if (existing === void 0) {
      storageCache.set(this.resourceName, requestPromise);
    } else if (existing instanceof Promise) {
      requestPromise = existing;
    } else {
      return;
    }
    try {
      const result = await requestPromise;
      storageCache.set(this.resourceName, result);
    } catch (error) {
      storageCache.delete(this.resourceName);
      return error;
    }
    return;
  }
  async getItem(key) {
    var _a, _b, _c;
    if (!this.canUseUserStorage) {
      return (_a = data.store.get(`${this.resourceName}:${key}`)) != null ? _a : null;
    }
    const releaseLock = await this.acquireLock();
    try {
      await this.init();
      const storageSpec = storageCache.get(this.resourceName);
      if (!storageSpec) {
        return (_b = data.store.get(`${this.resourceName}:${key}`)) != null ? _b : null;
      }
      if (storageSpec instanceof Promise) {
        const result = await storageSpec;
        return (_c = result == null ? void 0 : result.data[key]) != null ? _c : null;
      }
      return storageSpec.data[key];
    } finally {
      releaseLock();
    }
  }
  async setItem(key, value) {
    if (!this.canUseUserStorage) {
      data.store.set(`${this.resourceName}:${key}`, value);
      return;
    }
    const releaseLock = await this.acquireLock();
    try {
      const newData = { data: { [key]: value } };
      const error = await this.init();
      if (error) {
        data.store.set(`${this.resourceName}:${key}`, value);
        return;
      }
      let storageSpec = storageCache.get(this.resourceName);
      if (storageSpec instanceof Promise) {
        storageSpec = await storageSpec;
      }
      if (!storageSpec) {
        const createResult = await apiRequest({
          url: `/`,
          method: "POST",
          body: {
            metadata: { name: this.resourceName, labels: { user: this.userUID, service: this.service } },
            spec: newData
          },
          manageError: (error2) => {
            data.store.set(`${this.resourceName}:${key}`, value);
            return { error: error2 };
          }
        });
        if ("error" in createResult && createResult.error) {
          return;
        }
        storageCache.set(this.resourceName, newData);
        return;
      }
      const updatedSpec = {
        data: { ...storageSpec.data, [key]: value }
      };
      const updateResult = await apiRequest({
        headers: { "Content-Type": "application/merge-patch+json" },
        url: `/${this.resourceName}`,
        method: "PATCH",
        body: { spec: newData },
        manageError: (error2) => {
          data.store.set(`${this.resourceName}:${key}`, value);
          return { error: error2 };
        }
      });
      if ("error" in updateResult && updateResult.error) {
        return;
      }
      storageCache.set(this.resourceName, updatedSpec);
    } finally {
      releaseLock();
    }
  }
}
function usePluginUserStorage() {
  const context = data.usePluginContext();
  if (!context) {
    throw new Error(`No PluginContext found. The usePluginUserStorage() hook can only be used from a plugin.`);
  }
  return new UserStorage(context == null ? void 0 : context.meta.id);
}

exports.UserStorage = UserStorage;
exports._clearStorageCache = _clearStorageCache;
exports.usePluginUserStorage = usePluginUserStorage;
