// @grafana/ui/dist/cjs/components/DataLinks/DataLinksContextMenu.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var e2eSelectors = require('@grafana/e2e-selectors');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var dataLinks = require('../../utils/dataLinks.cjs');
var WithContextMenu = require('../ContextMenu/WithContextMenu.cjs');
var MenuGroup = require('../Menu/MenuGroup.cjs');
var MenuItem = require('../Menu/MenuItem.cjs');

"use strict";
const DataLinksContextMenu = ({ children, links, style }) => {
  const styles = ThemeContext.useStyles2(getStyles);
  const itemsGroup = [
    { items: dataLinks.linkModelToContextMenuItems(links), label: Boolean(links().length) ? "Data links" : "" }
  ];
  const linksCounter = itemsGroup[0].items.length;
  const renderMenuGroupItems = () => {
    return itemsGroup.map((group, groupIdx) => /* @__PURE__ */ jsxRuntime.jsx(MenuGroup.MenuGroup, { label: group.label, children: (group.items || []).map((item, itemIdx) => /* @__PURE__ */ jsxRuntime.jsx(
      MenuItem.MenuItem,
      {
        url: item.url,
        label: item.label,
        target: item.target,
        icon: item.icon,
        active: item.active,
        onClick: item.onClick,
        className: styles.itemWrapper
      },
      `${group.label}-${groupIdx}-${itemIdx}}`
    )) }, `${group.label}${groupIdx}`));
  };
  const targetClassName = css.css({
    cursor: "context-menu"
  });
  if (linksCounter > 1) {
    return /* @__PURE__ */ jsxRuntime.jsx(WithContextMenu.WithContextMenu, { renderMenuItems: renderMenuGroupItems, children: ({ openMenu }) => {
      return children({ openMenu, targetClassName });
    } });
  } else {
    const linkModel = links()[0];
    return /* @__PURE__ */ jsxRuntime.jsx(
      "a",
      {
        href: linkModel.href,
        onClick: linkModel.onClick,
        target: linkModel.target,
        title: linkModel.title,
        style: { ...style, overflow: "hidden", display: "flex" },
        "data-testid": e2eSelectors.selectors.components.DataLinksContextMenu.singleLink,
        children: children({})
      }
    );
  }
};
const getStyles = (theme) => ({
  itemWrapper: css.css({
    fontSize: 12
  })
});

exports.DataLinksContextMenu = DataLinksContextMenu;
