// @grafana/ui/dist/esm/components/InteractiveTable/Expander/index.mjs
import { jsx } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import { t, Trans } from '@grafana/i18n';
import { IconButton } from '../../IconButton/IconButton.mjs';

"use strict";
const expanderContainerStyles = css({
  display: "flex",
  alignItems: "center",
  height: "100%"
});
function ExpanderCell({ row, __rowID }) {
  return /* @__PURE__ */ jsx("div", { className: expanderContainerStyles, children: /* @__PURE__ */ jsx(
    IconButton,
    {
      tooltip: t("grafana-ui.interactive-table.expand-row-tooltip", "Toggle row expanded"),
      "aria-controls": row.isExpanded ? __rowID : void 0,
      name: row.isExpanded ? "angle-down" : "angle-right",
      "aria-expanded": row.isExpanded,
      ...row.getToggleRowExpandedProps(),
      size: "lg"
    }
  ) });
}
function EmptyExpanderHeader() {
  return /* @__PURE__ */ jsx("span", { className: "sr-only", children: /* @__PURE__ */ jsx(Trans, { i18nKey: "grafana-ui.interactive-table.expand-row-header", children: "Row expander" }) });
}
function ExpanderHeader({ isAllRowsExpanded, toggleAllRowsExpanded }) {
  return /* @__PURE__ */ jsx("div", { className: expanderContainerStyles, children: /* @__PURE__ */ jsx(
    IconButton,
    {
      "aria-label": !isAllRowsExpanded ? t("grafana-ui.interactive-table.aria-label-expand-all", "Expand all rows") : t("grafana-ui.interactive-table.aria-label-collapse-all", "Collapse all rows"),
      name: !isAllRowsExpanded ? "table-expand-all" : "table-collapse-all",
      onClick: () => toggleAllRowsExpanded(),
      size: "lg",
      tooltip: !isAllRowsExpanded ? t("grafana-ui.interactive-table.tooltip-expand-all", "Expand all rows") : t("grafana-ui.interactive-table.tooltip-collapse-all", "Collapse all rows"),
      variant: "secondary"
    }
  ) });
}

export { EmptyExpanderHeader, ExpanderCell, ExpanderHeader };
