// @grafana/plugin-e2e/dist/selectors/versionedAPIs.js
'use strict';

var minGrafanaVersion = require('./minGrafanaVersion.js');

const versionedAPIs = {
  Alerting: {
    eval: {
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: "/api/v1/eval"
    }
  },
  DataSource: {
    resourcePattern: {
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: "/api/datasources/*/resources"
    },
    resourceUIDPattern: {
      "9.4.4": "/api/datasources/uid/*/resources",
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: "/api/datasources/*/resources"
    },
    queryPattern: {
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: "*/**/api/ds/query*"
    },
    query: {
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: "/api/ds/query"
    },
    health: {
      "9.4.0": (uid, _) => `/api/datasources/uid/${uid}/health`,
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: (_, id) => `/api/datasources/${id}/health`
    },
    datasourceByUID: {
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: (uid) => `/api/datasources/uid/${uid}`
    },
    proxy: {
      "9.4.0": (uid, _) => `api/datasources/proxy/uid/${uid}`,
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: (_, id) => `/api/datasources/proxy/${id}`
    }
  },
  Dashboard: {
    delete: {
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: (uid) => `/api/dashboards/uid/${uid}`
    }
  },
  Plugin: {
    settings: {
      [minGrafanaVersion.MIN_GRAFANA_VERSION]: (pluginId) => `/api/plugins/${pluginId}/settings`
    }
  }
};

exports.versionedAPIs = versionedAPIs;
