// @grafana/ui/dist/cjs/components/PanelChrome/PanelStatus.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var e2eSelectors = require('@grafana/e2e-selectors');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var ToolbarButton = require('../ToolbarButton/ToolbarButton.cjs');

"use strict";
function PanelStatus({ message, onClick, ariaLabel = "status" }) {
  const styles = ThemeContext.useStyles2(getStyles);
  return /* @__PURE__ */ jsxRuntime.jsx(
    ToolbarButton.ToolbarButton,
    {
      className: styles.buttonStyles,
      onClick,
      variant: "destructive",
      icon: "exclamation-triangle",
      iconSize: "md",
      tooltip: message || "",
      "aria-label": ariaLabel,
      "data-testid": e2eSelectors.selectors.components.Panels.Panel.status("error")
    }
  );
}
const getStyles = (theme) => {
  return {
    buttonStyles: css.css({
      label: "panel-header-state-button",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      padding: theme.spacing(1),
      width: theme.spacing(theme.components.height.md),
      height: theme.spacing(theme.components.height.md),
      borderRadius: theme.shape.radius.default
    })
  };
};

exports.PanelStatus = PanelStatus;
