// @grafana/plugin-e2e/dist/models/pages/GrafanaPage.js
'use strict';

var utils = require('../utils.js');

class GrafanaPage {
  constructor(ctx, pageArgs = {}) {
    this.ctx = ctx;
    this.pageArgs = pageArgs;
  }
  async navigate(url, options) {
    let queryParams = options?.queryParams ? options.queryParams : this.pageArgs.queryParams;
    if (queryParams) {
      url += `?${queryParams.toString()}`;
    }
    await this.ctx.page.goto(url, {
      waitUntil: "load",
      ...this.pageArgs,
      ...options
    });
  }
  /**
   * Get a locator based on a Grafana E2E selector. A grafana E2E selector is defined in @grafana/e2e-selectors or in plugin-e2e/src/e2e-selectors.
   * An E2E selector is a string that identifies a specific element in the Grafana UI. The element referencing the E2E selector use the data-testid or aria-label attribute.
   */
  getByGrafanaSelector(selector, options) {
    return (options?.root ?? this.ctx.page).locator(utils.resolveGrafanaSelector(selector, options));
  }
  /**
   * Mocks the response of the datasource query call
   * @param json the json response to return
   * @param status the HTTP status code to return. Defaults to 200
   */
  async mockQueryDataResponse(json, status = 200) {
    await this.ctx.page.route(this.ctx.selectors.apis.DataSource.queryPattern, async (route) => {
      await route.fulfill({ json, status });
    });
  }
  /**
   * Mocks the response of the datasource resource request.
   * https://grafana.com/developers/plugin-tools/key-concepts/backend-plugins#resources
   *
   * @param path the path of the resource to mock
   * @param json the json response to return
   * @param status the HTTP status code to return. Defaults to 200
   */
  async mockResourceResponse(path, json, status = 200) {
    await this.ctx.page.route(`${this.ctx.selectors.apis.DataSource.resourceUIDPattern}/${path}`, async (route) => {
      await route.fulfill({ json, status });
    });
    await this.ctx.page.route(`${this.ctx.selectors.apis.DataSource.resourcePattern}/${path}`, async (route) => {
      await route.fulfill({ json, status });
    });
  }
  /**
   * Waits for a data source query data request to be made.
   *
   * @param cb optional callback to filter the request. Use this to filter by request body or other request properties
   */
  async waitForQueryDataRequest(cb) {
    return this.ctx.page.waitForRequest((request) => {
      if (request.url().includes(this.ctx.selectors.apis.DataSource.query) && request.method() === "POST") {
        return cb ? cb(request) : true;
      }
      return false;
    });
  }
  /**
   * Waits for a data source query data response
   *
   * @param cb optional callback to filter the response. Use this to filter by response body or other response properties
   */
  async waitForQueryDataResponse(cb) {
    return this.ctx.page.waitForResponse((response) => {
      if (response.url().includes(this.ctx.selectors.apis.DataSource.query)) {
        return cb ? cb(response) : true;
      }
      return false;
    });
  }
}

exports.GrafanaPage = GrafanaPage;
