// @grafana/runtime/dist/esm/services/pluginMeta/mappers/v0alpha1PanelMapper.mjs
import { PluginState, PluginSignatureStatus, PluginType } from '@grafana/data';
import { loadingStrategyMapper, angularMapper } from './shared.mjs';

"use strict";
function toCDNUrl(spec, path) {
  try {
    const normalizedBase = spec.baseURL.endsWith("/") ? spec.baseURL : `${spec.baseURL}/`;
    const url = new URL(path, normalizedBase);
    return url.toString();
  } catch (error) {
    return path;
  }
}
function screenshotsMapper(spec) {
  if (!spec.pluginJson.info.screenshots) {
    return [];
  }
  return spec.pluginJson.info.screenshots.map((s) => {
    var _a, _b;
    return {
      ...s,
      name: (_a = s.name) != null ? _a : "",
      path: toCDNUrl(spec, (_b = s.path) != null ? _b : "")
    };
  });
}
function logosMapper(spec) {
  return {
    ...spec.pluginJson.info.logos,
    large: toCDNUrl(spec, spec.pluginJson.info.logos.large),
    small: toCDNUrl(spec, spec.pluginJson.info.logos.small)
  };
}
function infoMapper(spec) {
  var _a, _b;
  const { updated, version, description = "", keywords } = spec.pluginJson.info;
  const author = { ...spec.pluginJson.info.author, name: (_b = (_a = spec.pluginJson.info.author) == null ? void 0 : _a.name) != null ? _b : "" };
  const links = (spec.pluginJson.info.links || []).map((l) => {
    var _a2, _b2;
    return { ...l, name: (_a2 = l.name) != null ? _a2 : "", url: (_b2 = l.url) != null ? _b2 : "" };
  });
  const screenshots = screenshotsMapper(spec);
  const build = {};
  const logos = logosMapper(spec);
  return {
    author,
    description,
    links,
    logos,
    build,
    screenshots,
    updated,
    version,
    keywords
  };
}
function stateMapper(spec) {
  const state = spec.pluginJson.state;
  if (state === PluginState.alpha) {
    return PluginState.alpha;
  }
  if (state === PluginState.beta) {
    return PluginState.beta;
  }
  if (state === PluginState.deprecated) {
    return PluginState.deprecated;
  }
  if (state === PluginState.stable) {
    return PluginState.stable;
  }
  return;
}
const idToSortMap = {
  timeseries: 1,
  barchart: 2,
  stat: 3,
  gauge: 4,
  bargauge: 5,
  table: 6,
  singlestat: 7,
  piechart: 8,
  "state-timeline": 9,
  heatmap: 10,
  "status-history": 11,
  histogram: 12,
  graph: 13,
  text: 14,
  alertlist: 15,
  dashlist: 16,
  news: 17
};
function sortMapper(spec) {
  var _a;
  return (_a = idToSortMap[spec.pluginJson.id]) != null ? _a : 100;
}
function signatureMapper(spec) {
  var _a;
  const signature = (_a = spec.signature) == null ? void 0 : _a.status;
  if (!signature) {
    return;
  }
  if (signature === PluginSignatureStatus.internal) {
    return PluginSignatureStatus.internal;
  }
  if (signature === PluginSignatureStatus.invalid) {
    return PluginSignatureStatus.invalid;
  }
  if (signature === PluginSignatureStatus.modified) {
    return PluginSignatureStatus.modified;
  }
  if (signature === PluginSignatureStatus.valid) {
    return PluginSignatureStatus.valid;
  }
  return;
}
function specMapper(spec) {
  const { id, name, hideFromList = false, skipDataQuery = false, suggestions } = spec.pluginJson;
  const state = stateMapper(spec);
  const info = infoMapper(spec);
  const loadingStrategy = loadingStrategyMapper(spec);
  const sort = sortMapper(spec);
  const type = PluginType.panel;
  const module = spec.module.path;
  const baseUrl = spec.baseURL;
  const signature = signatureMapper(spec);
  const angular = angularMapper(spec);
  const translations = spec.translations;
  const moduleHash = spec.module.hash;
  return {
    id,
    name,
    info,
    hideFromList,
    sort,
    skipDataQuery,
    suggestions,
    state,
    baseUrl,
    signature,
    module,
    angular,
    loadingStrategy,
    type,
    translations,
    moduleHash
  };
}
const v0alpha1PanelMapper = (response) => {
  const result = {};
  return response.items.reduce((acc, curr) => {
    if (curr.spec.pluginJson.type !== "panel") {
      return acc;
    }
    const config = specMapper(curr.spec);
    acc[config.id] = config;
    return acc;
  }, result);
};

export { v0alpha1PanelMapper };
