// @grafana/runtime/dist/esm/services/pluginMeta/mappers/v0alpha1AppMapper.mjs
import { angularMapper, loadingStrategyMapper } from './shared.mjs';

"use strict";
function dependenciesMapper(spec) {
  var _a, _b, _c, _d, _e;
  const plugins = ((_b = (_a = spec.pluginJson.dependencies) == null ? void 0 : _a.plugins) != null ? _b : []).map((v) => ({
    ...v,
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    type: v.type,
    version: ""
  }));
  const dependencies = {
    ...spec.pluginJson.dependencies,
    extensions: {
      exposedComponents: (_d = (_c = spec.pluginJson.dependencies.extensions) == null ? void 0 : _c.exposedComponents) != null ? _d : []
    },
    grafanaDependency: spec.pluginJson.dependencies.grafanaDependency,
    grafanaVersion: (_e = spec.pluginJson.dependencies.grafanaVersion) != null ? _e : "",
    plugins
  };
  return dependencies;
}
function extensionsMapper(spec) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j;
  const addedComponents = (_b = (_a = spec.pluginJson.extensions) == null ? void 0 : _a.addedComponents) != null ? _b : [];
  const addedFunctions = (_d = (_c = spec.pluginJson.extensions) == null ? void 0 : _c.addedFunctions) != null ? _d : [];
  const addedLinks = (_f = (_e = spec.pluginJson.extensions) == null ? void 0 : _e.addedLinks) != null ? _f : [];
  const exposedComponents = ((_h = (_g = spec.pluginJson.extensions) == null ? void 0 : _g.exposedComponents) != null ? _h : []).map((v) => {
    var _a2, _b2;
    return {
      ...v,
      description: (_a2 = v.description) != null ? _a2 : "",
      title: (_b2 = v.title) != null ? _b2 : ""
    };
  });
  const extensionPoints = ((_j = (_i = spec.pluginJson.extensions) == null ? void 0 : _i.extensionPoints) != null ? _j : []).map((v) => {
    var _a2, _b2;
    return {
      ...v,
      description: (_a2 = v.description) != null ? _a2 : "",
      title: (_b2 = v.title) != null ? _b2 : ""
    };
  });
  const extensions = {
    addedComponents,
    addedFunctions,
    addedLinks,
    exposedComponents,
    extensionPoints
  };
  return extensions;
}
function specMapper(spec) {
  var _a, _b, _c, _d;
  const { id, info, preload = false } = spec.pluginJson;
  const angular = angularMapper(spec);
  const dependencies = dependenciesMapper(spec);
  const extensions = extensionsMapper(spec);
  const loadingStrategy = loadingStrategyMapper(spec);
  const path = (_b = (_a = spec.module) == null ? void 0 : _a.path) != null ? _b : "";
  const version = info.version;
  const buildMode = (_c = spec.pluginJson.buildMode) != null ? _c : "production";
  const moduleHash = (_d = spec.module) == null ? void 0 : _d.hash;
  return {
    id,
    angular,
    dependencies,
    extensions,
    loadingStrategy,
    path,
    preload,
    version,
    buildMode,
    moduleHash
  };
}
const v0alpha1AppMapper = (response) => {
  const result = {};
  return response.items.reduce((acc, curr) => {
    if (curr.spec.pluginJson.type !== "app") {
      return acc;
    }
    const config = specMapper(curr.spec);
    acc[config.id] = config;
    return acc;
  }, result);
};

export { v0alpha1AppMapper };
