// @grafana/schema/dist/cjs/index.gen.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var types_gen = require('./raw/dashboard/x/types.gen.cjs');
var dashboard_types = require('./veneer/dashboard.types.cjs');

"use strict";

exports.DashboardCursorSync = types_gen.DashboardCursorSync;
exports.FieldColorModeId = types_gen.FieldColorModeId;
exports.MappingType = types_gen.MappingType;
exports.SpecialValueMatch = types_gen.SpecialValueMatch;
exports.ThresholdsMode = types_gen.ThresholdsMode;
exports.VariableRefresh = types_gen.VariableRefresh;
exports.VariableSort = types_gen.VariableSort;
exports.defaultAction = types_gen.defaultAction;
exports.defaultAnnotationPanelFilter = types_gen.defaultAnnotationPanelFilter;
exports.defaultAnnotationTarget = types_gen.defaultAnnotationTarget;
exports.defaultDashboardCursorSync = types_gen.defaultDashboardCursorSync;
exports.defaultDashboardLink = types_gen.defaultDashboardLink;
exports.defaultFetchOptions = types_gen.defaultFetchOptions;
exports.defaultGridPos = types_gen.defaultGridPos;
exports.defaultInfinityOptions = types_gen.defaultInfinityOptions;
exports.defaultThresholdsConfig = types_gen.defaultThresholdsConfig;
exports.VariableHide = dashboard_types.VariableHide;
exports.defaultAnnotationContainer = dashboard_types.defaultAnnotationContainer;
exports.defaultAnnotationQuery = dashboard_types.defaultAnnotationQuery;
exports.defaultDashboard = dashboard_types.defaultDashboard;
exports.defaultFieldConfig = dashboard_types.defaultFieldConfig;
exports.defaultFieldConfigSource = dashboard_types.defaultFieldConfigSource;
exports.defaultMatcherConfig = dashboard_types.defaultMatcherConfig;
exports.defaultPanel = dashboard_types.defaultPanel;
exports.defaultRowPanel = dashboard_types.defaultRowPanel;
exports.defaultTimePickerConfig = dashboard_types.defaultTimePickerConfig;
exports.defaultVariableModel = dashboard_types.defaultVariableModel;
