// @grafana/ui/dist/cjs/components/Table/TableNG/TableNG.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
require('react-data-grid/lib/styles.css');
var clsx = require('clsx');
var memoize = require('micro-memoize');
var React = require('react');
var reactDataGrid = require('react-data-grid');
var data = require('@grafana/data');
var i18n = require('@grafana/i18n');
var schema = require('@grafana/schema');
var ThemeContext = require('../../../themes/ThemeContext.cjs');
var colors = require('../../../utils/colors.cjs');
var Pagination = require('../../Pagination/Pagination.cjs');
require('../../PanelChrome/index.cjs');
var DataLinksActionsTooltip = require('../DataLinksActionsTooltip.cjs');
var TableCellInspector = require('../TableCellInspector.cjs');
require('../types.cjs');
var renderers = require('./Cells/renderers.cjs');
var HeaderCell = require('./components/HeaderCell.cjs');
var RowExpander = require('./components/RowExpander.cjs');
var SummaryCell = require('./components/SummaryCell.cjs');
var TableCellActions = require('./components/TableCellActions.cjs');
var TableCellTooltip = require('./components/TableCellTooltip.cjs');
var constants = require('./constants.cjs');
var hooks = require('./hooks.cjs');
var styles = require('./styles.cjs');
var utils = require('./utils.cjs');
var PanelContext = require('../../PanelChrome/PanelContext.cjs');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var memoize__default = /*#__PURE__*/_interopDefaultCompat(memoize);

"use strict";
const EXPANDED_COLUMN_KEY = "expanded";
function TableNG(props) {
  var _a, _b;
  const {
    cellHeight,
    data: data$1,
    disableKeyboardEvents,
    disableSanitizeHtml,
    enablePagination = false,
    enableSharedCrosshair = false,
    enableVirtualization,
    frozenColumns = 0,
    getActions = () => [],
    height,
    initialSortBy,
    maxRowHeight: _maxRowHeight,
    noHeader,
    onCellFilterAdded,
    onColumnResize,
    onSortByChange,
    showTypeIcons,
    structureRev,
    timeRange,
    transparent,
    width,
    initialRowIndex
  } = props;
  const uniqueId = React.useId();
  const theme = ThemeContext.useTheme2();
  const styles$1 = ThemeContext.useStyles2(styles.getGridStyles, enablePagination, transparent);
  const panelContext = PanelContext.usePanelContext();
  const userCanExecuteActions = React.useMemo(() => {
    var _a2, _b2;
    return (_b2 = (_a2 = panelContext.canExecuteActions) == null ? void 0 : _a2.call(panelContext)) != null ? _b2 : false;
  }, [panelContext]);
  const getCellActions = React.useCallback(
    (field, rowIdx) => {
      if (!userCanExecuteActions) {
        return [];
      }
      return getActions(data$1, field, rowIdx);
    },
    [getActions, data$1, userCanExecuteActions]
  );
  const visibleFields = React.useMemo(() => utils.getVisibleFields(data$1.fields), [data$1.fields]);
  const hasHeader = !noHeader;
  const hasFooter = React.useMemo(
    () => visibleFields.some((field) => {
      var _a2, _b2, _c;
      return Boolean((_c = (_b2 = (_a2 = field.config.custom) == null ? void 0 : _a2.footer) == null ? void 0 : _b2.reducers) == null ? void 0 : _c.length);
    }),
    [visibleFields]
  );
  const footerHeight = React.useMemo(
    () => hasFooter ? utils.calculateFooterHeight(visibleFields) : 0,
    [hasFooter, visibleFields]
  );
  const resizeHandler = hooks.useColumnResize(onColumnResize);
  const hasNestedFrames = React.useMemo(() => utils.getIsNestedTable(data$1.fields), [data$1]);
  const nestedFramesFieldName = React.useMemo(() => {
    if (!hasNestedFrames) {
      return;
    }
    const firstNestedField = data$1.fields.find((f) => f.type === data.FieldType.nestedFrames);
    if (!firstNestedField) {
      return;
    }
    return utils.getDisplayName(firstNestedField);
  }, [data$1, hasNestedFrames]);
  const rows = React.useMemo(() => utils.frameToRecords(data$1, nestedFramesFieldName), [data$1, nestedFramesFieldName]);
  const getTextColorForBackground = React.useMemo(() => memoize__default.default(colors.getTextColorForBackground, { maxSize: 1e3 }), []);
  const {
    rows: filteredRows,
    filter,
    setFilter,
    crossFilterOrder,
    crossFilterRows
  } = hooks.useFilteredRows(rows, data$1.fields, { hasNestedFrames });
  const {
    rows: sortedRows,
    sortColumns,
    setSortColumns
  } = hooks.useSortedRows(filteredRows, data$1.fields, { hasNestedFrames, initialSortBy });
  const [inspectCell, setInspectCell] = React.useState(null);
  const [tooltipState, setTooltipState] = React.useState();
  const [expandedRows, setExpandedRows] = React.useState(() => /* @__PURE__ */ new Set());
  const [selectedRows, setSelectedRows] = React.useState(() => /* @__PURE__ */ new Set());
  const defaultRowHeight = React.useMemo(
    () => utils.getDefaultRowHeight(theme, visibleFields, cellHeight),
    [theme, visibleFields, cellHeight]
  );
  const gridRef = React.useRef(null);
  const scrollbarWidth = hooks.useScrollbarWidth(gridRef, height);
  const availableWidth = React.useMemo(
    () => (hasNestedFrames ? width - constants.COLUMN.EXPANDER_WIDTH : width) - scrollbarWidth,
    [width, hasNestedFrames, scrollbarWidth]
  );
  const getCellColorInlineStyles = React.useMemo(() => utils.getCellColorInlineStylesFactory(theme), [theme]);
  const applyToRowBgFn = React.useMemo(
    () => {
      var _a2;
      return (_a2 = utils.getApplyToRowBgFn(data$1.fields, getCellColorInlineStyles)) != null ? _a2 : void 0;
    },
    [data$1.fields, getCellColorInlineStyles]
  );
  const typographyCtx = React.useMemo(
    () => utils.createTypographyContext(
      theme.typography.fontSize,
      theme.typography.fontFamily,
      utils.extractPixelValue(theme.typography.body.letterSpacing) * theme.typography.fontSize
    ),
    [theme]
  );
  const [widths, numFrozenColsFullyInView] = hooks.useColWidths(visibleFields, availableWidth, frozenColumns);
  const headerHeight = hooks.useHeaderHeight({
    columnWidths: widths,
    fields: visibleFields,
    enabled: hasHeader,
    sortColumns,
    showTypeIcons: showTypeIcons != null ? showTypeIcons : false,
    typographyCtx
  });
  const maxRowHeight = _maxRowHeight != null ? Math.max(constants.TABLE.LINE_HEIGHT, _maxRowHeight) : void 0;
  const rowHeight = hooks.useRowHeight({
    columnWidths: widths,
    fields: visibleFields,
    hasNestedFrames,
    defaultHeight: defaultRowHeight,
    expandedRows,
    typographyCtx,
    maxHeight: maxRowHeight
  });
  const {
    rows: paginatedRows,
    page,
    setPage,
    numPages,
    pageRangeStart,
    pageRangeEnd,
    smallPagination
  } = hooks.usePaginatedRows(sortedRows, {
    enabled: enablePagination,
    width: availableWidth,
    height,
    footerHeight,
    headerHeight: hasHeader ? constants.TABLE.HEADER_HEIGHT : 0,
    rowHeight
  });
  const [scrollToIndex, setScrollToIndex] = React.useState(initialRowIndex);
  React.useEffect(() => {
    var _a2;
    if (scrollToIndex !== void 0 && sortedRows && ((_a2 = gridRef.current) == null ? void 0 : _a2.scrollToCell)) {
      const rowIdx = sortedRows.findIndex((row) => row.__index === scrollToIndex);
      gridRef.current.scrollToCell({
        rowIdx
      });
      setScrollToIndex(void 0);
      setSelectedRows(/* @__PURE__ */ new Set([utils.rowKeyGetter(sortedRows[rowIdx])]));
    }
  }, [scrollToIndex, sortedRows]);
  const [footers, isUniformFooter] = React.useMemo(() => {
    var _a2, _b2, _c, _d, _e;
    const footers2 = [];
    let isUniformFooter2 = true;
    let firstReducers;
    for (const field of visibleFields) {
      const footer = (_b2 = (_a2 = field.config) == null ? void 0 : _a2.custom) == null ? void 0 : _b2.footer;
      footers2.push(footer);
      if (firstReducers === void 0 && ((_d = (_c = footer == null ? void 0 : footer.reducers) == null ? void 0 : _c.length) != null ? _d : 0) > 0) {
        firstReducers = footer == null ? void 0 : footer.reducers;
      } else if (firstReducers !== void 0) {
        const reducers = footer == null ? void 0 : footer.reducers;
        if ((_e = reducers == null ? void 0 : reducers.length) != null ? _e : 0 > 0) {
          if (reducers.length !== firstReducers.length || reducers.some((r, idx) => (firstReducers == null ? void 0 : firstReducers[idx]) !== r)) {
            isUniformFooter2 = false;
            break;
          }
        }
      }
    }
    return [footers2, isUniformFooter2];
  }, [visibleFields]);
  const rowHeightFn = React.useMemo(() => {
    if (typeof rowHeight === "function") {
      return rowHeight;
    }
    if (typeof rowHeight === "string") {
      return () => constants.TABLE.MAX_CELL_HEIGHT;
    }
    return () => rowHeight;
  }, [rowHeight]);
  const renderRow = React.useMemo(
    () => renderRowFactory(data$1.fields, panelContext, expandedRows, enableSharedCrosshair),
    [data$1, enableSharedCrosshair, expandedRows, panelContext]
  );
  const commonDataGridProps = React.useMemo(
    () => ({
      enableVirtualization: !utils.IS_SAFARI_26 && enableVirtualization !== false && rowHeight !== "auto",
      defaultColumnOptions: {
        minWidth: 50,
        resizable: true,
        sortable: true
        // draggable: true,
      },
      onColumnResize: resizeHandler,
      onSortColumnsChange: (newSortColumns) => {
        setSortColumns(newSortColumns);
        onSortByChange == null ? void 0 : onSortByChange(
          newSortColumns.map(({ columnKey, direction }) => ({
            displayName: columnKey,
            desc: direction === "DESC"
          }))
        );
      },
      sortColumns,
      rowHeight,
      bottomSummaryRows: hasFooter ? [{}] : void 0,
      summaryRowHeight: footerHeight,
      headerRowClass: styles$1.headerRow,
      headerRowHeight: noHeader ? 0 : constants.TABLE.HEADER_HEIGHT
    }),
    [
      enableVirtualization,
      hasFooter,
      resizeHandler,
      sortColumns,
      rowHeight,
      styles$1.headerRow,
      noHeader,
      setSortColumns,
      onSortByChange,
      footerHeight
    ]
  );
  const buildNestedTableExpanderColumn = React.useCallback(
    (nestedColumns, hasNestedHeaders, renderers) => ({
      key: EXPANDED_COLUMN_KEY,
      name: i18n.t("grafana-ui.table.nested-table.expander-column-name", "Expand nested rows"),
      field: {
        name: "",
        type: data.FieldType.other,
        config: {},
        values: []
      },
      cellClass(row) {
        if (row.__depth !== 0) {
          return styles$1.cellNested;
        }
        return;
      },
      colSpan(args) {
        return args.type === "ROW" && args.row.__depth === 1 ? data$1.fields.length : 1;
      },
      renderCell: ({ row }) => {
        const rowId = `${uniqueId}-nested-table-${row.__index}`;
        if (row.__depth === 0) {
          const rowIdx = row.__index;
          return /* @__PURE__ */ jsxRuntime.jsx(
            RowExpander.RowExpander,
            {
              rowId,
              isExpanded: expandedRows.has(rowIdx),
              onCellExpand: () => {
                if (expandedRows.has(rowIdx)) {
                  expandedRows.delete(rowIdx);
                } else {
                  expandedRows.add(rowIdx);
                }
                setExpandedRows(new Set(expandedRows));
              }
            }
          );
        }
        const nestedData = row.data;
        if (!nestedData) {
          return null;
        }
        const expandedRecords = utils.applySort(
          utils.frameToRecords(nestedData, nestedFramesFieldName),
          nestedData.fields,
          sortColumns
        );
        if (!expandedRecords.length) {
          return /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles$1.noDataNested, children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.table.nested-table.no-data", children: "No data" }) });
        }
        return /* @__PURE__ */ jsxRuntime.jsx("div", { id: rowId, children: /* @__PURE__ */ jsxRuntime.jsx(
          reactDataGrid.DataGrid,
          {
            ...commonDataGridProps,
            className: clsx.clsx(styles$1.grid, styles$1.gridNested),
            headerRowClass: clsx.clsx(styles$1.headerRow, { [styles$1.displayNone]: !hasNestedHeaders }),
            headerRowHeight: hasNestedHeaders ? constants.TABLE.HEADER_HEIGHT : 0,
            columns: nestedColumns,
            rows: expandedRecords,
            renderers
          }
        ) });
      },
      renderHeaderCell(props2) {
        return /* @__PURE__ */ jsxRuntime.jsx("div", { className: "sr-only", children: props2.column.name });
      },
      width: constants.COLUMN.EXPANDER_WIDTH,
      minWidth: constants.COLUMN.EXPANDER_WIDTH
    }),
    [
      styles$1.cellNested,
      styles$1.grid,
      styles$1.gridNested,
      styles$1.headerRow,
      styles$1.displayNone,
      styles$1.noDataNested,
      data$1.fields.length,
      uniqueId,
      nestedFramesFieldName,
      sortColumns,
      commonDataGridProps,
      expandedRows
    ]
  );
  const fromFields = React.useCallback(
    (f, widths2) => {
      const result = {
        columns: [],
        cellRootRenderers: {}
      };
      let lastRowIdx = -1;
      let rowCellStyle = {
        color: void 0,
        background: void 0
      };
      f.forEach((field, i) => {
        var _a2, _b2, _c, _d, _e, _f, _g, _h, _i, _j, _k;
        const cellOptions = utils.getCellOptions(field);
        const cellType = cellOptions.type;
        if (cellType === schema.TableCellDisplayMode.Pill && ((_b2 = (_a2 = field.config.mappings) == null ? void 0 : _a2.length) != null ? _b2 : 0 > 0)) {
          field = {
            ...field,
            config: {
              ...field.config,
              color: {
                ...field.config.color,
                mode: schema.FieldColorModeId.Fixed,
                fixedColor: (_d = (_c = field.config.color) == null ? void 0 : _c.fixedColor) != null ? _d : data.FALLBACK_COLOR
              }
            }
          };
          field.display = data.getDisplayProcessor({ field, theme });
        }
        if (cellType === schema.TableCellDisplayMode.JSONView || field.type === data.FieldType.other) {
          field.display = utils.displayJsonValue(field);
        }
        const textAlign = utils.getAlignment(field);
        const justifyContent = utils.getJustifyContent(textAlign);
        const displayName = utils.getDisplayName(field);
        const headerCellClass = styles.getHeaderCellStyles(theme, justifyContent);
        const CellType = renderers.getCellRenderer(field, cellOptions);
        const cellInspect = utils.isCellInspectEnabled(field);
        const showFilters = Boolean(field.config.filterable && onCellFilterAdded != null);
        const showActions = cellInspect || showFilters;
        const width2 = widths2[i];
        const cellActionClassName = showActions ? clsx.clsx("table-cell-actions", styles.getCellActionStyles(theme, textAlign)) : void 0;
        const shouldOverflow = !utils.IS_SAFARI_26 && rowHeight !== "auto" && (utils.shouldTextOverflow(field) || Boolean(maxRowHeight));
        const textWrap = rowHeight === "auto" || utils.shouldTextWrap(field);
        const canBeColorized = utils.canFieldBeColorized(cellType, applyToRowBgFn);
        const cellStyleOptions = {
          textAlign,
          textWrap,
          shouldOverflow,
          maxHeight: maxRowHeight
        };
        const defaultCellStyles = styles.getDefaultCellStyles(theme, cellStyleOptions);
        const cellSpecificStyles = renderers.getCellSpecificStyles(cellType, field, theme, cellStyleOptions);
        const linkStyles = styles.getLinkStyles(theme, canBeColorized);
        const cellParentStyles = clsx.clsx(defaultCellStyles, linkStyles);
        const maxHeightClassName = maxRowHeight ? styles.getMaxHeightCellStyles(theme, cellStyleOptions) : void 0;
        const styleFieldValue = (_e = field.config.custom) == null ? void 0 : _e.styleField;
        const styleField = styleFieldValue ? data$1.fields.find(utils.predicateByName(styleFieldValue)) : void 0;
        const styleFieldName = styleField ? utils.getDisplayName(styleField) : void 0;
        const hasValidStyleField = Boolean(styleFieldName);
        const renderCellRoot2 = (key, props2) => {
          const rowIdx = props2.row.__index;
          if (rowIdx !== lastRowIdx) {
            lastRowIdx = rowIdx;
            rowCellStyle.color = void 0;
            rowCellStyle.background = void 0;
            if (applyToRowBgFn != null) {
              rowCellStyle = { ...rowCellStyle, ...applyToRowBgFn(rowIdx) };
            }
          }
          let style = { ...rowCellStyle };
          if (canBeColorized) {
            const value = props2.row[props2.column.key];
            const displayValue = field.display(value);
            const cellColorStyles = getCellColorInlineStyles(cellOptions, displayValue, applyToRowBgFn != null);
            Object.assign(style, cellColorStyles);
          }
          if (hasValidStyleField) {
            style = { ...style, ...utils.parseStyleJson(props2.row[styleFieldName]) };
          }
          return /* @__PURE__ */ jsxRuntime.jsx(
            reactDataGrid.Cell,
            {
              ...props2,
              className: clsx.clsx(
                props2.className,
                cellParentStyles,
                cellSpecificStyles != null && { [cellSpecificStyles]: maxRowHeight == null }
              ),
              style
            },
            key
          );
        };
        result.cellRootRenderers[displayName] = renderCellRoot2;
        const renderBasicCellContent = (props2) => {
          const rowIdx = props2.row.__index;
          const value = props2.row[props2.column.key];
          const height2 = rowHeightFn(props2.row);
          const frame = data$1;
          let result2 = /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
            /* @__PURE__ */ jsxRuntime.jsx(
              CellType,
              {
                cellOptions,
                frame,
                field,
                height: height2,
                rowIdx,
                theme,
                value,
                width: width2,
                timeRange,
                cellInspect,
                showFilters,
                getActions: getCellActions,
                disableSanitizeHtml,
                getTextColorForBackground
              }
            ),
            showActions && /* @__PURE__ */ jsxRuntime.jsx(
              TableCellActions.TableCellActions,
              {
                field,
                value,
                displayName,
                cellInspect,
                showFilters,
                className: cellActionClassName,
                setInspectCell,
                onCellFilterAdded
              }
            )
          ] });
          if (maxRowHeight != null) {
            result2 = /* @__PURE__ */ jsxRuntime.jsx("div", { className: clsx.clsx(maxHeightClassName, cellSpecificStyles), children: result2 });
          }
          return result2;
        };
        let renderCellContent = renderBasicCellContent;
        const tooltipFieldName = (_g = (_f = field.config.custom) == null ? void 0 : _f.tooltip) == null ? void 0 : _g.field;
        if (tooltipFieldName) {
          const tooltipField = data$1.fields.find(utils.predicateByName(tooltipFieldName));
          if (tooltipField) {
            const tooltipDisplayName = utils.getDisplayName(tooltipField);
            const tooltipCellOptions = utils.getCellOptions(tooltipField);
            const tooltipFieldRenderer = renderers.getCellRenderer(tooltipField, tooltipCellOptions);
            const tooltipCellStyleOptions = {
              textAlign: utils.getAlignment(tooltipField),
              textWrap: utils.shouldTextWrap(tooltipField),
              shouldOverflow: false,
              maxHeight: maxRowHeight
            };
            const tooltipCanBeColorized = utils.canFieldBeColorized(tooltipCellOptions.type, applyToRowBgFn);
            const tooltipDefaultStyles = styles.getDefaultCellStyles(theme, tooltipCellStyleOptions);
            const tooltipSpecificStyles = renderers.getCellSpecificStyles(
              tooltipCellOptions.type,
              tooltipField,
              theme,
              tooltipCellStyleOptions
            );
            const tooltipLinkStyles = styles.getLinkStyles(theme, tooltipCanBeColorized);
            const tooltipClasses = styles.getTooltipStyles(theme, textAlign);
            const placement = (_j = (_i = (_h = field.config.custom) == null ? void 0 : _h.tooltip) == null ? void 0 : _i.placement) != null ? _j : schema.TableCellTooltipPlacement.Auto;
            const tooltipWidth = placement === schema.TableCellTooltipPlacement.Left || placement === schema.TableCellTooltipPlacement.Right ? (_k = tooltipField.config.custom) == null ? void 0 : _k.width : width2;
            const tooltipProps = {
              cellOptions: tooltipCellOptions,
              classes: tooltipClasses,
              className: clsx.clsx(
                tooltipClasses.tooltipContent,
                tooltipDefaultStyles,
                tooltipSpecificStyles,
                tooltipLinkStyles
              ),
              data: data$1,
              disableSanitizeHtml,
              field: tooltipField,
              getActions: getCellActions,
              getTextColorForBackground,
              gridRef,
              placement,
              renderer: tooltipFieldRenderer,
              tooltipField,
              theme,
              width: tooltipWidth
            };
            renderCellContent = (props2) => {
              const tooltipHeight = rowHeightFn(props2.row);
              let tooltipStyle = { ...rowCellStyle };
              if (tooltipCanBeColorized) {
                const tooltipDisplayValue = tooltipField.display(props2.row[tooltipDisplayName]);
                const tooltipCellColorStyles = getCellColorInlineStyles(
                  tooltipCellOptions,
                  tooltipDisplayValue,
                  applyToRowBgFn != null
                );
                Object.assign(tooltipStyle, tooltipCellColorStyles);
              }
              return /* @__PURE__ */ jsxRuntime.jsx(
                TableCellTooltip.TableCellTooltip,
                {
                  ...tooltipProps,
                  height: tooltipHeight,
                  rowIdx: props2.row.__index,
                  style: tooltipStyle,
                  children: renderBasicCellContent(props2)
                }
              );
            };
          }
        }
        result.columns.push({
          field,
          key: displayName,
          name: displayName,
          width: width2,
          headerCellClass,
          frozen: Math.min(frozenColumns, numFrozenColsFullyInView) > i,
          renderCell: renderCellContent,
          renderHeaderCell: ({ column, sortDirection }) => /* @__PURE__ */ jsxRuntime.jsx(
            HeaderCell.HeaderCell,
            {
              column,
              rows,
              field,
              filter,
              setFilter,
              disableKeyboardEvents,
              crossFilterOrder,
              crossFilterRows,
              direction: sortDirection,
              showTypeIcons,
              selectFirstCell: () => {
                var _a3;
                (_a3 = gridRef.current) == null ? void 0 : _a3.selectCell({ rowIdx: 0, idx: 0 });
              }
            }
          ),
          renderSummaryCell: () => /* @__PURE__ */ jsxRuntime.jsx(
            SummaryCell.SummaryCell,
            {
              rows: sortedRows,
              footers,
              field,
              colIdx: i,
              textAlign: utils.getSummaryCellTextAlign(textAlign, cellType),
              rowLabel: isUniformFooter && i === 0,
              hideLabel: isUniformFooter && i !== 0
            }
          )
        });
      });
      return result;
    },
    [
      theme,
      onCellFilterAdded,
      rowHeight,
      maxRowHeight,
      applyToRowBgFn,
      data$1,
      frozenColumns,
      numFrozenColsFullyInView,
      getCellColorInlineStyles,
      rowHeightFn,
      timeRange,
      getCellActions,
      disableSanitizeHtml,
      getTextColorForBackground,
      rows,
      filter,
      setFilter,
      disableKeyboardEvents,
      crossFilterOrder,
      crossFilterRows,
      showTypeIcons,
      sortedRows,
      footers,
      isUniformFooter
    ]
  );
  const firstRowNestedData = React.useMemo(
    () => {
      var _a2;
      return hasNestedFrames ? (_a2 = rows.find((r) => r.data)) == null ? void 0 : _a2.data : void 0;
    },
    [hasNestedFrames, rows]
  );
  const nestedVisibleFields = React.useMemo(
    () => firstRowNestedData ? utils.getVisibleFields(firstRowNestedData.fields) : [],
    [firstRowNestedData]
  );
  const [nestedFieldWidths] = hooks.useColWidths(nestedVisibleFields, availableWidth);
  const { columns, cellRootRenderers } = React.useMemo(() => {
    var _a2, _b2;
    const result = fromFields(visibleFields, widths);
    if (!firstRowNestedData) {
      return result;
    }
    const hasNestedHeaders = ((_b2 = (_a2 = firstRowNestedData.meta) == null ? void 0 : _a2.custom) == null ? void 0 : _b2.noHeader) !== true;
    const renderRow2 = renderRowFactory(firstRowNestedData.fields, panelContext, expandedRows, enableSharedCrosshair);
    const { columns: nestedColumns, cellRootRenderers: nestedCellRootRenderers } = fromFields(
      nestedVisibleFields,
      nestedFieldWidths
    );
    const expanderCellRenderer = (key, props2) => /* @__PURE__ */ jsxRuntime.jsx(reactDataGrid.Cell, { ...props2 }, key);
    result.cellRootRenderers[EXPANDED_COLUMN_KEY] = expanderCellRenderer;
    result.columns.unshift(
      buildNestedTableExpanderColumn(nestedColumns, hasNestedHeaders, {
        renderRow: renderRow2,
        renderCell: (key, props2) => nestedCellRootRenderers[props2.column.key](key, props2)
      })
    );
    return result;
  }, [
    buildNestedTableExpanderColumn,
    enableSharedCrosshair,
    expandedRows,
    firstRowNestedData,
    fromFields,
    nestedFieldWidths,
    nestedVisibleFields,
    panelContext,
    visibleFields,
    widths
  ]);
  const structureRevColumns = React.useMemo(() => columns, [columns, structureRev]);
  const renderCellRoot = React.useCallback(
    (key, props2) => cellRootRenderers[props2.column.key](key, props2),
    [cellRootRenderers]
  );
  const itemsRangeStart = pageRangeStart;
  const displayedEnd = pageRangeEnd;
  const numRows = sortedRows.length;
  let rendered = /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      reactDataGrid.DataGrid,
      {
        ...commonDataGridProps,
        role: hasNestedFrames ? "treegrid" : "grid",
        ref: gridRef,
        className: styles$1.grid,
        columns: structureRevColumns,
        rows: paginatedRows,
        rowKeyGetter: utils.rowKeyGetter,
        isRowSelectionDisabled: () => initialRowIndex !== void 0,
        selectedRows,
        onSelectedRowsChange: setSelectedRows,
        headerRowClass: clsx.clsx(styles$1.headerRow, { [styles$1.displayNone]: noHeader }),
        headerRowHeight: headerHeight,
        onCellClick: ({ column, row }, { clientX, clientY, preventGridDefault, target }) => {
          var _a2;
          const field = columns[column.idx].field;
          if (target instanceof HTMLElement && // this walks up the tree to find either a faux link wrapper or the cell root
          // it then only proceeds if we matched the faux link wrapper
          ((_a2 = target.closest("a[aria-haspopup], .rdg-cell")) == null ? void 0 : _a2.matches("a"))) {
            const rowIdx = row.__index;
            setTooltipState({
              coords: {
                clientX,
                clientY
              },
              links: utils.getCellLinks(field, rowIdx),
              actions: getCellActions(field, rowIdx)
            });
            preventGridDefault();
          }
        },
        onCellKeyDown: ({ column, row }, event) => {
          var _a2;
          if (column.key === columns[0].key && row.__index === 0 && event.shiftKey && event.key === "Tab") {
            event.preventGridDefault();
            (_a2 = gridRef.current) == null ? void 0 : _a2.selectCell({ rowIdx: -1, idx: columns.length - 1 });
            return;
          }
          if (disableKeyboardEvents || hasNestedFrames && event.isDefaultPrevented()) {
            event.preventGridDefault();
          }
        },
        renderers: { renderRow, renderCell: renderCellRoot }
      }
    ),
    enablePagination && /* @__PURE__ */ jsxRuntime.jsxs("div", { className: styles$1.paginationContainer, children: [
      /* @__PURE__ */ jsxRuntime.jsx(
        Pagination.Pagination,
        {
          className: "table-ng-pagination",
          currentPage: page + 1,
          numberOfPages: numPages,
          showSmallVersion: smallPagination,
          onNavigate: (toPage) => {
            setPage(toPage - 1);
          }
        }
      ),
      !smallPagination && /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles$1.paginationSummary, children: /* @__PURE__ */ jsxRuntime.jsxs(i18n.Trans, { i18nKey: "grafana-ui.table.pagination-summary", children: [
        { itemsRangeStart },
        " - ",
        { displayedEnd },
        " of ",
        { numRows },
        " rows"
      ] }) })
    ] }),
    tooltipState && /* @__PURE__ */ jsxRuntime.jsx(
      DataLinksActionsTooltip.DataLinksActionsTooltip,
      {
        links: (_a = tooltipState.links) != null ? _a : [],
        actions: tooltipState.actions,
        coords: tooltipState.coords,
        onTooltipClose: () => setTooltipState(void 0)
      }
    ),
    inspectCell && /* @__PURE__ */ jsxRuntime.jsx(
      TableCellInspector.TableCellInspector,
      {
        mode: (_b = inspectCell.mode) != null ? _b : TableCellInspector.TableCellInspectorMode.text,
        value: inspectCell.value,
        onDismiss: () => setInspectCell(null)
      }
    )
  ] });
  if (utils.IS_SAFARI_26) {
    rendered = /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles$1.safariWrapper, children: rendered });
  }
  return rendered;
}
const renderRowFactory = (fields, panelContext, expandedRows, enableSharedCrosshair) => (
  // eslint-disable-next-line react/display-name
  (key, props) => {
    const { row } = props;
    const rowIdx = row.__index;
    const isExpanded = expandedRows.has(rowIdx);
    if (row.__depth === 1 && !isExpanded) {
      return null;
    }
    if (row.data) {
      return /* @__PURE__ */ jsxRuntime.jsx(reactDataGrid.Row, { "aria-level": row.__index + 1, "aria-expanded": isExpanded, ...props }, key);
    }
    const handlers = {};
    if (enableSharedCrosshair) {
      const timeField = fields.find((f) => f.type === data.FieldType.time);
      if (timeField) {
        handlers.onMouseEnter = () => {
          panelContext.eventBus.publish(
            new data.DataHoverEvent({
              point: {
                time: timeField == null ? void 0 : timeField.values[rowIdx]
              }
            })
          );
        };
        handlers.onMouseLeave = () => {
          panelContext.eventBus.publish(new data.DataHoverClearEvent());
        };
      }
    }
    return /* @__PURE__ */ jsxRuntime.jsx(reactDataGrid.Row, { ...props, ...handlers }, key);
  }
);

exports.TableNG = TableNG;
