// @grafana/runtime/dist/cjs/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('./services/index.cjs');
var config = require('./config.cjs');
var types = require('./analytics/types.cjs');
var plugin = require('./utils/plugin.cjs');
var utils = require('./analytics/utils.cjs');
var licensing = require('./utils/licensing.cjs');
var logging = require('./utils/logging.cjs');
var DataSourceWithBackend = require('./utils/DataSourceWithBackend.cjs');
var queryResponse = require('./utils/queryResponse.cjs');
var PanelRenderer = require('./components/PanelRenderer.cjs');
var PanelDataErrorView = require('./components/PanelDataErrorView.cjs');
var toDataQueryError = require('./utils/toDataQueryError.cjs');
var QueryRunner = require('./services/QueryRunner.cjs');
var PluginPage = require('./components/PluginPage.cjs');
var DataSourcePicker = require('./components/DataSourcePicker.cjs');
var eventProperties = require('./analytics/plugins/eventProperties.cjs');
var usePluginInteractionReporter = require('./analytics/plugins/usePluginInteractionReporter.cjs');
var returnToPrevious = require('./utils/returnToPrevious.cjs');
var megaMenuOpen = require('./utils/megaMenuOpen.cjs');
var chromeHeaderHeight = require('./utils/chromeHeaderHeight.cjs');
var EmbeddedDashboard = require('./components/EmbeddedDashboard.cjs');
var rbac = require('./utils/rbac.cjs');
var QueryEditorWithMigration = require('./components/QueryEditorWithMigration.cjs');
var migrationHandler = require('./utils/migrationHandler.cjs');
var userStorage = require('./utils/userStorage.cjs');
var useFavoriteDatasources = require('./utils/useFavoriteDatasources.cjs');
var FolderPicker = require('./components/FolderPicker.cjs');
var CorrelationsService = require('./services/CorrelationsService.cjs');
var apps = require('./services/pluginMeta/apps.cjs');
var hooks = require('./services/pluginMeta/hooks.cjs');
var panels = require('./services/pluginMeta/panels.cjs');
var usePluginComponent = require('./services/pluginExtensions/usePluginComponent.cjs');
var usePluginComponents = require('./services/pluginExtensions/usePluginComponents.cjs');
var usePluginLinks = require('./services/pluginExtensions/usePluginLinks.cjs');
var usePluginFunctions = require('./services/pluginExtensions/usePluginFunctions.cjs');
var useHelpNavItem = require('./services/navigation/useHelpNavItem.cjs');
var getObservablePluginLinks = require('./services/pluginExtensions/getObservablePluginLinks.cjs');
var getObservablePluginComponents = require('./services/pluginExtensions/getObservablePluginComponents.cjs');
var utils$1 = require('./services/pluginExtensions/utils.cjs');
var user = require('./services/user.cjs');
var RuntimeDataSource = require('./services/RuntimeDataSource.cjs');
var ScopesContext = require('./services/ScopesContext.cjs');
var backendSrv = require('./services/backendSrv.cjs');
var dataSourceSrv = require('./services/dataSourceSrv.cjs');
var LocationSrv = require('./services/LocationSrv.cjs');
var EchoSrv = require('./services/EchoSrv.cjs');
var templateSrv = require('./services/templateSrv.cjs');
var data = require('@grafana/data');
var live = require('./services/live.cjs');
var LocationService = require('./services/LocationService.cjs');
var appEvents = require('./services/appEvents.cjs');

"use strict";

exports.GrafanaBootConfig = config.GrafanaBootConfig;
exports.config = config.config;
exports.MetaAnalyticsEventName = types.MetaAnalyticsEventName;
exports.isExperimentViewEvent = types.isExperimentViewEvent;
exports.isInteractionEvent = types.isInteractionEvent;
exports.isPageviewEvent = types.isPageviewEvent;
exports.getPluginImportUtils = plugin.getPluginImportUtils;
exports.loadPluginCss = plugin.loadPluginCss;
exports.setPluginImportUtils = plugin.setPluginImportUtils;
exports.reportExperimentView = utils.reportExperimentView;
exports.reportInteraction = utils.reportInteraction;
exports.reportMetaAnalytics = utils.reportMetaAnalytics;
exports.reportPageview = utils.reportPageview;
exports.featureEnabled = licensing.featureEnabled;
exports.createMonitoringLogger = logging.createMonitoringLogger;
exports.logDebug = logging.logDebug;
exports.logError = logging.logError;
exports.logInfo = logging.logInfo;
exports.logMeasurement = logging.logMeasurement;
exports.logWarning = logging.logWarning;
Object.defineProperty(exports, "DataSourceWithBackend", {
  enumerable: true,
  get: function () { return DataSourceWithBackend.DataSourceWithBackend; }
});
exports.HealthCheckError = DataSourceWithBackend.HealthCheckError;
exports.HealthStatus = DataSourceWithBackend.HealthStatus;
exports.isExpressionReference = DataSourceWithBackend.isExpressionReference;
exports.frameToMetricFindValue = queryResponse.frameToMetricFindValue;
exports.toDataQueryResponse = queryResponse.toDataQueryResponse;
Object.defineProperty(exports, "PanelRenderer", {
  enumerable: true,
  get: function () { return PanelRenderer.PanelRenderer; }
});
Object.defineProperty(exports, "PanelDataErrorView", {
  enumerable: true,
  get: function () { return PanelDataErrorView.PanelDataErrorView; }
});
exports.toDataQueryError = toDataQueryError.toDataQueryError;
exports.createQueryRunner = QueryRunner.createQueryRunner;
exports.getRunRequest = QueryRunner.getRunRequest;
exports.setQueryRunnerFactory = QueryRunner.setQueryRunnerFactory;
exports.setRunRequest = QueryRunner.setRunRequest;
Object.defineProperty(exports, "PluginPage", {
  enumerable: true,
  get: function () { return PluginPage.PluginPage; }
});
exports.DataSourcePicker = DataSourcePicker.DataSourcePicker;
exports.createDataSourcePluginEventProperties = eventProperties.createDataSourcePluginEventProperties;
exports.createPluginEventProperties = eventProperties.createPluginEventProperties;
exports.usePluginInteractionReporter = usePluginInteractionReporter.usePluginInteractionReporter;
exports.setReturnToPreviousHook = returnToPrevious.setReturnToPreviousHook;
exports.useReturnToPrevious = returnToPrevious.useReturnToPrevious;
exports.setMegaMenuOpenHook = megaMenuOpen.setMegaMenuOpenHook;
exports.useMegaMenuOpen = megaMenuOpen.useMegaMenuOpen;
exports.setChromeHeaderHeightHook = chromeHeaderHeight.setChromeHeaderHeightHook;
exports.useChromeHeaderHeight = chromeHeaderHeight.useChromeHeaderHeight;
Object.defineProperty(exports, "EmbeddedDashboard", {
  enumerable: true,
  get: function () { return EmbeddedDashboard.EmbeddedDashboard; }
});
exports.setEmbeddedDashboard = EmbeddedDashboard.setEmbeddedDashboard;
exports.hasAllPermissions = rbac.hasAllPermissions;
exports.hasAnyPermission = rbac.hasAnyPermission;
exports.hasPermission = rbac.hasPermission;
exports.hasPermissionInMetadata = rbac.hasPermissionInMetadata;
exports.QueryEditorWithMigration = QueryEditorWithMigration.QueryEditorWithMigration;
exports.isMigrationHandler = migrationHandler.isMigrationHandler;
exports.migrateQuery = migrationHandler.migrateQuery;
exports.migrateRequest = migrationHandler.migrateRequest;
exports.usePluginUserStorage = userStorage.usePluginUserStorage;
exports.useFavoriteDatasources = useFavoriteDatasources.useFavoriteDatasources;
exports.FolderPicker = FolderPicker.FolderPicker;
exports.setFolderPicker = FolderPicker.setFolderPicker;
exports.getCorrelationsService = CorrelationsService.getCorrelationsService;
exports.setCorrelationsService = CorrelationsService.setCorrelationsService;
exports.getAppPluginVersion = apps.getAppPluginVersion;
exports.isAppPluginInstalled = apps.isAppPluginInstalled;
exports.useAppPluginInstalled = hooks.useAppPluginInstalled;
exports.useAppPluginVersion = hooks.useAppPluginVersion;
exports.useListedPanelPluginIds = hooks.useListedPanelPluginIds;
exports.usePanelPluginInstalled = hooks.usePanelPluginInstalled;
exports.usePanelPluginVersion = hooks.usePanelPluginVersion;
exports.getListedPanelPluginIds = panels.getListedPanelPluginIds;
exports.getPanelPluginVersion = panels.getPanelPluginVersion;
exports.isPanelPluginInstalled = panels.isPanelPluginInstalled;
exports.setPluginComponentHook = usePluginComponent.setPluginComponentHook;
exports.usePluginComponent = usePluginComponent.usePluginComponent;
exports.setPluginComponentsHook = usePluginComponents.setPluginComponentsHook;
exports.usePluginComponents = usePluginComponents.usePluginComponents;
exports.setPluginLinksHook = usePluginLinks.setPluginLinksHook;
exports.usePluginLinks = usePluginLinks.usePluginLinks;
exports.setPluginFunctionsHook = usePluginFunctions.setPluginFunctionsHook;
exports.usePluginFunctions = usePluginFunctions.usePluginFunctions;
exports.setHelpNavItemHook = useHelpNavItem.setHelpNavItemHook;
exports.useHelpNavItem = useHelpNavItem.useHelpNavItem;
exports.getObservablePluginLinks = getObservablePluginLinks.getObservablePluginLinks;
exports.getObservablePluginComponents = getObservablePluginComponents.getObservablePluginComponents;
exports.getLimitedComponentsToRender = utils$1.getLimitedComponentsToRender;
exports.isPluginExtensionComponent = utils$1.isPluginExtensionComponent;
exports.isPluginExtensionLink = utils$1.isPluginExtensionLink;
exports.renderLimitedComponents = utils$1.renderLimitedComponents;
exports.setCurrentUser = user.setCurrentUser;
exports.RuntimeDataSource = RuntimeDataSource.RuntimeDataSource;
exports.ScopesContext = ScopesContext.ScopesContext;
exports.useScopes = ScopesContext.useScopes;
exports.getBackendSrv = backendSrv.getBackendSrv;
exports.isFetchError = backendSrv.isFetchError;
exports.setBackendSrv = backendSrv.setBackendSrv;
exports.getDataSourceSrv = dataSourceSrv.getDataSourceSrv;
exports.setDataSourceSrv = dataSourceSrv.setDataSourceSrv;
exports.getLocationSrv = LocationSrv.getLocationSrv;
exports.setLocationSrv = LocationSrv.setLocationSrv;
exports.EchoEventType = EchoSrv.EchoEventType;
exports.FakeEchoSrv = EchoSrv.FakeEchoSrv;
exports.getEchoSrv = EchoSrv.getEchoSrv;
exports.registerEchoBackend = EchoSrv.registerEchoBackend;
exports.setEchoSrv = EchoSrv.setEchoSrv;
exports.getTemplateSrv = templateSrv.getTemplateSrv;
exports.setTemplateSrv = templateSrv.setTemplateSrv;
Object.defineProperty(exports, "StreamingFrameAction", {
  enumerable: true,
  get: function () { return data.StreamingFrameAction; }
});
exports.getGrafanaLiveSrv = live.getGrafanaLiveSrv;
exports.setGrafanaLiveSrv = live.setGrafanaLiveSrv;
exports.HistoryWrapper = LocationService.HistoryWrapper;
exports.LocationServiceProvider = LocationService.LocationServiceProvider;
exports.locationSearchToObject = LocationService.locationSearchToObject;
Object.defineProperty(exports, "locationService", {
  enumerable: true,
  get: function () { return LocationService.locationService; }
});
exports.navigationLogger = LocationService.navigationLogger;
exports.setLocationService = LocationService.setLocationService;
exports.useLocationService = LocationService.useLocationService;
exports.CopyPanelEvent = appEvents.CopyPanelEvent;
exports.RefreshEvent = appEvents.RefreshEvent;
exports.ThemeChangedEvent = appEvents.ThemeChangedEvent;
exports.TimeRangeUpdatedEvent = appEvents.TimeRangeUpdatedEvent;
exports.getAppEvents = appEvents.getAppEvents;
exports.setAppEvents = appEvents.setAppEvents;
