// @grafana/ui/dist/cjs/graveyard/Graph/GraphWithLegend.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var CustomScrollbar = require('../../components/CustomScrollbar/CustomScrollbar.cjs');
var VizLegend = require('../../components/VizLegend/VizLegend.cjs');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var Graph = require('./Graph.cjs');

"use strict";
const shouldHideLegendItem = (data, hideEmpty = false, hideZero = false) => {
  const isZeroOnlySeries = data.reduce((acc, current) => acc + (current[1] || 0), 0) === 0;
  const isNullOnlySeries = !data.reduce((acc, current) => acc && current[1] !== null, true);
  return hideEmpty && isNullOnlySeries || hideZero && isZeroOnlySeries;
};
const GraphWithLegend = (props) => {
  const {
    series,
    timeRange,
    width,
    height,
    showBars,
    showLines,
    showPoints,
    sortLegendBy,
    sortLegendDesc,
    legendDisplayMode,
    legendVisibility,
    placement,
    onSeriesToggle,
    onToggleSort,
    hideEmpty,
    hideZero,
    isStacked,
    lineWidth,
    onHorizontalRegionSelected,
    timeZone,
    children,
    ariaLabel
  } = props;
  const { graphContainer, wrapper, legendContainer } = ThemeContext.useStyles2(getGraphWithLegendStyles, props.placement);
  const legendItems = series.reduce((acc, s) => {
    return shouldHideLegendItem(s.data, hideEmpty, hideZero) ? acc : acc.concat([
      {
        label: s.label,
        color: s.color || "",
        disabled: !s.isVisible,
        yAxis: s.yAxis.index,
        getDisplayValues: () => s.info || []
      }
    ]);
  }, []);
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: wrapper, "aria-label": ariaLabel, children: [
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: graphContainer, children: /* @__PURE__ */ jsxRuntime.jsx(
      Graph.Graph,
      {
        series,
        timeRange,
        timeZone,
        showLines,
        showPoints,
        showBars,
        width,
        height,
        isStacked,
        lineWidth,
        onHorizontalRegionSelected,
        children
      }
    ) }),
    legendVisibility && /* @__PURE__ */ jsxRuntime.jsx("div", { className: legendContainer, children: /* @__PURE__ */ jsxRuntime.jsx(CustomScrollbar.CustomScrollbar, { hideHorizontalTrack: true, children: /* @__PURE__ */ jsxRuntime.jsx(
      VizLegend.VizLegend,
      {
        items: legendItems,
        displayMode: legendDisplayMode,
        placement,
        sortBy: sortLegendBy,
        sortDesc: sortLegendDesc,
        onLabelClick: (item, event) => {
          if (onSeriesToggle) {
            onSeriesToggle(item.label, event);
          }
        },
        onToggleSort
      }
    ) }) })
  ] });
};
const getGraphWithLegendStyles = (_theme, placement) => ({
  wrapper: css.css({
    display: "flex",
    flexDirection: placement === "bottom" ? "column" : "row"
  }),
  graphContainer: css.css({
    minHeight: "65%",
    flexGrow: 1
  }),
  legendContainer: css.css({
    padding: "10px 0",
    maxHeight: placement === "bottom" ? "35%" : "none"
  })
});

exports.GraphWithLegend = GraphWithLegend;
