// @grafana/ui/dist/cjs/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var FormField = require('./components/FormField/FormField.cjs');
var Input = require('./components/Forms/Legacy/Input/Input.cjs');
var IndicatorsContainer = require('./components/Forms/Legacy/Select/IndicatorsContainer.cjs');
var NoOptionsMessage = require('./components/Forms/Legacy/Select/NoOptionsMessage.cjs');
var Select$1 = require('./components/Forms/Legacy/Select/Select.cjs');
var Switch$1 = require('./components/Forms/Legacy/Switch/Switch.cjs');
var SecretFormField = require('./components/SecretFormField/SecretFormField.cjs');
var index = require('./options/builder/index.cjs');
var mixins = require('./themes/mixins.cjs');
var dom = require('./utils/dom.cjs');
var floating = require('./utils/floating.cjs');
var reactUtils = require('./utils/reactUtils.cjs');
var Icon = require('./components/Icon/Icon.cjs');
var IconButton = require('./components/IconButton/IconButton.cjs');
var ConfirmButton = require('./components/ConfirmButton/ConfirmButton.cjs');
var DeleteButton = require('./components/ConfirmButton/DeleteButton.cjs');
var Tooltip = require('./components/Tooltip/Tooltip.cjs');
var PopoverController = require('./components/Tooltip/PopoverController.cjs');
var Popover = require('./components/Tooltip/Popover.cjs');
var Toggletip = require('./components/Toggletip/Toggletip.cjs');
var Portal = require('./components/Portal/Portal.cjs');
var CustomScrollbar = require('./components/CustomScrollbar/CustomScrollbar.cjs');
var TabbedContainer = require('./components/TabbedContainer/TabbedContainer.cjs');
var ClipboardButton = require('./components/ClipboardButton/ClipboardButton.cjs');
var Cascader = require('./components/Cascader/Cascader.cjs');
var Carousel = require('./components/Carousel/Carousel.cjs');
var ButtonCascader = require('./components/ButtonCascader/ButtonCascader.cjs');
var InlineToast = require('./components/InlineToast/InlineToast.cjs');
var LoadingPlaceholder = require('./components/LoadingPlaceholder/LoadingPlaceholder.cjs');
var LoadingBar = require('./components/LoadingBar/LoadingBar.cjs');
var ColorPicker = require('./components/ColorPicker/ColorPicker.cjs');
var ColorPickerInput = require('./components/ColorPicker/ColorPickerInput.cjs');
var SeriesColorPickerPopover = require('./components/ColorPicker/SeriesColorPickerPopover.cjs');
var EmptySearchResult = require('./components/EmptySearchResult/EmptySearchResult.cjs');
var EmptyState = require('./components/EmptyState/EmptyState.cjs');
var UnitPicker = require('./components/UnitPicker/UnitPicker.cjs');
var StatsPicker = require('./components/StatsPicker/StatsPicker.cjs');
var RefreshPicker = require('./components/RefreshPicker/RefreshPicker.cjs');
var TimeRangePicker = require('./components/DateTimePickers/TimeRangePicker.cjs');
var TimeRangeContext = require('./components/DateTimePickers/TimeRangeContext.cjs');
var TimeRangeLabel = require('./components/DateTimePickers/TimeRangePicker/TimeRangeLabel.cjs');
var TimeOfDayPicker = require('./components/DateTimePickers/TimeOfDayPicker.cjs');
var TimeZonePicker = require('./components/DateTimePickers/TimeZonePicker.cjs');
var WeekStartPicker = require('./components/DateTimePickers/WeekStartPicker.cjs');
var DatePicker = require('./components/DateTimePickers/DatePicker/DatePicker.cjs');
var DatePickerWithInput = require('./components/DateTimePickers/DatePickerWithInput/DatePickerWithInput.cjs');
var DateTimePicker = require('./components/DateTimePickers/DateTimePicker/DateTimePicker.cjs');
var List = require('./components/List/List.cjs');
var InteractiveTable = require('./components/InteractiveTable/InteractiveTable.cjs');
var TagsInput = require('./components/TagsInput/TagsInput.cjs');
var AutoSaveField = require('./components/AutoSaveField/AutoSaveField.cjs');
var Pagination = require('./components/Pagination/Pagination.cjs');
var Tag = require('./components/Tags/Tag.cjs');
var TagList = require('./components/Tags/TagList.cjs');
var FilterPill = require('./components/FilterPill/FilterPill.cjs');
var ConfirmModal = require('./components/ConfirmModal/ConfirmModal.cjs');
var QueryField = require('./components/QueryField/QueryField.cjs');
var CodeEditor = require('./components/Monaco/CodeEditor.cjs');
var ReactMonacoEditorLazy = require('./components/Monaco/ReactMonacoEditorLazy.cjs');
var types = require('./components/Monaco/types.cjs');
var utils = require('./components/Monaco/utils.cjs');
var Modal = require('./components/Modal/Modal.cjs');
var ModalHeader = require('./components/Modal/ModalHeader.cjs');
var ModalTabsHeader = require('./components/Modal/ModalTabsHeader.cjs');
var ModalTabContent = require('./components/Modal/ModalTabContent.cjs');
var ModalsContext = require('./components/Modal/ModalsContext.cjs');
var PageToolbar = require('./components/PageLayout/PageToolbar.cjs');
var SetInterval = require('./components/SetInterval/SetInterval.cjs');
require('./components/Table/Table.cjs');
var TableCellInspector = require('./components/Table/TableCellInspector.cjs');
var TableInputCSV = require('./components/TableInputCSV/TableInputCSV.cjs');
var TabsBar = require('./components/Tabs/TabsBar.cjs');
var Tab = require('./components/Tabs/Tab.cjs');
var VerticalTab = require('./components/Tabs/VerticalTab.cjs');
var TabContent = require('./components/Tabs/TabContent.cjs');
var Counter = require('./components/Tabs/Counter.cjs');
var RenderUserContentAsHTML = require('./components/RenderUserContentAsHTML/RenderUserContentAsHTML.cjs');
var BigValue = require('./components/BigValue/BigValue.cjs');
var Sparkline = require('./components/Sparkline/Sparkline.cjs');
var Gauge = require('./components/Gauge/Gauge.cjs');
var BarGauge = require('./components/BarGauge/BarGauge.cjs');
require('./components/VizTooltip/index.cjs');
var VizRepeater = require('./components/VizRepeater/VizRepeater.cjs');
var index$1 = require('./components/PanelChrome/index.cjs');
var VizLayout = require('./components/VizLayout/VizLayout.cjs');
var types$1 = require('./components/VizLegend/types.cjs');
var VizLegend = require('./components/VizLegend/VizLegend.cjs');
var VizLegendListItem = require('./components/VizLegend/VizLegendListItem.cjs');
var Alert = require('./components/Alert/Alert.cjs');
var Collapse = require('./components/Collapse/Collapse.cjs');
var CollapsableSection = require('./components/Collapse/CollapsableSection.cjs');
var DataLinkButton = require('./components/DataLinks/DataLinkButton.cjs');
var FieldLinkList = require('./components/DataLinks/FieldLinkList.cjs');
var FullWidthButtonContainer = require('./components/Button/FullWidthButtonContainer.cjs');
var ClickOutsideWrapper = require('./components/ClickOutsideWrapper/ClickOutsideWrapper.cjs');
var SingleStatBaseOptions = require('./components/SingleStatShared/SingleStatBaseOptions.cjs');
var CallToActionCard = require('./components/CallToActionCard/CallToActionCard.cjs');
var ContextMenu = require('./components/ContextMenu/ContextMenu.cjs');
var Menu = require('./components/Menu/Menu.cjs');
var MenuGroup = require('./components/Menu/MenuGroup.cjs');
var MenuItem = require('./components/Menu/MenuItem.cjs');
var WithContextMenu = require('./components/ContextMenu/WithContextMenu.cjs');
var DataLinksInlineEditor = require('./components/DataLinks/DataLinksInlineEditor/DataLinksInlineEditor.cjs');
var DataLinksInlineEditorBase = require('./components/DataLinks/DataLinksInlineEditor/DataLinksInlineEditorBase.cjs');
var DataLinkInput = require('./components/DataLinks/DataLinkInput.cjs');
var DataLinksContextMenu = require('./components/DataLinks/DataLinksContextMenu.cjs');
var SeriesIcon = require('./components/VizLegend/SeriesIcon.cjs');
var InfoBox = require('./components/InfoBox/InfoBox.cjs');
var FeatureInfoBox = require('./components/InfoBox/FeatureInfoBox.cjs');
var FeatureBadge = require('./components/FeatureBadge/FeatureBadge.cjs');
var JSONFormatter = require('./components/JSONFormatter/JSONFormatter.cjs');
var json_explorer = require('./components/JSONFormatter/json_explorer/json_explorer.cjs');
var ErrorBoundary = require('./components/ErrorBoundary/ErrorBoundary.cjs');
var ErrorWithStack = require('./components/ErrorBoundary/ErrorWithStack.cjs');
var DataSourceHttpSettings = require('./components/DataSourceSettings/DataSourceHttpSettings.cjs');
var CustomHeadersSettings = require('./components/DataSourceSettings/CustomHeadersSettings.cjs');
var AlertingSettings = require('./components/DataSourceSettings/AlertingSettings.cjs');
var SecureSocksProxySettings = require('./components/DataSourceSettings/SecureSocksProxySettings.cjs');
var TLSAuthSettings = require('./components/DataSourceSettings/TLSAuthSettings.cjs');
var CertificationKey = require('./components/DataSourceSettings/CertificationKey.cjs');
var Spinner = require('./components/Spinner/Spinner.cjs');
var FadeTransition = require('./components/transitions/FadeTransition.cjs');
var SlideOutTransition = require('./components/transitions/SlideOutTransition.cjs');
var Segment = require('./components/Segment/Segment.cjs');
var SegmentAsync = require('./components/Segment/SegmentAsync.cjs');
var SegmentInput = require('./components/Segment/SegmentInput.cjs');
var SegmentSelect = require('./components/Segment/SegmentSelect.cjs');
var SegmentSection = require('./components/Segment/SegmentSection.cjs');
var Drawer = require('./components/Drawer/Drawer.cjs');
var Slider = require('./components/Slider/Slider.cjs');
var RangeSlider = require('./components/Slider/RangeSlider.cjs');
var Form = require('./components/Forms/Form.cjs');
var commonStyles = require('./components/Forms/commonStyles.cjs');
var InputControl = require('./components/InputControl.cjs');
var Button = require('./components/Button/Button.cjs');
var ButtonGroup = require('./components/Button/ButtonGroup.cjs');
var ToolbarButton = require('./components/ToolbarButton/ToolbarButton.cjs');
var ToolbarButtonRow = require('./components/ToolbarButton/ToolbarButtonRow.cjs');
var ValuePicker = require('./components/ValuePicker/ValuePicker.cjs');
var fieldMatchersUI = require('./components/MatchersUI/fieldMatchersUI.cjs');
var Link = require('./components/Link/Link.cjs');
var TextLink = require('./components/Link/TextLink.cjs');
var Text = require('./components/Text/Text.cjs');
var Box = require('./components/Layout/Box/Box.cjs');
var Stack = require('./components/Layout/Stack/Stack.cjs');
var Grid = require('./components/Layout/Grid/Grid.cjs');
var Space = require('./components/Layout/Space.cjs');
var ScrollContainer = require('./components/ScrollContainer/ScrollContainer.cjs');
var Label = require('./components/Forms/Label.cjs');
var Field = require('./components/Forms/Field.cjs');
var Legend = require('./components/Forms/Legend.cjs');
var FieldSet = require('./components/Forms/FieldSet.cjs');
var FieldValidationMessage = require('./components/Forms/FieldValidationMessage.cjs');
var InlineField = require('./components/Forms/InlineField.cjs');
var InlineSegmentGroup = require('./components/Forms/InlineSegmentGroup.cjs');
var InlineLabel = require('./components/Forms/InlineLabel.cjs');
var InlineFieldRow = require('./components/Forms/InlineFieldRow.cjs');
var FieldArray = require('./components/Forms/FieldArray.cjs');
var resetSelectStyles = require('./components/Select/resetSelectStyles.cjs');
var Select = require('./components/Select/Select.cjs');
var SelectMenu = require('./components/Select/SelectMenu.cjs');
var getSelectStyles = require('./components/Select/getSelectStyles.cjs');
var types$3 = require('./components/Select/types.cjs');
var Combobox = require('./components/Combobox/Combobox.cjs');
var MultiCombobox = require('./components/Combobox/MultiCombobox.cjs');
var Layout = require('./components/Layout/Layout.cjs');
var Badge = require('./components/Badge/Badge.cjs');
var RadioButtonGroup = require('./components/Forms/RadioButtonGroup/RadioButtonGroup.cjs');
var RadioButtonDot = require('./components/Forms/RadioButtonList/RadioButtonDot.cjs');
var RadioButtonList = require('./components/Forms/RadioButtonList/RadioButtonList.cjs');
var Input$1 = require('./components/Input/Input.cjs');
var AutoSizeInput = require('./components/Input/AutoSizeInput.cjs');
var FilterInput = require('./components/FilterInput/FilterInput.cjs');
require('./components/SecretInput/index.cjs');
require('./components/SecretTextArea/index.cjs');
var Switch = require('./components/Switch/Switch.cjs');
var Checkbox = require('./components/Forms/Checkbox.cjs');
var TextArea = require('./components/TextArea/TextArea.cjs');
var FileUpload = require('./components/FileUpload/FileUpload.cjs');
var FileDropzone = require('./components/FileDropzone/FileDropzone.cjs');
var FileListItem = require('./components/FileDropzone/FileListItem.cjs');
var TimeRangeInput = require('./components/DateTimePickers/TimeRangeInput.cjs');
var RelativeTimeRangePicker = require('./components/DateTimePickers/RelativeTimeRangePicker/RelativeTimeRangePicker.cjs');
var Card = require('./components/Card/Card.cjs');
var CardContainer = require('./components/Card/CardContainer.cjs');
var FormattedValueDisplay = require('./components/FormattedValueDisplay/FormattedValueDisplay.cjs');
var ButtonSelect = require('./components/Dropdown/ButtonSelect.cjs');
var Dropdown = require('./components/Dropdown/Dropdown.cjs');
var PluginSignatureBadge = require('./components/PluginSignatureBadge/PluginSignatureBadge.cjs');
var UserIcon = require('./components/UsersIndicator/UserIcon.cjs');
var UsersIndicator = require('./components/UsersIndicator/UsersIndicator.cjs');
var Avatar = require('./components/UsersIndicator/Avatar.cjs');
var FormLabel = require('./components/FormLabel/FormLabel.cjs');
var Divider = require('./components/Divider/Divider.cjs');
var DragHandle = require('./components/DragHandle/DragHandle.cjs');
var useSplitter = require('./components/Splitter/useSplitter.cjs');
var config = require('./components/uPlot/config.cjs');
var UPlotConfigBuilder = require('./components/uPlot/config/UPlotConfigBuilder.cjs');
var UPlotAxisBuilder = require('./components/uPlot/config/UPlotAxisBuilder.cjs');
var Plot = require('./components/uPlot/Plot.cjs');
var PlotLegend = require('./components/uPlot/PlotLegend.cjs');
var XYCanvas = require('./components/uPlot/geometries/XYCanvas.cjs');
var Marker = require('./components/uPlot/geometries/Marker.cjs');
var EventsCanvas = require('./components/uPlot/geometries/EventsCanvas.cjs');
var TooltipPlugin2 = require('./components/uPlot/plugins/TooltipPlugin2.cjs');
var EventBusPlugin = require('./components/uPlot/plugins/EventBusPlugin.cjs');
var KeyboardPlugin = require('./components/uPlot/plugins/KeyboardPlugin.cjs');
var XAxisInteractionAreaPlugin = require('./components/uPlot/plugins/XAxisInteractionAreaPlugin.cjs');
var types$2 = require('./components/uPlot/types.cjs');
var types$4 = require('./components/PanelChrome/types.cjs');
var Label$1 = require('./components/BrowserLabel/Label.cjs');
var PanelContainer = require('./components/PanelContainer/PanelContainer.cjs');
var VariablesInputModal = require('./components/Actions/VariablesInputModal.cjs');
var Graph = require('./graveyard/Graph/Graph.cjs');
var GraphWithLegend = require('./graveyard/Graph/GraphWithLegend.cjs');
var GraphContextMenu = require('./graveyard/Graph/GraphContextMenu.cjs');
var utils$1 = require('./graveyard/Graph/utils.cjs');
var GraphSeriesToggler = require('./graveyard/Graph/GraphSeriesToggler.cjs');
var GraphNG = require('./graveyard/GraphNG/GraphNG.cjs');
var TimeSeries = require('./graveyard/TimeSeries/TimeSeries.cjs');
var hooks = require('./graveyard/GraphNG/hooks.cjs');
var utils$2 = require('./graveyard/GraphNG/utils.cjs');
var ZoomPlugin = require('./graveyard/uPlot/plugins/ZoomPlugin.cjs');
var TooltipPlugin = require('./graveyard/uPlot/plugins/TooltipPlugin.cjs');
var ElementSelectionContext = require('./components/ElementSelectionContext/ElementSelectionContext.cjs');
var completion = require('./types/completion.cjs');
var icon = require('./types/icon.cjs');
var colors = require('./utils/colors.cjs');
var validate = require('./utils/validate.cjs');
var slate = require('./utils/slate.cjs');
var dataLinks = require('./utils/dataLinks.cjs');
var tags = require('./utils/tags.cjs');
var scrollbar = require('./utils/scrollbar.cjs');
var table = require('./utils/table.cjs');
var measureText = require('./utils/measureText.cjs');
var usePointerDistance = require('./utils/usePointerDistance.cjs');
var useForceUpdate = require('./utils/useForceUpdate.cjs');
var searchFunctions = require('./utils/searchFunctions.cjs');
var logger = require('./utils/logger.cjs');
var debug = require('./utils/debug.cjs');
var nodeGraph = require('./utils/nodeGraph.cjs');
var fuzzy = require('./utils/fuzzy.cjs');
var logOptions = require('./utils/logOptions.cjs');
var data = require('@grafana/data');
var ThemeContext = require('./themes/ThemeContext.cjs');
var getTheme = require('./themes/getTheme.cjs');
var stylesFactory = require('./themes/stylesFactory.cjs');
var GlobalStyles = require('./themes/GlobalStyles/GlobalStyles.cjs');
var braces = require('./slate-plugins/braces.cjs');
var clear = require('./slate-plugins/clear.cjs');
var clipboard = require('./slate-plugins/clipboard.cjs');
var indentation = require('./slate-plugins/indentation.cjs');
var newline = require('./slate-plugins/newline.cjs');
var runner = require('./slate-plugins/runner.cjs');
var selection_shortcuts = require('./slate-plugins/selection_shortcuts.cjs');
var index$2 = require('./slate-plugins/slate-prism/index.cjs');
var suggestions = require('./slate-plugins/suggestions.cjs');
var Sidebar = require('./components/Sidebar/Sidebar.cjs');
var schema = require('@grafana/schema');
var Table = require('./components/Table/TableRT/Table.cjs');
var VizTooltip = require('./components/VizTooltip/VizTooltip.cjs');
var VizTooltipContainer = require('./components/VizTooltip/VizTooltipContainer.cjs');
var SeriesTable = require('./components/VizTooltip/SeriesTable.cjs');
var LoadingIndicator = require('./components/PanelChrome/LoadingIndicator.cjs');
var PanelContext = require('./components/PanelChrome/PanelContext.cjs');
var useSidebar = require('./components/Sidebar/useSidebar.cjs');
var SelectContainer = require('./components/Select/SelectContainer.cjs');
var SecretInput = require('./components/SecretInput/SecretInput.cjs');
var SecretTextArea = require('./components/SecretTextArea/SecretTextArea.cjs');

"use strict";
const LegacyForms = {
  SecretFormField: SecretFormField.SecretFormField,
  FormField: FormField.FormField,
  Select: Select$1.Select,
  AsyncSelect: Select$1.AsyncSelect,
  IndicatorsContainer: IndicatorsContainer.IndicatorsContainer,
  NoOptionsMessage: NoOptionsMessage.NoOptionsMessage,
  Input: Input.Input,
  Switch: Switch$1.Switch
};

exports.LegacyInputStatus = Input.LegacyInputStatus;
exports.commonOptionsBuilder = index;
exports.styleMixins = mixins;
exports.DOMUtil = dom;
exports.floatingUtils = floating;
exports.ReactUtils = reactUtils;
exports.Icon = Icon.Icon;
exports.IconButton = IconButton.IconButton;
exports.ConfirmButton = ConfirmButton.ConfirmButton;
exports.DeleteButton = DeleteButton.DeleteButton;
exports.Tooltip = Tooltip.Tooltip;
exports.PopoverController = PopoverController.PopoverController;
exports.Popover = Popover.Popover;
exports.Toggletip = Toggletip.Toggletip;
exports.Portal = Portal.Portal;
exports.PortalContainer = Portal.PortalContainer;
exports.getPortalContainer = Portal.getPortalContainer;
exports.CustomScrollbar = CustomScrollbar.CustomScrollbar;
exports.TabbedContainer = TabbedContainer.TabbedContainer;
exports.ClipboardButton = ClipboardButton.ClipboardButton;
exports.Cascader = Cascader.Cascader;
exports.Carousel = Carousel.Carousel;
exports.ButtonCascader = ButtonCascader.ButtonCascader;
exports.InlineToast = InlineToast.InlineToast;
exports.LoadingPlaceholder = LoadingPlaceholder.LoadingPlaceholder;
exports.LoadingBar = LoadingBar.LoadingBar;
exports.ColorPicker = ColorPicker.ColorPicker;
exports.SeriesColorPicker = ColorPicker.SeriesColorPicker;
exports.ColorPickerInput = ColorPickerInput.ColorPickerInput;
exports.SeriesColorPickerPopover = SeriesColorPickerPopover.SeriesColorPickerPopover;
exports.SeriesColorPickerPopoverWithTheme = SeriesColorPickerPopover.SeriesColorPickerPopoverWithTheme;
exports.EmptySearchResult = EmptySearchResult.EmptySearchResult;
exports.EmptyState = EmptyState.EmptyState;
exports.UnitPicker = UnitPicker.UnitPicker;
exports.StatsPicker = StatsPicker.StatsPicker;
exports.RefreshPicker = RefreshPicker.RefreshPicker;
exports.defaultIntervals = RefreshPicker.defaultIntervals;
exports.TimePickerTooltip = TimeRangePicker.TimePickerTooltip;
exports.TimeRangePicker = TimeRangePicker.TimeRangePicker;
exports.TimeRangeProvider = TimeRangeContext.TimeRangeProvider;
exports.TimeRangeLabel = TimeRangeLabel.TimeRangeLabel;
exports.TimeOfDayPicker = TimeOfDayPicker.TimeOfDayPicker;
exports.TimeZonePicker = TimeZonePicker.TimeZonePicker;
exports.WeekStartPicker = WeekStartPicker.WeekStartPicker;
exports.getWeekStart = WeekStartPicker.getWeekStart;
exports.isWeekStart = WeekStartPicker.isWeekStart;
exports.DatePicker = DatePicker.DatePicker;
exports.DatePickerWithInput = DatePickerWithInput.DatePickerWithInput;
exports.DateTimePicker = DateTimePicker.DateTimePicker;
exports.List = List.List;
exports.InteractiveTable = InteractiveTable.InteractiveTable;
exports.TagsInput = TagsInput.TagsInput;
exports.AutoSaveField = AutoSaveField.AutoSaveField;
exports.Pagination = Pagination.Pagination;
exports.Tag = Tag.Tag;
exports.TagList = TagList.TagList;
exports.FilterPill = FilterPill.FilterPill;
exports.ConfirmModal = ConfirmModal.ConfirmModal;
exports.QueryField = QueryField.QueryField;
exports.CodeEditor = CodeEditor.CodeEditor;
exports.ReactMonacoEditor = ReactMonacoEditorLazy.ReactMonacoEditorLazy;
exports.CodeEditorSuggestionItemKind = types.CodeEditorSuggestionItemKind;
exports.variableSuggestionToCodeEditorSuggestion = utils.variableSuggestionToCodeEditorSuggestion;
exports.Modal = Modal.Modal;
exports.ModalHeader = ModalHeader.ModalHeader;
exports.ModalTabsHeader = ModalTabsHeader.ModalTabsHeader;
exports.ModalTabContent = ModalTabContent.ModalTabContent;
exports.ModalRoot = ModalsContext.ModalRoot;
exports.ModalsContext = ModalsContext.ModalsContext;
exports.ModalsController = ModalsContext.ModalsController;
exports.ModalsProvider = ModalsContext.ModalsProvider;
exports.PageToolbar = PageToolbar.PageToolbar;
exports.SetInterval = SetInterval.SetInterval;
exports.TableCellInspector = TableCellInspector.TableCellInspector;
exports.TableCellInspectorMode = TableCellInspector.TableCellInspectorMode;
exports.TableInputCSV = TableInputCSV.TableInputCSV;
exports.TabsBar = TabsBar.TabsBar;
exports.Tab = Tab.Tab;
exports.VerticalTab = VerticalTab.VerticalTab;
exports.TabContent = TabContent.TabContent;
exports.Counter = Counter.Counter;
exports.RenderUserContentAsHTML = RenderUserContentAsHTML.RenderUserContentAsHTML;
exports.BigValue = BigValue.BigValue;
exports.BigValueColorMode = BigValue.BigValueColorMode;
exports.BigValueGraphMode = BigValue.BigValueGraphMode;
exports.BigValueJustifyMode = BigValue.BigValueJustifyMode;
exports.BigValueTextMode = BigValue.BigValueTextMode;
exports.Sparkline = Sparkline.Sparkline;
exports.Gauge = Gauge.Gauge;
exports.BarGauge = BarGauge.BarGauge;
exports.VizRepeater = VizRepeater.VizRepeater;
exports.PanelChrome = index$1.PanelChrome;
exports.VizLayout = VizLayout.VizLayout;
exports.SeriesVisibilityChangeBehavior = types$1.SeriesVisibilityChangeBehavior;
exports.VizLegend = VizLegend.VizLegend;
exports.VizLegendListItem = VizLegendListItem.VizLegendListItem;
exports.Alert = Alert.Alert;
exports.Collapse = Collapse.Collapse;
exports.ControlledCollapse = Collapse.ControlledCollapse;
exports.CollapsableSection = CollapsableSection.CollapsableSection;
exports.DataLinkButton = DataLinkButton.DataLinkButton;
exports.FieldLinkList = FieldLinkList.FieldLinkList;
exports.FullWidthButtonContainer = FullWidthButtonContainer.FullWidthButtonContainer;
exports.ClickOutsideWrapper = ClickOutsideWrapper.ClickOutsideWrapper;
exports.convertOldAngularValueMapping = SingleStatBaseOptions.convertOldAngularValueMapping;
exports.sharedSingleStatMigrationHandler = SingleStatBaseOptions.sharedSingleStatMigrationHandler;
exports.sharedSingleStatPanelChangedHandler = SingleStatBaseOptions.sharedSingleStatPanelChangedHandler;
exports.CallToActionCard = CallToActionCard.CallToActionCard;
exports.ContextMenu = ContextMenu.ContextMenu;
exports.Menu = Menu.Menu;
exports.MenuGroup = MenuGroup.MenuGroup;
exports.MenuItem = MenuItem.MenuItem;
exports.WithContextMenu = WithContextMenu.WithContextMenu;
exports.DataLinksInlineEditor = DataLinksInlineEditor.DataLinksInlineEditor;
exports.DataLinksInlineEditorBase = DataLinksInlineEditorBase.DataLinksInlineEditorBase;
exports.DataLinkInput = DataLinkInput.DataLinkInput;
exports.DataLinksContextMenu = DataLinksContextMenu.DataLinksContextMenu;
exports.SeriesIcon = SeriesIcon.SeriesIcon;
exports.InfoBox = InfoBox.InfoBox;
exports.FeatureInfoBox = FeatureInfoBox.FeatureInfoBox;
exports.FeatureBadge = FeatureBadge.FeatureBadge;
exports.JSONFormatter = JSONFormatter.JSONFormatter;
exports.JsonExplorer = json_explorer.JsonExplorer;
exports.ErrorBoundary = ErrorBoundary.ErrorBoundary;
exports.ErrorBoundaryAlert = ErrorBoundary.ErrorBoundaryAlert;
exports.withErrorBoundary = ErrorBoundary.withErrorBoundary;
exports.ErrorWithStack = ErrorWithStack.ErrorWithStack;
exports.DataSourceHttpSettings = DataSourceHttpSettings.DataSourceHttpSettings;
exports.CustomHeadersSettings = CustomHeadersSettings.CustomHeadersSettings;
exports.AlertingSettings = AlertingSettings.AlertingSettings;
exports.SecureSocksProxySettings = SecureSocksProxySettings.SecureSocksProxySettings;
exports.TLSAuthSettings = TLSAuthSettings.TLSAuthSettings;
exports.CertificationKey = CertificationKey.CertificationKey;
exports.Spinner = Spinner.Spinner;
exports.FadeTransition = FadeTransition.FadeTransition;
exports.SlideOutTransition = SlideOutTransition.SlideOutTransition;
exports.Segment = Segment.Segment;
exports.SegmentAsync = SegmentAsync.SegmentAsync;
exports.SegmentInput = SegmentInput.SegmentInput;
exports.SegmentSelect = SegmentSelect.SegmentSelect;
exports.SegmentSection = SegmentSection.SegmentSection;
exports.Drawer = Drawer.Drawer;
exports.Slider = Slider.Slider;
exports.RangeSlider = RangeSlider.RangeSlider;
exports.Form = Form.Form;
exports.sharedInputStyle = commonStyles.sharedInputStyle;
exports.InputControl = InputControl.InputControl;
exports.Button = Button.Button;
exports.LinkButton = Button.LinkButton;
exports.clearButtonStyles = Button.clearButtonStyles;
exports.ButtonGroup = ButtonGroup.ButtonGroup;
exports.ToolbarButton = ToolbarButton.ToolbarButton;
exports.ToolbarButtonRow = ToolbarButtonRow.ToolbarButtonRow;
exports.ValuePicker = ValuePicker.ValuePicker;
exports.fieldMatchersUI = fieldMatchersUI.fieldMatchersUI;
exports.Link = Link.Link;
exports.TextLink = TextLink.TextLink;
exports.Text = Text.Text;
exports.Box = Box.Box;
exports.Stack = Stack.Stack;
exports.Grid = Grid.Grid;
exports.Space = Space.Space;
exports.ScrollContainer = ScrollContainer.ScrollContainer;
exports.Label = Label.Label;
exports.Field = Field.Field;
exports.Legend = Legend.Legend;
exports.FieldSet = FieldSet.FieldSet;
exports.FieldValidationMessage = FieldValidationMessage.FieldValidationMessage;
exports.InlineField = InlineField.InlineField;
exports.InlineSegmentGroup = InlineSegmentGroup.InlineSegmentGroup;
exports.InlineLabel = InlineLabel.InlineLabel;
exports.InlineFieldRow = InlineFieldRow.InlineFieldRow;
exports.FieldArray = FieldArray.FieldArray;
exports.resetSelectStyles = resetSelectStyles.default;
exports.AsyncMultiSelect = Select.AsyncMultiSelect;
exports.AsyncSelect = Select.AsyncSelect;
exports.AsyncVirtualizedSelect = Select.AsyncVirtualizedSelect;
exports.MultiSelect = Select.MultiSelect;
exports.Select = Select.Select;
exports.VirtualizedSelect = Select.VirtualizedSelect;
exports.SelectMenuOptions = SelectMenu.SelectMenuOptions;
exports.getSelectStyles = getSelectStyles.getSelectStyles;
exports.ToggleAllState = types$3.ToggleAllState;
exports.Combobox = Combobox.Combobox;
exports.MultiCombobox = MultiCombobox.MultiCombobox;
exports.Container = Layout.Container;
exports.HorizontalGroup = Layout.HorizontalGroup;
exports.VerticalGroup = Layout.VerticalGroup;
exports.Badge = Badge.Badge;
exports.RadioButtonGroup = RadioButtonGroup.RadioButtonGroup;
exports.RadioButtonDot = RadioButtonDot.RadioButtonDot;
exports.RadioButtonList = RadioButtonList.RadioButtonList;
exports.Input = Input$1.Input;
exports.getInputStyles = Input$1.getInputStyles;
exports.AutoSizeInput = AutoSizeInput.AutoSizeInput;
exports.FilterInput = FilterInput.FilterInput;
exports.InlineSwitch = Switch.InlineSwitch;
exports.Switch = Switch.Switch;
exports.Checkbox = Checkbox.Checkbox;
exports.TextArea = TextArea.TextArea;
exports.FileUpload = FileUpload.FileUpload;
exports.FileDropzone = FileDropzone.FileDropzone;
exports.FileDropzoneDefaultChildren = FileDropzone.FileDropzoneDefaultChildren;
exports.FileListItem = FileListItem.FileListItem;
exports.TimeRangeInput = TimeRangeInput.TimeRangeInput;
exports.RelativeTimeRangePicker = RelativeTimeRangePicker.RelativeTimeRangePicker;
exports.Card = Card.Card;
exports.getCardStyles = Card.getCardStyles;
exports.CardContainer = CardContainer.CardContainer;
exports.FormattedValueDisplay = FormattedValueDisplay.FormattedValueDisplay;
exports.ButtonSelect = ButtonSelect.ButtonSelect;
exports.Dropdown = Dropdown.Dropdown;
exports.PluginSignatureBadge = PluginSignatureBadge.PluginSignatureBadge;
exports.UserIcon = UserIcon.UserIcon;
exports.UsersIndicator = UsersIndicator.UsersIndicator;
exports.Avatar = Avatar.Avatar;
exports.InlineFormLabel = FormLabel.InlineFormLabel;
exports.Divider = Divider.Divider;
exports.getDragStyles = DragHandle.getDragStyles;
exports.useSplitter = useSplitter.useSplitter;
exports.getGraphFieldOptions = config.getGraphFieldOptions;
exports.graphFieldOptions = config.graphFieldOptions;
exports.UPlotConfigBuilder = UPlotConfigBuilder.UPlotConfigBuilder;
exports.UPLOT_AXIS_FONT_SIZE = UPlotAxisBuilder.UPLOT_AXIS_FONT_SIZE;
exports.UPlotChart = Plot.UPlotChart;
exports.PlotLegend = PlotLegend.PlotLegend;
exports.XYCanvas = XYCanvas.XYCanvas;
exports.Marker = Marker.Marker;
exports.EventsCanvas = EventsCanvas.EventsCanvas;
exports.TooltipPlugin2 = TooltipPlugin2.TooltipPlugin2;
exports.EventBusPlugin = EventBusPlugin.EventBusPlugin;
exports.KeyboardPlugin = KeyboardPlugin.KeyboardPlugin;
exports.XAxisInteractionAreaPlugin = XAxisInteractionAreaPlugin.XAxisInteractionAreaPlugin;
exports.FIXED_UNIT = types$2.FIXED_UNIT;
exports.SeriesVisibilityChangeMode = types$4.SeriesVisibilityChangeMode;
exports.BrowserLabel = Label$1.Label;
exports.PanelContainer = PanelContainer.PanelContainer;
exports.VariablesInputModal = VariablesInputModal.VariablesInputModal;
exports.Graph = Graph.Graph;
exports.GraphWithLegend = GraphWithLegend.GraphWithLegend;
exports.GraphContextMenu = GraphContextMenu.GraphContextMenu;
exports.GraphContextMenuHeader = GraphContextMenu.GraphContextMenuHeader;
exports.graphTickFormatter = utils$1.graphTickFormatter;
exports.graphTimeFormat = utils$1.graphTimeFormat;
exports.GraphSeriesToggler = GraphSeriesToggler.GraphSeriesToggler;
exports.GraphNG = GraphNG.GraphNG;
exports.TimeSeries = TimeSeries.TimeSeries;
exports.useGraphNGContext = hooks.useGraphNGContext;
exports.buildScaleKey = utils$2.buildScaleKey;
exports.preparePlotFrame = utils$2.preparePlotFrame;
exports.ZoomPlugin = ZoomPlugin.ZoomPlugin;
exports.TooltipPlugin = TooltipPlugin.TooltipPlugin;
exports.ElementSelectionContext = ElementSelectionContext.ElementSelectionContext;
exports.useElementSelection = ElementSelectionContext.useElementSelection;
exports.CompletionItemKind = completion.CompletionItemKind;
exports.getAvailableIcons = icon.getAvailableIcons;
exports.getFieldTypeIcon = icon.getFieldTypeIcon;
exports.getFieldTypeIconName = icon.getFieldTypeIconName;
exports.isIconSize = icon.isIconSize;
exports.ALERTING_COLOR = colors.ALERTING_COLOR;
exports.DEFAULT_ANNOTATION_COLOR = colors.DEFAULT_ANNOTATION_COLOR;
exports.NO_DATA_COLOR = colors.NO_DATA_COLOR;
exports.OK_COLOR = colors.OK_COLOR;
exports.PENDING_COLOR = colors.PENDING_COLOR;
exports.REGION_FILL_ALPHA = colors.REGION_FILL_ALPHA;
exports.colors = colors.colors;
exports.getTextColorForAlphaBackground = colors.getTextColorForAlphaBackground;
exports.getTextColorForBackground = colors.getTextColorForBackground;
exports.sortedColors = colors.sortedColors;
exports.EventsWithValidation = validate.EventsWithValidation;
exports.hasValidationEvent = validate.hasValidationEvent;
exports.regexValidation = validate.regexValidation;
exports.validate = validate.validate;
exports.SCHEMA = slate.SCHEMA;
exports.makeFragment = slate.makeFragment;
exports.makeValue = slate.makeValue;
exports.linkModelToContextMenuItems = dataLinks.linkModelToContextMenuItems;
exports.getTagColor = tags.getTagColor;
exports.getTagColorIndexFromName = tags.getTagColorIndexFromName;
exports.getTagColorsFromName = tags.getTagColorsFromName;
exports.getScrollbarWidth = scrollbar.getScrollbarWidth;
exports.getCellLinks = table.getCellLinks;
exports.calculateFontSize = measureText.calculateFontSize;
exports.getCanvasContext = measureText.getCanvasContext;
exports.measureText = measureText.measureText;
exports.createPointerDistance = usePointerDistance.createPointerDistance;
exports.usePointerDistance = usePointerDistance.usePointerDistance;
exports.useForceUpdate = useForceUpdate.useForceUpdate;
exports.SearchFunctionType = searchFunctions.SearchFunctionType;
exports.createLogger = logger.createLogger;
exports.attachDebugger = debug.attachDebugger;
exports.NodeGraphDataFrameFieldNames = nodeGraph.NodeGraphDataFrameFieldNames;
exports.fuzzyMatch = fuzzy.fuzzyMatch;
exports.logOptions = logOptions.logOptions;
Object.defineProperty(exports, "ThemeContext", {
  enumerable: true,
  get: function () { return data.ThemeContext; }
});
Object.defineProperty(exports, "toIconName", {
  enumerable: true,
  get: function () { return data.toIconName; }
});
exports.mockThemeContext = ThemeContext.mockThemeContext;
exports.useStyles = ThemeContext.useStyles;
exports.useStyles2 = ThemeContext.useStyles2;
exports.useTheme = ThemeContext.useTheme;
exports.useTheme2 = ThemeContext.useTheme2;
exports.withTheme = ThemeContext.withTheme;
exports.withTheme2 = ThemeContext.withTheme2;
exports.getTheme = getTheme.getTheme;
exports.mockTheme = getTheme.mockTheme;
exports.stylesFactory = stylesFactory.stylesFactory;
exports.GlobalStyles = GlobalStyles.GlobalStyles;
exports.BracesPlugin = braces.BracesPlugin;
exports.ClearPlugin = clear.ClearPlugin;
exports.ClipboardPlugin = clipboard.ClipboardPlugin;
exports.IndentationPlugin = indentation.IndentationPlugin;
exports.NewlinePlugin = newline.NewlinePlugin;
exports.RunnerPlugin = runner.RunnerPlugin;
exports.SelectionShortcutsPlugin = selection_shortcuts.SelectionShortcutsPlugin;
exports.SlatePrism = index$2.SlatePrism;
exports.SuggestionsPlugin = suggestions.SuggestionsPlugin;
exports.Sidebar = Sidebar.Sidebar;
Object.defineProperty(exports, "AxisPlacement", {
  enumerable: true,
  get: function () { return schema.AxisPlacement; }
});
Object.defineProperty(exports, "BarAlignment", {
  enumerable: true,
  get: function () { return schema.BarAlignment; }
});
Object.defineProperty(exports, "BarGaugeDisplayMode", {
  enumerable: true,
  get: function () { return schema.BarGaugeDisplayMode; }
});
Object.defineProperty(exports, "BarValueVisibility", {
  enumerable: true,
  get: function () { return schema.VisibilityMode; }
});
Object.defineProperty(exports, "DrawStyle", {
  enumerable: true,
  get: function () { return schema.GraphDrawStyle; }
});
Object.defineProperty(exports, "GraphGradientMode", {
  enumerable: true,
  get: function () { return schema.GraphGradientMode; }
});
Object.defineProperty(exports, "GraphThresholdsStyleMode", {
  enumerable: true,
  get: function () { return schema.GraphThresholdsStyleMode; }
});
Object.defineProperty(exports, "LegendDisplayMode", {
  enumerable: true,
  get: function () { return schema.LegendDisplayMode; }
});
Object.defineProperty(exports, "LineInterpolation", {
  enumerable: true,
  get: function () { return schema.LineInterpolation; }
});
Object.defineProperty(exports, "PointVisibility", {
  enumerable: true,
  get: function () { return schema.VisibilityMode; }
});
Object.defineProperty(exports, "ScaleDirection", {
  enumerable: true,
  get: function () { return schema.ScaleDirection; }
});
Object.defineProperty(exports, "ScaleDistribution", {
  enumerable: true,
  get: function () { return schema.ScaleDistribution; }
});
Object.defineProperty(exports, "ScaleOrientation", {
  enumerable: true,
  get: function () { return schema.ScaleOrientation; }
});
Object.defineProperty(exports, "StackingMode", {
  enumerable: true,
  get: function () { return schema.StackingMode; }
});
Object.defineProperty(exports, "TableCellDisplayMode", {
  enumerable: true,
  get: function () { return schema.TableCellDisplayMode; }
});
Object.defineProperty(exports, "TooltipDisplayMode", {
  enumerable: true,
  get: function () { return schema.TooltipDisplayMode; }
});
exports.Table = Table.Table;
exports.VizTooltip = VizTooltip.VizTooltip;
exports.VizTooltipContainer = VizTooltipContainer.VizTooltipContainer;
exports.SeriesTable = SeriesTable.SeriesTable;
exports.SeriesTableRow = SeriesTable.SeriesTableRow;
exports.PanelChromeLoadingIndicator = LoadingIndicator.LoadingIndicator;
exports.PanelContextProvider = PanelContext.PanelContextProvider;
exports.PanelContextRoot = PanelContext.PanelContextRoot;
exports.usePanelContext = PanelContext.usePanelContext;
exports.useSidebar = useSidebar.useSidebar;
exports.SelectContainer = SelectContainer.SelectContainer;
exports.SecretInput = SecretInput.SecretInput;
exports.SecretTextArea = SecretTextArea.SecretTextArea;
exports.LegacyForms = LegacyForms;
