// @grafana/ui/dist/cjs/components/Select/MultiValue.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var IconButton = require('../IconButton/IconButton.cjs');
var getSelectStyles = require('./getSelectStyles.cjs');

"use strict";
const MultiValueContainer = ({ innerProps, children }) => {
  const theme = ThemeContext.useTheme2();
  const styles = getSelectStyles.getSelectStyles(theme);
  return /* @__PURE__ */ jsxRuntime.jsx("div", { ...innerProps, className: styles.multiValueContainer, children });
};
const MultiValueRemove = ({ children, innerProps }) => {
  const theme = ThemeContext.useTheme2();
  const styles = getSelectStyles.getSelectStyles(theme);
  return /* @__PURE__ */ jsxRuntime.jsx(
    IconButton.IconButton,
    {
      ...innerProps,
      name: "times",
      size: "sm",
      className: styles.multiValueRemove,
      tooltip: i18n.t("grafana-ui.select.multi-value-remove", "Remove")
    }
  );
};

exports.MultiValueContainer = MultiValueContainer;
exports.MultiValueRemove = MultiValueRemove;
