// @grafana/ui/dist/cjs/components/Select/Select.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var SelectBase = require('./SelectBase.cjs');
var SelectContainer = require('./SelectContainer.cjs');

"use strict";
function Select(props) {
  return /* @__PURE__ */ jsxRuntime.jsx(SelectBase.SelectBase, { ...props });
}
function MultiSelect(props) {
  return /* @__PURE__ */ jsxRuntime.jsx(SelectBase.SelectBase, { ...props, isMulti: true });
}
function AsyncSelect(props) {
  return /* @__PURE__ */ jsxRuntime.jsx(SelectBase.SelectBase, { ...props });
}
function VirtualizedSelect(props) {
  return /* @__PURE__ */ jsxRuntime.jsx(SelectBase.SelectBase, { virtualized: true, ...props });
}
function AsyncVirtualizedSelect(props) {
  return /* @__PURE__ */ jsxRuntime.jsx(SelectBase.SelectBase, { virtualized: true, ...props });
}
function AsyncMultiSelect(props) {
  return /* @__PURE__ */ jsxRuntime.jsx(SelectBase.SelectBase, { ...props, isMulti: true });
}

exports.SelectContainer = SelectContainer.SelectContainer;
exports.AsyncMultiSelect = AsyncMultiSelect;
exports.AsyncSelect = AsyncSelect;
exports.AsyncVirtualizedSelect = AsyncVirtualizedSelect;
exports.MultiSelect = MultiSelect;
exports.Select = Select;
exports.VirtualizedSelect = VirtualizedSelect;
