// @grafana/ui/dist/cjs/components/Select/ValueContainer.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var lodash = require('lodash');
var React = require('react');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var getSelectStyles = require('./getSelectStyles.cjs');

"use strict";
class UnthemedValueContainer extends React.Component {
  constructor() {
    super(...arguments);
    this.ref = React.createRef();
  }
  componentDidUpdate(prevProps) {
    if (this.ref.current && this.props.selectProps.autoWidth && !this.props.selectProps.maxVisibleValues && !lodash.isEqual(prevProps.selectProps.value, this.props.selectProps.value)) {
      this.ref.current.style.minWidth = "0px";
      const width = this.ref.current.offsetWidth;
      this.ref.current.style.minWidth = `${width}px`;
    }
  }
  render() {
    const { children } = this.props;
    const { selectProps } = this.props;
    if (selectProps && Array.isArray(children) && Array.isArray(children[0]) && selectProps.maxVisibleValues !== void 0 && !(selectProps.showAllSelectedWhenOpen && selectProps.menuIsOpen)) {
      const [valueChildren, ...otherChildren] = children;
      const truncatedValues = valueChildren.slice(0, selectProps.maxVisibleValues);
      return this.renderContainer([truncatedValues, ...otherChildren]);
    }
    return this.renderContainer(children);
  }
  renderContainer(children) {
    var _a, _b;
    const { isMulti, theme, selectProps } = this.props;
    const noWrap = ((_a = this.props.selectProps) == null ? void 0 : _a.noMultiValueWrap) && !((_b = this.props.selectProps) == null ? void 0 : _b.menuIsOpen);
    const styles = getSelectStyles.getSelectStyles(theme);
    const dataTestid = selectProps["data-testid"];
    const className = css.cx(styles.valueContainer, {
      [styles.valueContainerMulti]: isMulti && !noWrap,
      [styles.valueContainerMultiNoWrap]: isMulti && noWrap
    });
    return /* @__PURE__ */ jsxRuntime.jsx("div", { ref: this.ref, "data-testid": dataTestid, className, children });
  }
}
const ValueContainer = ThemeContext.withTheme2(UnthemedValueContainer);

exports.ValueContainer = ValueContainer;
