// @grafana/ui/dist/esm/components/Table/TableNG/Filter/FilterPopup.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import { memo, useMemo, useState, useRef, useEffect, useCallback } from 'react';
import { selectors } from '@grafana/e2e-selectors';
import { t, Trans } from '@grafana/i18n';
import { useStyles2 } from '../../../../themes/ThemeContext.mjs';
import { Button } from '../../../Button/Button.mjs';
import { ButtonSelect } from '../../../Dropdown/ButtonSelect.mjs';
import { FilterInput } from '../../../FilterInput/FilterInput.mjs';
import { Label } from '../../../Forms/Label.mjs';
import { Stack } from '../../../Layout/Stack/Stack.mjs';
import { getDisplayName } from '../utils.mjs';
import { FilterList } from './FilterList.mjs';
import { calculateUniqueFieldValues, valuesToOptions, getFilteredOptions, operatorSelectableValues } from './utils.mjs';

"use strict";
const FilterPopup = memo(
  ({
    name,
    rows,
    filterValue,
    setFilter,
    onClose,
    field,
    searchFilter,
    setSearchFilter,
    operator,
    setOperator,
    buttonElement
  }) => {
    var _a;
    const uniqueValues = useMemo(() => calculateUniqueFieldValues(rows, field), [rows, field]);
    const options = useMemo(() => valuesToOptions(uniqueValues), [uniqueValues]);
    const filteredOptions = useMemo(() => getFilteredOptions(options, filterValue), [options, filterValue]);
    const [values, setValues] = useState(filteredOptions);
    const [matchCase, setMatchCase] = useState(false);
    const containerRef = useRef(null);
    const inputRef = useRef(null);
    const operators = Object.values(operatorSelectableValues());
    useEffect(() => {
      if (inputRef.current) {
        inputRef.current.focus();
      }
    }, []);
    useEffect(() => {
      function handleEscape(e) {
        if (e.key === "Escape") {
          onClose();
          buttonElement == null ? void 0 : buttonElement.focus();
        }
      }
      document.addEventListener("keyup", handleEscape);
      return () => {
        document.removeEventListener("keyup", handleEscape);
      };
    }, [onClose, buttonElement]);
    const onFilter = useCallback(() => {
      if (values.length !== 0) {
        const filteredSet = new Set(values.map((item) => item.value));
        setFilter((filter) => ({
          ...filter,
          [name]: { filtered: values, filteredSet, searchFilter, operator }
        }));
      } else {
        setFilter((filter) => {
          const newFilter = { ...filter };
          delete newFilter[name];
          return newFilter;
        });
      }
      onClose();
    }, [name, operator, searchFilter, setFilter, values, onClose]);
    const onClearFilter = useCallback(() => {
      setFilter((filter) => {
        const newFilter = { ...filter };
        delete newFilter[name];
        return newFilter;
      });
      onClose();
    }, [name, setFilter, onClose]);
    useEffect(() => {
      const onOutsideClick = (event) => {
        const domNode = containerRef.current;
        if (!domNode) {
          return;
        }
        if (event.target instanceof Node && !domNode.contains(event.target)) {
          console.log("closing from outside click");
          onClose();
        }
      };
      window.addEventListener("click", onOutsideClick);
      return () => {
        window.removeEventListener("click", onOutsideClick);
      };
    }, [onClose]);
    const filterInputPlaceholder = t("grafana-ui.table.filter-popup-input-placeholder", "Filter values");
    const clearFilterVisible = useMemo(() => filterValue !== void 0, [filterValue]);
    const styles = useStyles2(getStyles);
    return /* @__PURE__ */ jsx(
      "div",
      {
        className: styles.filterContainer,
        "data-testid": selectors.components.Panels.Visualization.TableNG.Filters.Container,
        ref: containerRef,
        children: /* @__PURE__ */ jsxs(Stack, { direction: "column", children: [
          /* @__PURE__ */ jsx(Stack, { alignItems: "center", children: field && /* @__PURE__ */ jsx(Label, { className: styles.label, children: getDisplayName(field) }) }),
          /* @__PURE__ */ jsxs(Stack, { gap: 1, children: [
            /* @__PURE__ */ jsx("div", { className: styles.inputContainer, children: /* @__PURE__ */ jsx(
              FilterInput,
              {
                ref: inputRef,
                placeholder: filterInputPlaceholder,
                title: filterInputPlaceholder,
                onChange: setSearchFilter,
                value: searchFilter,
                suffix: /* @__PURE__ */ jsx(
                  ButtonSelect,
                  {
                    className: styles.buttonSelectOverrides,
                    options: operators,
                    onChange: setOperator,
                    value: operator,
                    tooltip: operator.description,
                    narrow: true,
                    root: (_a = containerRef.current) != null ? _a : void 0
                  }
                )
              }
            ) }),
            /* @__PURE__ */ jsx(
              Button,
              {
                tooltip: t("grafana-ui.table.filter-popup-aria-label-match-case", "Match case"),
                variant: matchCase ? "primary" : "secondary",
                onClick: () => setMatchCase((s) => !s),
                "aria-pressed": matchCase,
                icon: "text-fields"
              }
            )
          ] }),
          /* @__PURE__ */ jsx(
            FilterList,
            {
              onChange: setValues,
              values,
              options,
              caseSensitive: matchCase,
              searchFilter,
              operator
            }
          ),
          /* @__PURE__ */ jsxs(Stack, { justifyContent: "end", direction: "row-reverse", children: [
            /* @__PURE__ */ jsx(Button, { size: "sm", onClick: onFilter, children: /* @__PURE__ */ jsx(Trans, { i18nKey: "grafana-ui.table.filter-popup-apply", children: "Ok" }) }),
            /* @__PURE__ */ jsx(Button, { size: "sm", variant: "secondary", onClick: onClose, children: /* @__PURE__ */ jsx(Trans, { i18nKey: "grafana-ui.table.filter-popup-cancel", children: "Cancel" }) }),
            clearFilterVisible && /* @__PURE__ */ jsx(Button, { fill: "text", size: "sm", onClick: onClearFilter, children: /* @__PURE__ */ jsx(Trans, { i18nKey: "grafana-ui.table.filter-popup-clear", children: "Clear filter" }) })
          ] })
        ] })
      }
    );
  }
);
FilterPopup.displayName = "FilterPopup";
const getStyles = (theme) => ({
  filterContainer: css({
    label: "filterContainer",
    width: "100%",
    minWidth: "320px",
    height: "100%",
    backgroundColor: theme.colors.background.primary,
    border: `1px solid ${theme.colors.border.weak}`,
    padding: theme.spacing(2),
    boxShadow: theme.shadows.z3,
    borderRadius: theme.shape.radius.default
  }),
  label: css({
    marginBottom: 0
  }),
  inputContainer: css({
    width: 300
  }),
  buttonSelectOverrides: css({
    fontSize: 12,
    "&:hover, &:focus, &:active": {
      color: theme.colors.text.primary,
      background: "transparent"
    }
  })
});

export { FilterPopup };
