// @grafana/ui/dist/esm/components/Table/TableNG/Filter/Filter.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import { clsx } from 'clsx';
import { memo, useRef, useState } from 'react';
import { selectors } from '@grafana/e2e-selectors';
import { t } from '@grafana/i18n';
import { useStyles2 } from '../../../../themes/ThemeContext.mjs';
import { Icon } from '../../../Icon/Icon.mjs';
import { Popover } from '../../../Tooltip/Popover.mjs';
import { FilterOperator } from '../types.mjs';
import { getDisplayName } from '../utils.mjs';
import { FilterPopup } from './FilterPopup.mjs';
import { operatorSelectableValues } from './utils.mjs';

"use strict";
const Filter = memo(
  ({ name, rows, filter, setFilter, field, crossFilterOrder, crossFilterRows, iconClassName }) => {
    var _a, _b, _c;
    const filterValue = (_a = filter[name]) == null ? void 0 : _a.filtered;
    const filterIndex = crossFilterOrder.indexOf(name);
    let filteredRows;
    if (filterIndex > 0) {
      const previousFilterName = crossFilterOrder[filterIndex - 1];
      filteredRows = crossFilterRows[previousFilterName];
    } else if (filterIndex === -1 && crossFilterOrder.length > 0) {
      const previousFilterName = crossFilterOrder[crossFilterOrder.length - 1];
      filteredRows = crossFilterRows[previousFilterName];
    } else {
      filteredRows = rows;
    }
    const ref = useRef(null);
    const [isPopoverVisible, setPopoverVisible] = useState(false);
    const styles = useStyles2(getStyles);
    const filterEnabled = Boolean(filterValue);
    const [searchFilter, setSearchFilter] = useState(((_b = filter[name]) == null ? void 0 : _b.searchFilter) || "");
    const [operator, setOperator] = useState(
      ((_c = filter[name]) == null ? void 0 : _c.operator) || operatorSelectableValues()[FilterOperator.CONTAINS]
    );
    return /* @__PURE__ */ jsxs(
      "button",
      {
        className: styles.headerFilter,
        ref,
        type: "button",
        "aria-haspopup": "dialog",
        "aria-pressed": isPopoverVisible,
        "aria-label": t("grafana-ui.table.filter.button", `Filter {{name}}`, {
          name: field ? getDisplayName(field) : ""
        }),
        "data-testid": selectors.components.Panels.Visualization.TableNG.Filters.HeaderButton,
        tabIndex: 0,
        onKeyDown: (ev) => {
          if (ev.target === ref.current && (ev.key === "Enter" || ev.key === " ")) {
            setPopoverVisible(true);
            ev.stopPropagation();
            ev.preventDefault();
          }
        },
        onClick: (ev) => {
          ev.stopPropagation();
          if (!isPopoverVisible) {
            setPopoverVisible(true);
          }
        },
        children: [
          /* @__PURE__ */ jsx(Icon, { name: "filter", className: clsx(iconClassName, filterEnabled ? styles.filterIconEnabled : void 0) }),
          isPopoverVisible && ref.current && /* @__PURE__ */ jsx(
            Popover,
            {
              content: /* @__PURE__ */ jsx(
                FilterPopup,
                {
                  name,
                  rows: filteredRows,
                  filterValue,
                  setFilter,
                  field,
                  onClose: () => setPopoverVisible(false),
                  searchFilter,
                  setSearchFilter,
                  operator,
                  setOperator,
                  buttonElement: ref.current
                }
              ),
              placement: "bottom-start",
              referenceElement: ref.current,
              show: true
            }
          )
        ]
      }
    );
  }
);
Filter.displayName = "Filter";
const getStyles = (theme) => ({
  headerFilter: css({
    background: "transparent",
    border: "none",
    label: "headerFilter",
    padding: 0,
    alignSelf: "flex-end",
    borderRadius: theme.spacing(0.25)
  }),
  filterIconEnabled: css({
    label: "filterIconEnabled",
    color: theme.colors.primary.text
  })
});

export { Filter };
