// @grafana/ui/dist/cjs/components/Gauge/utils.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var data = require('@grafana/data');
var schema = require('@grafana/schema');

"use strict";
const DEFAULT_THRESHOLDS = {
  mode: data.ThresholdsMode.Absolute,
  steps: [
    { value: -Infinity, color: "green" },
    { value: 80, color: "red" }
  ]
};
function calculateGaugeAutoProps(width, height, title, orientation) {
  const showLabel = title !== null && title !== void 0;
  const titleFontSizeDimension = orientation === schema.VizOrientation.Vertical ? height : width;
  const titleFontSize = Math.min(titleFontSizeDimension * 0.15 / 1.5, 20);
  const titleHeight = titleFontSize * 1.5;
  const availableHeight = showLabel ? height - titleHeight : height;
  const gaugeHeight = Math.min(availableHeight, width);
  return {
    showLabel,
    gaugeHeight,
    titleFontSize
  };
}
function getFormattedThresholds(decimals, field, theme, offsetColor = true) {
  var _a, _b, _c;
  const thresholds = (_a = field.thresholds) != null ? _a : DEFAULT_THRESHOLDS;
  const isPercent = thresholds.mode === data.ThresholdsMode.Percentage;
  const steps = thresholds.steps;
  let min = (_b = field.min) != null ? _b : data.GAUGE_DEFAULT_MINIMUM;
  let max = (_c = field.max) != null ? _c : data.GAUGE_DEFAULT_MAXIMUM;
  if (isPercent) {
    min = 0;
    max = 100;
  }
  const first = data.getActiveThreshold(min, steps);
  const last = data.getActiveThreshold(max, steps);
  const formatted = [];
  if (offsetColor) {
    formatted.push({
      value: parseFloat(min.toFixed(decimals)),
      color: theme.visualization.getColorByName(first.color)
    });
  }
  let skip = offsetColor;
  for (let i = 0; i < steps.length; i++) {
    const step = steps[i];
    if (skip) {
      if (first === step) {
        skip = false;
      }
      continue;
    }
    const prev = steps[i - 1];
    formatted.push({
      value: isFinite(step.value) ? step.value : 0,
      color: theme.visualization.getColorByName((offsetColor ? prev : step).color)
    });
    if (step === last) {
      break;
    }
  }
  if (max > last.value) {
    formatted.push({ value: parseFloat(max.toFixed(decimals)), color: theme.visualization.getColorByName(last.color) });
  }
  return formatted;
}

exports.DEFAULT_THRESHOLDS = DEFAULT_THRESHOLDS;
exports.calculateGaugeAutoProps = calculateGaugeAutoProps;
exports.getFormattedThresholds = getFormattedThresholds;
