// @grafana/ui/dist/cjs/components/StatsPicker/StatsPicker.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var lodash = require('lodash');
var React = require('react');
var data = require('@grafana/data');
var Select = require('../Select/Select.cjs');

"use strict";
const StatsPicker = React.memo(
  ({
    placeholder,
    onChange,
    stats,
    allowMultiple = false,
    defaultStat,
    className,
    width,
    menuPlacement,
    inputId,
    filterOptions
  }) => {
    React.useEffect(() => {
      const current = data.fieldReducers.list(stats);
      if (current.length !== stats.length) {
        const found = current.map((v) => v.id);
        const notFound = lodash.difference(stats, found);
        console.warn("Unknown stats", notFound, stats);
        onChange(current.map((stat) => stat.id));
      }
      if (!allowMultiple && stats.length > 1) {
        console.warn("Removing extra stat", stats);
        onChange([stats[0]]);
      }
      if (defaultStat && stats.length < 1) {
        onChange([defaultStat]);
      }
    }, [stats, allowMultiple, defaultStat, onChange]);
    const onSelectionChange = (item) => {
      if (Array.isArray(item)) {
        onChange(item.map((v) => v.value));
      } else {
        onChange(item && item.value ? [item.value] : []);
      }
    };
    const select = data.fieldReducers.selectOptions(stats, filterOptions);
    return /* @__PURE__ */ jsxRuntime.jsx(
      Select.Select,
      {
        value: select.current,
        className,
        isClearable: !defaultStat,
        isMulti: allowMultiple,
        width,
        isSearchable: true,
        options: select.options,
        placeholder,
        onChange: onSelectionChange,
        menuPlacement,
        inputId
      }
    );
  }
);
StatsPicker.displayName = "StatsPicker";

exports.StatsPicker = StatsPicker;
