// @grafana/schema/dist/esm/raw/composable/azuremonitor/dataquery/x/types.gen.mjs
"use strict";
const pluginVersion = "%VERSION%";
const defaultAzureMonitorQuery = {
  keepCookies: [],
  subscriptions: []
};
var AzureQueryType = /* @__PURE__ */ ((AzureQueryType2) => {
  AzureQueryType2["AzureMonitor"] = "Azure Monitor";
  AzureQueryType2["AzureResourceGraph"] = "Azure Resource Graph";
  AzureQueryType2["AzureTraces"] = "Azure Traces";
  AzureQueryType2["CustomMetricNamesQuery"] = "Azure Custom Metric Names";
  AzureQueryType2["CustomNamespacesQuery"] = "Azure Custom Namespaces";
  AzureQueryType2["GrafanaTemplateVariableFn"] = "Grafana Template Variable Function";
  AzureQueryType2["LocationsQuery"] = "Azure Regions";
  AzureQueryType2["LogAnalytics"] = "Azure Log Analytics";
  AzureQueryType2["MetricNamesQuery"] = "Azure Metric Names";
  AzureQueryType2["NamespacesQuery"] = "Azure Namespaces";
  AzureQueryType2["ResourceGroupsQuery"] = "Azure Resource Groups";
  AzureQueryType2["ResourceNamesQuery"] = "Azure Resource Names";
  AzureQueryType2["SubscriptionsQuery"] = "Azure Subscriptions";
  AzureQueryType2["TraceExemplar"] = "traceql";
  AzureQueryType2["WorkspacesQuery"] = "Azure Workspaces";
  return AzureQueryType2;
})(AzureQueryType || {});
const defaultAzureMetricQuery = {
  allowedTimeGrainsMs: [],
  dimensionFilters: [],
  resources: []
};
const defaultAzureLogsQuery = {
  resources: []
};
const defaultAzureTracesQuery = {
  filters: [],
  resources: [],
  traceTypes: []
};
const defaultAzureTracesFilter = {
  filters: []
};
var ResultFormat = /* @__PURE__ */ ((ResultFormat2) => {
  ResultFormat2["Logs"] = "logs";
  ResultFormat2["Table"] = "table";
  ResultFormat2["TimeSeries"] = "time_series";
  ResultFormat2["Trace"] = "trace";
  return ResultFormat2;
})(ResultFormat || {});
var LogsEditorMode = /* @__PURE__ */ ((LogsEditorMode2) => {
  LogsEditorMode2["Builder"] = "builder";
  LogsEditorMode2["Raw"] = "raw";
  return LogsEditorMode2;
})(LogsEditorMode || {});
var BuilderQueryEditorExpressionType = /* @__PURE__ */ ((BuilderQueryEditorExpressionType2) => {
  BuilderQueryEditorExpressionType2["And"] = "and";
  BuilderQueryEditorExpressionType2["Function_parameter"] = "function_parameter";
  BuilderQueryEditorExpressionType2["Group_by"] = "group_by";
  BuilderQueryEditorExpressionType2["Operator"] = "operator";
  BuilderQueryEditorExpressionType2["Or"] = "or";
  BuilderQueryEditorExpressionType2["Order_by"] = "order_by";
  BuilderQueryEditorExpressionType2["Property"] = "property";
  BuilderQueryEditorExpressionType2["Reduce"] = "reduce";
  return BuilderQueryEditorExpressionType2;
})(BuilderQueryEditorExpressionType || {});
var BuilderQueryEditorPropertyType = /* @__PURE__ */ ((BuilderQueryEditorPropertyType2) => {
  BuilderQueryEditorPropertyType2["Boolean"] = "boolean";
  BuilderQueryEditorPropertyType2["Datetime"] = "datetime";
  BuilderQueryEditorPropertyType2["Function"] = "function";
  BuilderQueryEditorPropertyType2["Interval"] = "interval";
  BuilderQueryEditorPropertyType2["Number"] = "number";
  BuilderQueryEditorPropertyType2["String"] = "string";
  BuilderQueryEditorPropertyType2["Time_span"] = "time_span";
  return BuilderQueryEditorPropertyType2;
})(BuilderQueryEditorPropertyType || {});
var BuilderQueryEditorOrderByOptions = /* @__PURE__ */ ((BuilderQueryEditorOrderByOptions2) => {
  BuilderQueryEditorOrderByOptions2["Asc"] = "asc";
  BuilderQueryEditorOrderByOptions2["Desc"] = "desc";
  return BuilderQueryEditorOrderByOptions2;
})(BuilderQueryEditorOrderByOptions || {});
var BuilderQueryEditorReduceParameterTypes = /* @__PURE__ */ ((BuilderQueryEditorReduceParameterTypes2) => {
  BuilderQueryEditorReduceParameterTypes2["Generic"] = "generic";
  BuilderQueryEditorReduceParameterTypes2["Numeric"] = "numeric";
  return BuilderQueryEditorReduceParameterTypes2;
})(BuilderQueryEditorReduceParameterTypes || {});
const defaultBuilderQueryEditorColumnsExpression = {
  columns: []
};
const defaultBuilderQueryEditorWhereExpression = {
  expressions: []
};
const defaultBuilderQueryEditorWhereExpressionArray = {
  expressions: []
};
const defaultBuilderQueryEditorReduceExpression = {
  parameters: []
};
const defaultBuilderQueryEditorReduceExpressionArray = {
  expressions: []
};
const defaultBuilderQueryEditorGroupByExpressionArray = {
  expressions: []
};
const defaultBuilderQueryEditorOrderByExpressionArray = {
  expressions: []
};
var ARGScope = /* @__PURE__ */ ((ARGScope2) => {
  ARGScope2["Directory"] = "directory";
  ARGScope2["Subscription"] = "subscription";
  return ARGScope2;
})(ARGScope || {});
const defaultAzureMetricDimension = {
  filters: []
};

export { ARGScope, AzureQueryType, BuilderQueryEditorExpressionType, BuilderQueryEditorOrderByOptions, BuilderQueryEditorPropertyType, BuilderQueryEditorReduceParameterTypes, LogsEditorMode, ResultFormat, defaultAzureLogsQuery, defaultAzureMetricDimension, defaultAzureMetricQuery, defaultAzureMonitorQuery, defaultAzureTracesFilter, defaultAzureTracesQuery, defaultBuilderQueryEditorColumnsExpression, defaultBuilderQueryEditorGroupByExpressionArray, defaultBuilderQueryEditorOrderByExpressionArray, defaultBuilderQueryEditorReduceExpression, defaultBuilderQueryEditorReduceExpressionArray, defaultBuilderQueryEditorWhereExpression, defaultBuilderQueryEditorWhereExpressionArray, pluginVersion };
