// @grafana/runtime/dist/cjs/services/pluginMeta/apps.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var config = require('../../config.cjs');
var index = require('../../internal/openFeature/index.cjs');
var mappers = require('./mappers/mappers.cjs');
var plugins = require('./plugins.cjs');

"use strict";
let apps = {};
function initialized() {
  return Boolean(Object.keys(apps).length);
}
async function initAppPluginMetas() {
  if (!index.getFeatureFlagClient().getBooleanValue("useMTPlugins", false)) {
    apps = config.config.apps;
    return;
  }
  const metas = await plugins.initPluginMetas();
  const mapper = mappers.getAppPluginMapper();
  apps = mapper(metas);
}
async function getAppPluginMetas() {
  if (!initialized()) {
    await initAppPluginMetas();
  }
  return Object.values(structuredClone(apps));
}
async function getAppPluginMeta(pluginId) {
  if (!initialized()) {
    await initAppPluginMetas();
  }
  const app = apps[pluginId];
  return app ? structuredClone(app) : null;
}
async function isAppPluginInstalled(pluginId) {
  const app = await getAppPluginMeta(pluginId);
  return Boolean(app);
}
async function getAppPluginVersion(pluginId) {
  var _a;
  const app = await getAppPluginMeta(pluginId);
  return (_a = app == null ? void 0 : app.version) != null ? _a : null;
}
function setAppPluginMetas(override) {
  if (process.env.NODE_ENV !== "test") {
    throw new Error("setAppPluginMetas() function can only be called from tests.");
  }
  apps = structuredClone(override);
}

exports.getAppPluginMeta = getAppPluginMeta;
exports.getAppPluginMetas = getAppPluginMetas;
exports.getAppPluginVersion = getAppPluginVersion;
exports.isAppPluginInstalled = isAppPluginInstalled;
exports.setAppPluginMetas = setAppPluginMetas;
