// @grafana/runtime/dist/cjs/services/pluginMeta/hooks.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var reactUse = require('react-use');
var apps = require('./apps.cjs');
var panels = require('./panels.cjs');

"use strict";
function useAppPluginMetas() {
  const { loading, error, value } = reactUse.useAsync(async () => apps.getAppPluginMetas());
  return { loading, error, value };
}
function useAppPluginMeta(pluginId) {
  const { loading, error, value } = reactUse.useAsync(async () => apps.getAppPluginMeta(pluginId));
  return { loading, error, value };
}
function usePanelPluginMetas() {
  const { loading, error, value } = reactUse.useAsync(async () => panels.getPanelPluginMetas());
  return { loading, error, value };
}
function usePanelPluginMeta(pluginId) {
  const { loading, error, value } = reactUse.useAsync(async () => panels.getPanelPluginMeta(pluginId));
  return { loading, error, value };
}
function useAppPluginInstalled(pluginId) {
  const { loading, error, value } = reactUse.useAsync(async () => apps.isAppPluginInstalled(pluginId));
  return { loading, error, value };
}
function useAppPluginVersion(pluginId) {
  const { loading, error, value } = reactUse.useAsync(async () => apps.getAppPluginVersion(pluginId));
  return { loading, error, value };
}
function usePanelPluginInstalled(pluginId) {
  const { loading, error, value } = reactUse.useAsync(async () => panels.isPanelPluginInstalled(pluginId));
  return { loading, error, value };
}
function usePanelPluginVersion(pluginId) {
  const { loading, error, value } = reactUse.useAsync(async () => panels.getPanelPluginVersion(pluginId));
  return { loading, error, value };
}
function useListedPanelPluginIds() {
  const { loading, error, value } = reactUse.useAsync(async () => panels.getListedPanelPluginIds());
  return { loading, error, value };
}

exports.useAppPluginInstalled = useAppPluginInstalled;
exports.useAppPluginMeta = useAppPluginMeta;
exports.useAppPluginMetas = useAppPluginMetas;
exports.useAppPluginVersion = useAppPluginVersion;
exports.useListedPanelPluginIds = useListedPanelPluginIds;
exports.usePanelPluginInstalled = usePanelPluginInstalled;
exports.usePanelPluginMeta = usePanelPluginMeta;
exports.usePanelPluginMetas = usePanelPluginMetas;
exports.usePanelPluginVersion = usePanelPluginVersion;
