// @grafana/runtime/dist/cjs/services/pluginMeta/panels.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var config = require('../../config.cjs');
var index = require('../../internal/openFeature/index.cjs');
var mappers = require('./mappers/mappers.cjs');
var plugins = require('./plugins.cjs');

"use strict";
let panels = {};
function initialized() {
  return Boolean(Object.keys(panels).length);
}
async function initPanelPluginMetas() {
  if (!index.getFeatureFlagClient().getBooleanValue("useMTPlugins", false)) {
    panels = config.config.panels;
    return;
  }
  const metas = await plugins.initPluginMetas();
  const mapper = mappers.getPanelPluginMapper();
  panels = mapper(metas);
}
async function getPanelPluginMetas() {
  if (!initialized()) {
    await initPanelPluginMetas();
  }
  return Object.values(structuredClone(panels));
}
async function getPanelPluginMeta(pluginId) {
  if (!initialized()) {
    await initPanelPluginMetas();
  }
  const panel = panels[pluginId];
  return panel ? structuredClone(panel) : null;
}
async function isPanelPluginInstalled(pluginId) {
  const panel = await getPanelPluginMeta(pluginId);
  return Boolean(panel);
}
async function getPanelPluginVersion(pluginId) {
  var _a;
  const panel = await getPanelPluginMeta(pluginId);
  return (_a = panel == null ? void 0 : panel.info.version) != null ? _a : null;
}
async function getListedPanelPluginIds() {
  const panels2 = await getPanelPluginMetas();
  return panels2.filter((p) => p.hideFromList === false).map((p) => p.id);
}
function setPanelPluginMetas(override) {
  if (process.env.NODE_ENV !== "test") {
    throw new Error("setPanelPluginMetas() function can only be called from tests.");
  }
  panels = structuredClone(override);
}

exports.getListedPanelPluginIds = getListedPanelPluginIds;
exports.getPanelPluginMeta = getPanelPluginMeta;
exports.getPanelPluginMetas = getPanelPluginMetas;
exports.getPanelPluginVersion = getPanelPluginVersion;
exports.isPanelPluginInstalled = isPanelPluginInstalled;
exports.setPanelPluginMetas = setPanelPluginMetas;
