// @grafana/ui/dist/cjs/components/Sidebar/SidebarPaneHeader.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var e2eSelectors = require('@grafana/e2e-selectors');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var IconButton = require('../IconButton/IconButton.cjs');
var Text = require('../Text/Text.cjs');
var useSidebar = require('./useSidebar.cjs');

"use strict";
function SidebarPaneHeader({ children, title }) {
  const styles = ThemeContext.useStyles2(getStyles);
  const context = React.useContext(useSidebar.SidebarContext);
  if (!context) {
    throw new Error("SidebarPaneHeader must be used within a Sidebar");
  }
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: styles.wrapper, children: [
    context.onClosePane && /* @__PURE__ */ jsxRuntime.jsx(
      IconButton.IconButton,
      {
        variant: "secondary",
        size: "lg",
        name: "times",
        onClick: context.onClosePane,
        "aria-label": i18n.t("grafana-ui.sidebar.close", "Close"),
        tooltip: i18n.t("grafana-ui.sidebar.close", "Close"),
        "data-testid": e2eSelectors.selectors.components.Sidebar.closePane
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsx(Text.Text, { weight: "medium", variant: "h6", truncate: true, "data-testid": "sidebar-pane-header-title", children: title }),
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles.flexGrow }),
    children
  ] });
}
const getStyles = (theme) => {
  return {
    wrapper: css.css({
      display: "flex",
      alignItems: "center",
      padding: theme.spacing(1.5),
      height: theme.spacing(6),
      gap: theme.spacing(1),
      borderBottom: `1px solid ${theme.colors.border.weak}`
    }),
    flexGrow: css.css({
      flexGrow: 1
    })
  };
};

exports.SidebarPaneHeader = SidebarPaneHeader;
exports.getStyles = getStyles;
