// @grafana/ui/dist/cjs/components/Sidebar/useSidebar.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var lodash = require('lodash');
var React = require('react');
var data = require('@grafana/data');
var ThemeContext = require('../../themes/ThemeContext.cjs');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefaultCompat(React);

"use strict";
const SidebarContext = React__default.default.createContext(void 0);
const SIDE_BAR_WIDTH_ICON_ONLY = 5;
const SIDE_BAR_WIDTH_WITH_TEXT = 8;
function useSidebar({
  hasOpenPane,
  position = "right",
  tabsMode,
  defaultToCompact = true,
  defaultToDocked = false,
  bottomMargin = 2,
  edgeMargin = 2,
  contentMargin = 2,
  persistanceKey,
  onClosePane,
  isHidden = false
}) {
  const theme = ThemeContext.useTheme2();
  const [isDocked, setIsDocked] = useSidebarSavedState(persistanceKey, "docked", defaultToDocked);
  const [compact, setCompact] = useSidebarSavedState(persistanceKey, "compact", defaultToCompact);
  const [paneWidth, setPaneWidth] = useSidebarSavedState(persistanceKey, "size", 280);
  const [_, setCompactDrag] = React__default.default.useState(0);
  const onToggleDock = React.useCallback(() => {
    setIsDocked((prev) => {
      return !prev;
    });
  }, [setIsDocked]);
  const prop = position === "right" ? "paddingRight" : "paddingLeft";
  const toolbarWidth = ((compact ? SIDE_BAR_WIDTH_ICON_ONLY : SIDE_BAR_WIDTH_WITH_TEXT) + edgeMargin + contentMargin) * theme.spacing.gridSize;
  const outerWrapperProps = isHidden ? {} : {
    style: {
      [prop]: isDocked && hasOpenPane ? paneWidth + toolbarWidth : toolbarWidth
    }
  };
  const onResize = React.useCallback(
    (diff) => {
      setPaneWidth((prevWidth) => {
        if (!hasOpenPane) {
          setCompactDrag((prevDrag) => {
            const newDrag = prevDrag + diff;
            if (newDrag < -20 && !compact) {
              setCompact(() => true);
              return 0;
            } else if (newDrag > 20 && compact) {
              setCompact(() => false);
              return 0;
            }
            return newDrag;
          });
          return prevWidth;
        }
        return lodash.clamp(prevWidth + diff, 100, 500);
      });
    },
    [hasOpenPane, setCompact, setPaneWidth, compact]
  );
  return {
    isDocked,
    onToggleDock,
    onResize,
    outerWrapperProps,
    position,
    compact,
    hasOpenPane,
    tabsMode,
    paneWidth,
    edgeMargin,
    bottomMargin,
    contentMargin,
    isHidden,
    onClosePane
  };
}
function useSidebarSavedState(persistanceKey, subKey, defaultValue) {
  const [state, setState] = React__default.default.useState(() => {
    if (!persistanceKey) {
      return defaultValue;
    }
    if (typeof defaultValue === "boolean") {
      return data.store.getBool(`grafana.ui.sidebar.${persistanceKey}.${subKey}`, defaultValue);
    }
    if (typeof defaultValue === "number") {
      const value = Number.parseInt(data.store.get(`grafana.ui.sidebar.${persistanceKey}.${subKey}`), 10);
      if (Number.isNaN(value)) {
        return defaultValue;
      }
      return value;
    }
    return defaultValue;
  });
  const setPersisted = React.useCallback(
    (cb) => {
      setState((prevState) => {
        const newState = cb(prevState);
        if (!persistanceKey) {
          return newState;
        }
        if (persistanceKey) {
          data.store.set(`grafana.ui.sidebar.${persistanceKey}.${subKey}`, String(newState));
        }
        return newState;
      });
    },
    [persistanceKey, subKey]
  );
  return [state, setPersisted];
}

exports.SIDE_BAR_WIDTH_ICON_ONLY = SIDE_BAR_WIDTH_ICON_ONLY;
exports.SIDE_BAR_WIDTH_WITH_TEXT = SIDE_BAR_WIDTH_WITH_TEXT;
exports.SidebarContext = SidebarContext;
exports.useSidebar = useSidebar;
