// @grafana/ui/dist/esm/components/Sidebar/SidebarPaneHeader.mjs
import { jsxs, jsx } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import { useContext } from 'react';
import { selectors } from '@grafana/e2e-selectors';
import { t } from '@grafana/i18n';
import { useStyles2 } from '../../themes/ThemeContext.mjs';
import { IconButton } from '../IconButton/IconButton.mjs';
import { Text } from '../Text/Text.mjs';
import { SidebarContext } from './useSidebar.mjs';

"use strict";
function SidebarPaneHeader({ children, title }) {
  const styles = useStyles2(getStyles);
  const context = useContext(SidebarContext);
  if (!context) {
    throw new Error("SidebarPaneHeader must be used within a Sidebar");
  }
  return /* @__PURE__ */ jsxs("div", { className: styles.wrapper, children: [
    context.onClosePane && /* @__PURE__ */ jsx(
      IconButton,
      {
        variant: "secondary",
        size: "lg",
        name: "times",
        onClick: context.onClosePane,
        "aria-label": t("grafana-ui.sidebar.close", "Close"),
        tooltip: t("grafana-ui.sidebar.close", "Close"),
        "data-testid": selectors.components.Sidebar.closePane
      }
    ),
    /* @__PURE__ */ jsx(Text, { weight: "medium", variant: "h6", truncate: true, "data-testid": "sidebar-pane-header-title", children: title }),
    /* @__PURE__ */ jsx("div", { className: styles.flexGrow }),
    children
  ] });
}
const getStyles = (theme) => {
  return {
    wrapper: css({
      display: "flex",
      alignItems: "center",
      padding: theme.spacing(1.5),
      height: theme.spacing(6),
      gap: theme.spacing(1),
      borderBottom: `1px solid ${theme.colors.border.weak}`
    }),
    flexGrow: css({
      flexGrow: 1
    })
  };
};

export { SidebarPaneHeader, getStyles };
