// @grafana/ui/dist/esm/components/Sidebar/Sidebar.mjs
import { jsx, jsxs } from 'react/jsx-runtime';
import { cx, css } from '@emotion/css';
import { useContext } from 'react';
import { selectors } from '@grafana/e2e-selectors';
import { t } from '@grafana/i18n';
import { useStyles2, useTheme2 } from '../../themes/ThemeContext.mjs';
import { getPortalContainer } from '../Portal/Portal.mjs';
import { SidebarButton } from './SidebarButton.mjs';
import { SidebarPaneHeader } from './SidebarPaneHeader.mjs';
import { SidebarResizer } from './SidebarResizer.mjs';
import { SidebarContext, SIDE_BAR_WIDTH_ICON_ONLY, SIDE_BAR_WIDTH_WITH_TEXT } from './useSidebar.mjs';
export { useSidebar } from './useSidebar.mjs';
import { useCustomClickAway } from './useSidebarClickAway.mjs';

"use strict";
function SidebarComp({ children, contextValue }) {
  const styles = useStyles2(getStyles);
  const theme = useTheme2();
  const { isDocked, position, tabsMode, hasOpenPane, edgeMargin, bottomMargin } = contextValue;
  const className = cx({
    [styles.container]: true,
    [styles.undockedPaneOpen]: hasOpenPane && !isDocked,
    [styles.containerLeft]: position === "left",
    [styles.containerTabsMode]: tabsMode,
    [styles.containerHidden]: !!contextValue.isHidden
  });
  const style = { [position]: theme.spacing(edgeMargin), bottom: theme.spacing(bottomMargin) };
  const ref = useCustomClickAway((evt) => {
    var _a;
    const portalContainer = getPortalContainer();
    if (evt.target instanceof Node && portalContainer && portalContainer.contains(evt.target)) {
      return;
    }
    if (!isDocked && hasOpenPane) {
      (_a = contextValue.onClosePane) == null ? void 0 : _a.call(contextValue);
    }
  });
  return /* @__PURE__ */ jsx(SidebarContext.Provider, { value: contextValue, children: /* @__PURE__ */ jsxs(
    "div",
    {
      ref,
      className,
      style,
      id: "sidebar-container",
      "data-testid": selectors.components.Sidebar.container,
      "aria-hidden": contextValue.isHidden,
      children: [
        !tabsMode && /* @__PURE__ */ jsx(SidebarResizer, {}),
        children
      ]
    }
  ) });
}
function SiderbarToolbar({ children }) {
  const styles = useStyles2(getStyles);
  const context = useContext(SidebarContext);
  if (!context) {
    throw new Error("Sidebar.Toolbar must be used within a Sidebar component");
  }
  return /* @__PURE__ */ jsxs("div", { className: cx(styles.toolbar, context.compact && styles.toolbarIconsOnly), children: [
    children,
    /* @__PURE__ */ jsx("div", { className: styles.flexGrow }),
    context.hasOpenPane && /* @__PURE__ */ jsx(
      SidebarButton,
      {
        icon: "web-section-alt",
        onClick: context.onToggleDock,
        title: context.isDocked ? t("grafana-ui.sidebar.undock", "Undock") : t("grafana-ui.sidebar.dock", "Dock"),
        "data-testid": selectors.components.Sidebar.dockToggle
      }
    )
  ] });
}
function SidebarDivider() {
  const styles = useStyles2(getStyles);
  return /* @__PURE__ */ jsx("div", { className: styles.divider });
}
function SidebarOpenPane({ children }) {
  const styles = useStyles2(getStyles);
  const context = useContext(SidebarContext);
  if (!context) {
    throw new Error("Sidebar.OpenPane must be used within a Sidebar component");
  }
  const className = cx(styles.openPane, context.position === "right" ? styles.openPaneRight : styles.openPaneLeft);
  return /* @__PURE__ */ jsx("div", { className, style: { width: context.paneWidth }, children });
}
const getStyles = (theme) => {
  return {
    container: css({
      display: "flex",
      position: "absolute",
      flexDirection: "row",
      flex: "1 1 0",
      border: `1px solid ${theme.colors.border.weak}`,
      background: theme.colors.background.primary,
      borderRadius: theme.shape.radius.default,
      zIndex: theme.zIndex.navbarFixed,
      bottom: 0,
      top: 0,
      right: 0,
      width: "calc-size(auto, size)",
      [theme.transitions.handleMotion("no-preference")]: {
        transition: theme.transitions.create("width", {
          duration: theme.transitions.duration.standard
        })
      }
    }),
    containerHidden: css({
      width: 0,
      border: 0,
      overflow: "hidden"
    }),
    containerTabsMode: css({
      position: "relative"
    }),
    containerLeft: css({
      right: "unset",
      flexDirection: "row-reverse",
      left: 0,
      borderRadius: theme.shape.radius.default
    }),
    undockedPaneOpen: css({
      boxShadow: theme.shadows.z3
    }),
    toolbar: css({
      display: "flex",
      flexDirection: "column",
      alignItems: "center",
      padding: theme.spacing(1, 0),
      flexGrow: 0,
      gap: theme.spacing(2),
      overflow: "hidden",
      width: theme.spacing(SIDE_BAR_WIDTH_WITH_TEXT)
    }),
    toolbarIconsOnly: css({
      width: theme.spacing(SIDE_BAR_WIDTH_ICON_ONLY)
    }),
    divider: css({
      height: "1px",
      background: theme.colors.border.weak,
      width: "70%"
    }),
    flexGrow: css({
      flexGrow: 1
    }),
    openPane: css({
      width: "280px",
      flexGrow: 1,
      paddingBottom: theme.spacing(2)
    }),
    openPaneRight: css({
      borderRight: `1px solid ${theme.colors.border.weak}`
    }),
    openPaneLeft: css({
      borderLeft: `1px solid ${theme.colors.border.weak}`
    })
  };
};
const Sidebar = Object.assign(SidebarComp, {
  Toolbar: SiderbarToolbar,
  Button: SidebarButton,
  OpenPane: SidebarOpenPane,
  Divider: SidebarDivider,
  PaneHeader: SidebarPaneHeader
});

export { Sidebar, SidebarComp, SidebarDivider, SidebarOpenPane, SiderbarToolbar, getStyles };
