// @grafana/ui/dist/esm/components/Sidebar/useSidebar.mjs
import { clamp } from 'lodash';
import React__default, { useCallback } from 'react';
import { store } from '@grafana/data';
import { useTheme2 } from '../../themes/ThemeContext.mjs';

"use strict";
const SidebarContext = React__default.createContext(void 0);
const SIDE_BAR_WIDTH_ICON_ONLY = 5;
const SIDE_BAR_WIDTH_WITH_TEXT = 8;
function useSidebar({
  hasOpenPane,
  position = "right",
  tabsMode,
  defaultToCompact = true,
  defaultToDocked = false,
  bottomMargin = 2,
  edgeMargin = 2,
  contentMargin = 2,
  persistanceKey,
  onClosePane,
  isHidden = false
}) {
  const theme = useTheme2();
  const [isDocked, setIsDocked] = useSidebarSavedState(persistanceKey, "docked", defaultToDocked);
  const [compact, setCompact] = useSidebarSavedState(persistanceKey, "compact", defaultToCompact);
  const [paneWidth, setPaneWidth] = useSidebarSavedState(persistanceKey, "size", 280);
  const [_, setCompactDrag] = React__default.useState(0);
  const onToggleDock = useCallback(() => {
    setIsDocked((prev) => {
      return !prev;
    });
  }, [setIsDocked]);
  const prop = position === "right" ? "paddingRight" : "paddingLeft";
  const toolbarWidth = ((compact ? SIDE_BAR_WIDTH_ICON_ONLY : SIDE_BAR_WIDTH_WITH_TEXT) + edgeMargin + contentMargin) * theme.spacing.gridSize;
  const outerWrapperProps = isHidden ? {} : {
    style: {
      [prop]: isDocked && hasOpenPane ? paneWidth + toolbarWidth : toolbarWidth
    }
  };
  const onResize = useCallback(
    (diff) => {
      setPaneWidth((prevWidth) => {
        if (!hasOpenPane) {
          setCompactDrag((prevDrag) => {
            const newDrag = prevDrag + diff;
            if (newDrag < -20 && !compact) {
              setCompact(() => true);
              return 0;
            } else if (newDrag > 20 && compact) {
              setCompact(() => false);
              return 0;
            }
            return newDrag;
          });
          return prevWidth;
        }
        return clamp(prevWidth + diff, 100, 500);
      });
    },
    [hasOpenPane, setCompact, setPaneWidth, compact]
  );
  return {
    isDocked,
    onToggleDock,
    onResize,
    outerWrapperProps,
    position,
    compact,
    hasOpenPane,
    tabsMode,
    paneWidth,
    edgeMargin,
    bottomMargin,
    contentMargin,
    isHidden,
    onClosePane
  };
}
function useSidebarSavedState(persistanceKey, subKey, defaultValue) {
  const [state, setState] = React__default.useState(() => {
    if (!persistanceKey) {
      return defaultValue;
    }
    if (typeof defaultValue === "boolean") {
      return store.getBool(`grafana.ui.sidebar.${persistanceKey}.${subKey}`, defaultValue);
    }
    if (typeof defaultValue === "number") {
      const value = Number.parseInt(store.get(`grafana.ui.sidebar.${persistanceKey}.${subKey}`), 10);
      if (Number.isNaN(value)) {
        return defaultValue;
      }
      return value;
    }
    return defaultValue;
  });
  const setPersisted = useCallback(
    (cb) => {
      setState((prevState) => {
        const newState = cb(prevState);
        if (!persistanceKey) {
          return newState;
        }
        if (persistanceKey) {
          store.set(`grafana.ui.sidebar.${persistanceKey}.${subKey}`, String(newState));
        }
        return newState;
      });
    },
    [persistanceKey, subKey]
  );
  return [state, setPersisted];
}

export { SIDE_BAR_WIDTH_ICON_ONLY, SIDE_BAR_WIDTH_WITH_TEXT, SidebarContext, useSidebar };
