// @grafana/ui/dist/cjs/components/Table/TableNG/components/HeaderCell.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var ThemeContext = require('../../../../themes/ThemeContext.cjs');
var icon = require('../../../../types/icon.cjs');
var Icon = require('../../../Icon/Icon.cjs');
var Stack = require('../../../Layout/Stack/Stack.cjs');
var Filter = require('../Filter/Filter.cjs');
var utils = require('../utils.cjs');

"use strict";
const HeaderCell = ({
  column,
  crossFilterOrder,
  crossFilterRows,
  direction,
  disableKeyboardEvents,
  field,
  filter,
  rows,
  selectFirstCell,
  setFilter,
  showTypeIcons
}) => {
  var _a, _b, _c, _d;
  const ref = React.useRef(null);
  const headerCellWrap = (_b = (_a = field.config.custom) == null ? void 0 : _a.wrapHeaderText) != null ? _b : false;
  const styles = ThemeContext.useStyles2(getStyles, headerCellWrap);
  const displayName = utils.getDisplayName(field);
  const filterable = (_d = (_c = field.config.custom) == null ? void 0 : _c.filterable) != null ? _d : false;
  React.useEffect(() => {
    if (!filterable && filter[displayName]) {
      setFilter((filter2) => {
        const newFilter = { ...filter2 };
        delete newFilter[displayName];
        return newFilter;
      });
    }
  }, [filterable, displayName, filter, setFilter]);
  return /* @__PURE__ */ jsxRuntime.jsxs(
    Stack.Stack,
    {
      ref,
      direction: "row",
      gap: 0.5,
      alignItems: "center",
      onKeyDown: disableKeyboardEvents ? void 0 : (ev) => {
        var _a2, _b2;
        ev.stopPropagation();
        if (!(ev.key === "Tab" && !ev.shiftKey)) {
          return;
        }
        const tableTabbedElement = ev.target;
        if (!(tableTabbedElement instanceof HTMLElement)) {
          return;
        }
        const headerContent = ref.current;
        const headerCell = (_a2 = ref.current) == null ? void 0 : _a2.parentNode;
        const row = headerCell == null ? void 0 : headerCell.parentNode;
        const isLastElementInHeader = ((_b2 = headerContent == null ? void 0 : headerContent.lastElementChild) == null ? void 0 : _b2.contains(tableTabbedElement)) && headerCell === (row == null ? void 0 : row.lastElementChild);
        if (isLastElementInHeader) {
          selectFirstCell();
        }
      },
      children: [
        showTypeIcons && /* @__PURE__ */ jsxRuntime.jsx(Icon.Icon, { className: styles.headerCellIcon, name: icon.getFieldTypeIcon(field), title: field == null ? void 0 : field.type, size: "sm" }),
        /* @__PURE__ */ jsxRuntime.jsxs("button", { tabIndex: 0, className: styles.headerCellLabel, title: displayName, children: [
          displayName,
          direction && /* @__PURE__ */ jsxRuntime.jsx(Icon.Icon, { className: styles.headerCellIcon, size: "lg", name: direction === "ASC" ? "arrow-up" : "arrow-down" })
        ] }),
        filterable && /* @__PURE__ */ jsxRuntime.jsx(
          Filter.Filter,
          {
            name: column.key,
            rows,
            filter,
            setFilter,
            field,
            crossFilterOrder,
            crossFilterRows,
            iconClassName: styles.headerCellIcon
          }
        )
      ]
    }
  );
};
const getStyles = (theme, headerTextWrap) => ({
  headerCellLabel: css.css({
    all: "unset",
    cursor: "pointer",
    fontWeight: theme.typography.fontWeightMedium,
    color: theme.colors.text.secondary,
    overflow: "hidden",
    textOverflow: "ellipsis",
    whiteSpace: headerTextWrap ? "pre-line" : "nowrap",
    borderRadius: theme.spacing(0.25),
    lineHeight: "20px",
    "&:hover": {
      textDecoration: "underline"
    },
    "&::selection": {
      backgroundColor: "var(--rdg-background-color)",
      color: theme.colors.text.secondary
    }
  }),
  headerCellIcon: css.css({
    color: theme.colors.text.secondary
  })
});

exports.HeaderCell = HeaderCell;
