// @grafana/ui/dist/cjs/components/Table/TableNG/components/RowExpander.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var e2eSelectors = require('@grafana/e2e-selectors');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../../../themes/ThemeContext.cjs');
var Icon = require('../../../Icon/Icon.cjs');

"use strict";
function RowExpander({ onCellExpand, isExpanded, rowId }) {
  const styles = ThemeContext.useStyles2(getStyles);
  function handleKeyDown(e) {
    if (e.key === " " || e.key === "Enter") {
      e.preventDefault();
      onCellExpand(e);
    }
  }
  const label = isExpanded ? i18n.t("grafana-ui.row-expander-ng.aria-label-collapse", "Collapse row") : i18n.t("grafana-ui.row-expander.aria-label-expand", "Expand row");
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      role: "button",
      tabIndex: 0,
      className: styles.expanderCell,
      onClick: onCellExpand,
      onKeyDown: handleKeyDown,
      "aria-label": label,
      "data-testid": e2eSelectors.selectors.components.Panels.Visualization.TableNG.RowExpander,
      "aria-expanded": isExpanded,
      "aria-controls": rowId,
      children: /* @__PURE__ */ jsxRuntime.jsx(Icon.Icon, { name: isExpanded ? "angle-down" : "angle-right", size: "lg", "aria-hidden": "true" })
    }
  );
}
const getStyles = (_theme) => ({
  expanderCell: css.css({
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    cursor: "pointer"
  })
});

exports.RowExpander = RowExpander;
