// @grafana/ui/dist/cjs/components/Table/TableNG/components/SummaryCell.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var clsx = require('clsx');
var React = require('react');
var data = require('@grafana/data');
var e2eSelectors = require('@grafana/e2e-selectors');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../../../themes/ThemeContext.cjs');
var hooks = require('../hooks.cjs');
var styles = require('../styles.cjs');
var utils = require('../utils.cjs');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var clsx__default = /*#__PURE__*/_interopDefaultCompat(clsx);

"use strict";
const getReducerName = (reducerId) => {
  var _a;
  if (reducerId === data.ReducerID.countAll) {
    return i18n.t("grafana-ui.table.footer.reducer.count", "Count");
  }
  return ((_a = data.fieldReducers.get(reducerId)) == null ? void 0 : _a.name) || reducerId;
};
const SummaryCell = ({
  rows,
  footers,
  field,
  colIdx,
  hideLabel = false,
  rowLabel = false,
  textAlign
}) => {
  const styles$1 = ThemeContext.useStyles2(getStyles, textAlign, hideLabel);
  const theme = ThemeContext.useTheme2();
  const defaultFooterCellStyles = styles.getDefaultCellStyles(theme, {
    textAlign: "left",
    // alignment is set in footerItem
    shouldOverflow: true,
    textWrap: false
  });
  const displayName = utils.getDisplayName(field);
  const reducerResultsEntries = hooks.useReducerEntries(field, rows, displayName, colIdx);
  const cellClass = clsx__default.default(styles$1.footerCell, defaultFooterCellStyles);
  const firstFooterReducers = React.useMemo(() => {
    var _a, _b;
    for (const footer of footers) {
      if ((_b = (_a = footer == null ? void 0 : footer.reducers) == null ? void 0 : _a.length) != null ? _b : 0 > 0) {
        return footer.reducers;
      }
    }
    return;
  }, [footers]);
  const renderRowLabel = rowLabel && reducerResultsEntries.length === 0 && Boolean(firstFooterReducers);
  const SummaryCellItem = ({ children }) => /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles$1.footerItem, children });
  const SummaryCellLabel = ({ children }) => /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      "data-testid": e2eSelectors.selectors.components.Panels.Visualization.TableNG.Footer.ReducerLabel,
      className: styles$1.footerItemLabel,
      children
    }
  );
  const SummaryCellValue = ({ children }) => /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      "data-testid": e2eSelectors.selectors.components.Panels.Visualization.TableNG.Footer.Value,
      className: styles$1.footerItemValue,
      children
    }
  );
  return /* @__PURE__ */ jsxRuntime.jsxs(
    "div",
    {
      className: cellClass,
      "data-testid": reducerResultsEntries.length === 0 && !renderRowLabel ? "summary-cell-empty" : void 0,
      children: [
        reducerResultsEntries.map(([reducerId, reducerResult]) => {
          if (reducerResult === null) {
            return /* @__PURE__ */ jsxRuntime.jsx(SummaryCellItem, { children: rowLabel ? /* @__PURE__ */ jsxRuntime.jsx(SummaryCellLabel, { children: getReducerName(reducerId) }) : /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, { children: "\xA0" }) }, reducerId);
          }
          return /* @__PURE__ */ jsxRuntime.jsxs(SummaryCellItem, { children: [
            !hideLabel && /* @__PURE__ */ jsxRuntime.jsx(SummaryCellLabel, { children: getReducerName(reducerId) }),
            /* @__PURE__ */ jsxRuntime.jsx(SummaryCellValue, { children: reducerResult })
          ] }, reducerId);
        }),
        renderRowLabel && firstFooterReducers.map((reducerId) => /* @__PURE__ */ jsxRuntime.jsx(SummaryCellItem, { children: /* @__PURE__ */ jsxRuntime.jsx(SummaryCellLabel, { children: getReducerName(reducerId) }) }, reducerId))
      ]
    }
  );
};
const getStyles = (theme, textAlign, hideLabel) => ({
  footerCell: css.css({
    flexDirection: "column",
    minHeight: "100%",
    width: "100%"
  }),
  footerItem: css.css({
    display: "flex",
    flexDirection: "row",
    justifyContent: hideLabel ? utils.getJustifyContent(textAlign) : "space-between",
    alignItems: "flex-start",
    width: "100%",
    gap: theme.spacing(0.5)
  }),
  footerItemLabel: css.css({
    flexShrink: 0,
    overflow: "hidden",
    textOverflow: "ellipsis",
    whiteSpace: "nowrap",
    color: theme.colors.text.secondary,
    fontSize: theme.typography.bodySmall.fontSize,
    fontWeight: theme.typography.fontWeightLight,
    textTransform: "uppercase",
    lineHeight: "22px"
  }),
  footerItemValue: css.css({
    overflow: "hidden",
    textOverflow: "ellipsis",
    fontWeight: theme.typography.fontWeightMedium
  })
});

exports.SummaryCell = SummaryCell;
exports.getStyles = getStyles;
