// @grafana/ui/dist/cjs/components/DataLinks/DataLinksInlineEditor/DataLinkEditorModalContent.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var i18n = require('@grafana/i18n');
var Button = require('../../Button/Button.cjs');
var Modal = require('../../Modal/Modal.cjs');
var DataLinkEditor = require('../DataLinkEditor.cjs');

"use strict";
const DataLinkEditorModalContent = ({
  link,
  index,
  onSave,
  onCancel,
  getSuggestions,
  showOneClick
}) => {
  const [dirtyLink, setDirtyLink] = React.useState(link);
  return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      DataLinkEditor.DataLinkEditor,
      {
        value: dirtyLink,
        index,
        isLast: false,
        onChange: (index2, link2) => {
          setDirtyLink(link2);
        },
        suggestions: getSuggestions(),
        showOneClick
      }
    ),
    /* @__PURE__ */ jsxRuntime.jsxs(Modal.Modal.ButtonRow, { children: [
      /* @__PURE__ */ jsxRuntime.jsx(Button.Button, { variant: "secondary", onClick: () => onCancel(index), fill: "outline", children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.data-link-editor-modal.cancel", children: "Cancel" }) }),
      /* @__PURE__ */ jsxRuntime.jsx(
        Button.Button,
        {
          onClick: () => {
            onSave(index, dirtyLink);
          },
          disabled: dirtyLink.title.trim() === "" || dirtyLink.url.trim() === "",
          children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.data-link-editor-modal.save", children: "Save" })
        }
      )
    ] })
  ] });
};

exports.DataLinkEditorModalContent = DataLinkEditorModalContent;
