// @grafana/ui/dist/cjs/components/Table/TableNG/Filter/Filter.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var clsx = require('clsx');
var React = require('react');
var e2eSelectors = require('@grafana/e2e-selectors');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../../../themes/ThemeContext.cjs');
var Icon = require('../../../Icon/Icon.cjs');
var Popover = require('../../../Tooltip/Popover.cjs');
var types = require('../types.cjs');
var utils$1 = require('../utils.cjs');
var FilterPopup = require('./FilterPopup.cjs');
var utils = require('./utils.cjs');

"use strict";
const Filter = React.memo(
  ({ name, rows, filter, setFilter, field, crossFilterOrder, crossFilterRows, iconClassName }) => {
    var _a, _b, _c;
    const filterValue = (_a = filter[name]) == null ? void 0 : _a.filtered;
    const filterIndex = crossFilterOrder.indexOf(name);
    let filteredRows;
    if (filterIndex > 0) {
      const previousFilterName = crossFilterOrder[filterIndex - 1];
      filteredRows = crossFilterRows[previousFilterName];
    } else if (filterIndex === -1 && crossFilterOrder.length > 0) {
      const previousFilterName = crossFilterOrder[crossFilterOrder.length - 1];
      filteredRows = crossFilterRows[previousFilterName];
    } else {
      filteredRows = rows;
    }
    const ref = React.useRef(null);
    const [isPopoverVisible, setPopoverVisible] = React.useState(false);
    const styles = ThemeContext.useStyles2(getStyles);
    const filterEnabled = Boolean(filterValue);
    const [searchFilter, setSearchFilter] = React.useState(((_b = filter[name]) == null ? void 0 : _b.searchFilter) || "");
    const [operator, setOperator] = React.useState(
      ((_c = filter[name]) == null ? void 0 : _c.operator) || utils.operatorSelectableValues()[types.FilterOperator.CONTAINS]
    );
    return /* @__PURE__ */ jsxRuntime.jsxs(
      "button",
      {
        className: styles.headerFilter,
        ref,
        type: "button",
        "aria-haspopup": "dialog",
        "aria-pressed": isPopoverVisible,
        "aria-label": i18n.t("grafana-ui.table.filter.button", `Filter {{name}}`, {
          name: field ? utils$1.getDisplayName(field) : ""
        }),
        "data-testid": e2eSelectors.selectors.components.Panels.Visualization.TableNG.Filters.HeaderButton,
        tabIndex: 0,
        onKeyDown: (ev) => {
          if (ev.target === ref.current && (ev.key === "Enter" || ev.key === " ")) {
            setPopoverVisible(true);
            ev.stopPropagation();
            ev.preventDefault();
          }
        },
        onClick: (ev) => {
          ev.stopPropagation();
          if (!isPopoverVisible) {
            setPopoverVisible(true);
          }
        },
        children: [
          /* @__PURE__ */ jsxRuntime.jsx(Icon.Icon, { name: "filter", className: clsx.clsx(iconClassName, filterEnabled ? styles.filterIconEnabled : void 0) }),
          isPopoverVisible && ref.current && /* @__PURE__ */ jsxRuntime.jsx(
            Popover.Popover,
            {
              content: /* @__PURE__ */ jsxRuntime.jsx(
                FilterPopup.FilterPopup,
                {
                  name,
                  rows: filteredRows,
                  filterValue,
                  setFilter,
                  field,
                  onClose: () => setPopoverVisible(false),
                  searchFilter,
                  setSearchFilter,
                  operator,
                  setOperator,
                  buttonElement: ref.current
                }
              ),
              placement: "bottom-start",
              referenceElement: ref.current,
              show: true
            }
          )
        ]
      }
    );
  }
);
Filter.displayName = "Filter";
const getStyles = (theme) => ({
  headerFilter: css.css({
    background: "transparent",
    border: "none",
    label: "headerFilter",
    padding: 0,
    alignSelf: "flex-end",
    borderRadius: theme.spacing(0.25)
  }),
  filterIconEnabled: css.css({
    label: "filterIconEnabled",
    color: theme.colors.primary.text
  })
});

exports.Filter = Filter;
