// @grafana/ui/dist/cjs/components/Table/TableNG/Filter/FilterPopup.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var e2eSelectors = require('@grafana/e2e-selectors');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../../../themes/ThemeContext.cjs');
var Button = require('../../../Button/Button.cjs');
var ButtonSelect = require('../../../Dropdown/ButtonSelect.cjs');
var FilterInput = require('../../../FilterInput/FilterInput.cjs');
var Label = require('../../../Forms/Label.cjs');
var Stack = require('../../../Layout/Stack/Stack.cjs');
var utils$1 = require('../utils.cjs');
var FilterList = require('./FilterList.cjs');
var utils = require('./utils.cjs');

"use strict";
const FilterPopup = React.memo(
  ({
    name,
    rows,
    filterValue,
    setFilter,
    onClose,
    field,
    searchFilter,
    setSearchFilter,
    operator,
    setOperator,
    buttonElement
  }) => {
    var _a;
    const uniqueValues = React.useMemo(() => utils.calculateUniqueFieldValues(rows, field), [rows, field]);
    const options = React.useMemo(() => utils.valuesToOptions(uniqueValues), [uniqueValues]);
    const filteredOptions = React.useMemo(() => utils.getFilteredOptions(options, filterValue), [options, filterValue]);
    const [values, setValues] = React.useState(filteredOptions);
    const [matchCase, setMatchCase] = React.useState(false);
    const containerRef = React.useRef(null);
    const inputRef = React.useRef(null);
    const operators = Object.values(utils.operatorSelectableValues());
    React.useEffect(() => {
      if (inputRef.current) {
        inputRef.current.focus();
      }
    }, []);
    React.useEffect(() => {
      function handleEscape(e) {
        if (e.key === "Escape") {
          onClose();
          buttonElement == null ? void 0 : buttonElement.focus();
        }
      }
      document.addEventListener("keyup", handleEscape);
      return () => {
        document.removeEventListener("keyup", handleEscape);
      };
    }, [onClose, buttonElement]);
    const onFilter = React.useCallback(() => {
      if (values.length !== 0) {
        const filteredSet = new Set(values.map((item) => item.value));
        setFilter((filter) => ({
          ...filter,
          [name]: { filtered: values, filteredSet, searchFilter, operator }
        }));
      } else {
        setFilter((filter) => {
          const newFilter = { ...filter };
          delete newFilter[name];
          return newFilter;
        });
      }
      onClose();
    }, [name, operator, searchFilter, setFilter, values, onClose]);
    const onClearFilter = React.useCallback(() => {
      setFilter((filter) => {
        const newFilter = { ...filter };
        delete newFilter[name];
        return newFilter;
      });
      onClose();
    }, [name, setFilter, onClose]);
    React.useEffect(() => {
      const onOutsideClick = (event) => {
        const domNode = containerRef.current;
        if (!domNode) {
          return;
        }
        if (event.target instanceof Node && !domNode.contains(event.target)) {
          console.log("closing from outside click");
          onClose();
        }
      };
      window.addEventListener("click", onOutsideClick);
      return () => {
        window.removeEventListener("click", onOutsideClick);
      };
    }, [onClose]);
    const filterInputPlaceholder = i18n.t("grafana-ui.table.filter-popup-input-placeholder", "Filter values");
    const clearFilterVisible = React.useMemo(() => filterValue !== void 0, [filterValue]);
    const styles = ThemeContext.useStyles2(getStyles);
    return /* @__PURE__ */ jsxRuntime.jsx(
      "div",
      {
        className: styles.filterContainer,
        "data-testid": e2eSelectors.selectors.components.Panels.Visualization.TableNG.Filters.Container,
        ref: containerRef,
        children: /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { direction: "column", children: [
          /* @__PURE__ */ jsxRuntime.jsx(Stack.Stack, { alignItems: "center", children: field && /* @__PURE__ */ jsxRuntime.jsx(Label.Label, { className: styles.label, children: utils$1.getDisplayName(field) }) }),
          /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { gap: 1, children: [
            /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles.inputContainer, children: /* @__PURE__ */ jsxRuntime.jsx(
              FilterInput.FilterInput,
              {
                ref: inputRef,
                placeholder: filterInputPlaceholder,
                title: filterInputPlaceholder,
                onChange: setSearchFilter,
                value: searchFilter,
                suffix: /* @__PURE__ */ jsxRuntime.jsx(
                  ButtonSelect.ButtonSelect,
                  {
                    className: styles.buttonSelectOverrides,
                    options: operators,
                    onChange: setOperator,
                    value: operator,
                    tooltip: operator.description,
                    narrow: true,
                    root: (_a = containerRef.current) != null ? _a : void 0
                  }
                )
              }
            ) }),
            /* @__PURE__ */ jsxRuntime.jsx(
              Button.Button,
              {
                tooltip: i18n.t("grafana-ui.table.filter-popup-aria-label-match-case", "Match case"),
                variant: matchCase ? "primary" : "secondary",
                onClick: () => setMatchCase((s) => !s),
                "aria-pressed": matchCase,
                icon: "text-fields"
              }
            )
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx(
            FilterList.FilterList,
            {
              onChange: setValues,
              values,
              options,
              caseSensitive: matchCase,
              searchFilter,
              operator
            }
          ),
          /* @__PURE__ */ jsxRuntime.jsxs(Stack.Stack, { justifyContent: "end", direction: "row-reverse", children: [
            /* @__PURE__ */ jsxRuntime.jsx(Button.Button, { size: "sm", onClick: onFilter, children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.table.filter-popup-apply", children: "Ok" }) }),
            /* @__PURE__ */ jsxRuntime.jsx(Button.Button, { size: "sm", variant: "secondary", onClick: onClose, children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.table.filter-popup-cancel", children: "Cancel" }) }),
            clearFilterVisible && /* @__PURE__ */ jsxRuntime.jsx(Button.Button, { fill: "text", size: "sm", onClick: onClearFilter, children: /* @__PURE__ */ jsxRuntime.jsx(i18n.Trans, { i18nKey: "grafana-ui.table.filter-popup-clear", children: "Clear filter" }) })
          ] })
        ] })
      }
    );
  }
);
FilterPopup.displayName = "FilterPopup";
const getStyles = (theme) => ({
  filterContainer: css.css({
    label: "filterContainer",
    width: "100%",
    minWidth: "320px",
    height: "100%",
    backgroundColor: theme.colors.background.primary,
    border: `1px solid ${theme.colors.border.weak}`,
    padding: theme.spacing(2),
    boxShadow: theme.shadows.z3,
    borderRadius: theme.shape.radius.default
  }),
  label: css.css({
    marginBottom: 0
  }),
  inputContainer: css.css({
    width: 300
  }),
  buttonSelectOverrides: css.css({
    fontSize: 12,
    "&:hover, &:focus, &:active": {
      color: theme.colors.text.primary,
      background: "transparent"
    }
  })
});

exports.FilterPopup = FilterPopup;
