// @grafana/ui/dist/cjs/components/Table/TableNG/Filter/utils.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var data = require('@grafana/data');
var i18n = require('@grafana/i18n');
var types = require('../types.cjs');
var utils = require('../utils.cjs');

"use strict";
function calculateUniqueFieldValues(rows, field) {
  if (!field || rows.length === 0) {
    return {};
  }
  const set = {};
  for (let index = 0; index < rows.length; index++) {
    const row = rows[index];
    const fieldValue = row[utils.getDisplayName(field)];
    const value = field.display ? data.formattedValueToString(field.display(fieldValue)) : String(fieldValue);
    set[value || "(Blanks)"] = value;
  }
  return set;
}
function getFilteredOptions(options, filterValues) {
  if (!filterValues) {
    return [];
  }
  return options.filter((option) => filterValues.some((filtered) => filtered.value === option.value));
}
function valuesToOptions(unique) {
  return Object.keys(unique).map((key) => ({ value: unique[key], label: key })).sort(sortOptions);
}
function sortOptions(a, b) {
  if (a.label === void 0 && b.label === void 0) {
    return 0;
  }
  if (a.label === void 0 && b.label !== void 0) {
    return -1;
  }
  if (a.label !== void 0 && b.label === void 0) {
    return 1;
  }
  if (a.label < b.label) {
    return -1;
  }
  if (a.label > b.label) {
    return 1;
  }
  return 0;
}
const operatorSelectableValues = () => {
  const toSelectableValue = (operator, label, description = label) => ({
    label: operator,
    ariaLabel: label,
    value: operator,
    description
  });
  return {
    [types.FilterOperator.CONTAINS]: toSelectableValue(
      types.FilterOperator.CONTAINS,
      i18n.t("grafana-ui.table.filter.operator.contains", "Contains")
    ),
    [types.FilterOperator.EQUALS]: toSelectableValue(
      types.FilterOperator.EQUALS,
      i18n.t("grafana-ui.table.filter.operator.equals", "Equals")
    ),
    [types.FilterOperator.NOT_EQUALS]: toSelectableValue(
      types.FilterOperator.NOT_EQUALS,
      i18n.t("grafana-ui.table.filter.operator.not-equals", "Not equals")
    ),
    [types.FilterOperator.GREATER]: toSelectableValue(
      types.FilterOperator.GREATER,
      i18n.t("grafana-ui.table.filter.operator.greater", "Greater")
    ),
    [types.FilterOperator.GREATER_OR_EQUAL]: toSelectableValue(
      types.FilterOperator.GREATER_OR_EQUAL,
      i18n.t("grafana-ui.table.filter.operator.greater-or-equal", "Greater or Equal")
    ),
    [types.FilterOperator.LESS]: toSelectableValue(types.FilterOperator.LESS, i18n.t("grafana-ui.table.filter.operator.less", "Less")),
    [types.FilterOperator.LESS_OR_EQUAL]: toSelectableValue(
      types.FilterOperator.LESS_OR_EQUAL,
      i18n.t("grafana-ui.table.filter.operator.less-or-equal", "Less or Equal")
    ),
    [types.FilterOperator.EXPRESSION]: toSelectableValue(
      types.FilterOperator.EXPRESSION,
      i18n.t("grafana-ui.table.filter.operator.expression", "Expression"),
      i18n.t(
        "grafana-ui.table.filter.operator.expression-description",
        'Bool Expression (Char $ represents the column value in the expression, e.g. "$ >= 10 && $ <= 12")'
      )
    )
  };
};

exports.calculateUniqueFieldValues = calculateUniqueFieldValues;
exports.getFilteredOptions = getFilteredOptions;
exports.operatorSelectableValues = operatorSelectableValues;
exports.valuesToOptions = valuesToOptions;
