// @grafana/ui/dist/cjs/components/DateTimePickers/TimeZonePicker.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var lodash = require('lodash');
var React = require('react');
var data = require('@grafana/data');
var i18n = require('@grafana/i18n');
var Select = require('../Select/Select.cjs');
var TimeZoneGroup = require('./TimeZonePicker/TimeZoneGroup.cjs');
var TimeZoneOffset = require('./TimeZonePicker/TimeZoneOffset.cjs');
var TimeZoneOption = require('./TimeZonePicker/TimeZoneOption.cjs');

"use strict";
const TimeZonePicker = (props) => {
  const {
    onChange,
    width,
    autoFocus = false,
    onBlur,
    value,
    includeInternal = false,
    disabled = false,
    inputId,
    menuShouldPortal = true,
    openMenuOnFocus = false
  } = props;
  const groupedTimeZones = useTimeZones(includeInternal);
  const selected = useSelectedTimeZone(groupedTimeZones, value);
  const filterBySearchIndex = useFilterBySearchIndex();
  const TimeZoneOption$1 = width && width <= 45 ? TimeZoneOption.CompactTimeZoneOption : TimeZoneOption.WideTimeZoneOption;
  const onChangeTz = React.useCallback(
    (selectable) => {
      if (!selectable || !lodash.isString(selectable.value)) {
        return onChange(value);
      }
      onChange(selectable.value);
    },
    [onChange, value]
  );
  return /* @__PURE__ */ jsxRuntime.jsx(
    Select.Select,
    {
      inputId,
      value: selected,
      placeholder: i18n.t("time-picker.zone.select-search-input", "Type to search (country, city, abbreviation)"),
      autoFocus,
      menuShouldPortal,
      openMenuOnFocus,
      width,
      filterOption: filterBySearchIndex,
      options: groupedTimeZones,
      onChange: onChangeTz,
      onBlur,
      components: { Option: TimeZoneOption$1, Group: TimeZoneGroup.TimeZoneGroup },
      disabled,
      "aria-label": i18n.t("time-picker.zone.select-aria-label", "Time zone picker")
    }
  );
};
const useTimeZones = (includeInternal) => {
  const now = Date.now();
  const timeZoneGroups = React.useMemo(() => {
    return data.getTimeZoneGroups(includeInternal).map((group) => {
      const options = group.zones.reduce((options2, zone) => {
        const info = data.getTimeZoneInfo(zone, now);
        if (!info) {
          return options2;
        }
        const name = info.name.replace(/_/g, " ");
        options2.push({
          label: name,
          value: info.zone,
          searchIndex: getSearchIndex(name, info, now)
        });
        return options2;
      }, []);
      return {
        label: group.name,
        options
      };
    });
  }, [includeInternal, now]);
  return timeZoneGroups;
};
const useSelectedTimeZone = (groups, timeZone) => {
  return React.useMemo(() => {
    if (timeZone === void 0) {
      return void 0;
    }
    const tz = lodash.toLower(timeZone);
    const group = groups.find((group2) => {
      if (!group2.label) {
        return isInternal(tz);
      }
      return tz.startsWith(lodash.toLower(group2.label));
    });
    return group == null ? void 0 : group.options.find((option) => {
      if (lodash.isEmpty(tz)) {
        return option.value === data.InternalTimeZones.default;
      }
      return lodash.toLower(option.value) === tz;
    });
  }, [groups, timeZone]);
};
const isInternal = (timeZone) => {
  switch (timeZone) {
    case data.InternalTimeZones.default:
    case data.InternalTimeZones.localBrowserTime:
    case data.InternalTimeZones.utc:
      return true;
    default:
      return false;
  }
};
const useFilterBySearchIndex = () => {
  return React.useCallback((option, searchQuery) => {
    if (!searchQuery || !option.data || !option.data.searchIndex) {
      return true;
    }
    return option.data.searchIndex.indexOf(lodash.toLower(searchQuery)) > -1;
  }, []);
};
const getSearchIndex = (label, info, timestamp) => {
  const parts = [
    lodash.toLower(info.zone),
    lodash.toLower(info.abbreviation),
    lodash.toLower(TimeZoneOffset.formatUtcOffset(timestamp, info.zone))
  ];
  if (label !== info.zone) {
    parts.push(lodash.toLower(label));
  }
  for (const country of info.countries) {
    parts.push(lodash.toLower(country.name));
    parts.push(lodash.toLower(country.code));
  }
  return parts.join("|");
};

exports.TimeZonePicker = TimeZonePicker;
