// @grafana/ui/dist/cjs/options/builder/legend.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var data = require('@grafana/data');
var i18n = require('@grafana/i18n');
var schema = require('@grafana/schema');

"use strict";
function addLegendOptions(builder, includeLegendCalcs = true, showLegend = true) {
  const category = [i18n.t("grafana-ui.builder.legend.category", "Legend")];
  builder.addBooleanSwitch({
    path: "legend.showLegend",
    name: i18n.t("grafana-ui.builder.legend.name-visibility", "Visibility"),
    category,
    description: "",
    defaultValue: showLegend
  }).addRadio({
    path: "legend.displayMode",
    name: i18n.t("grafana-ui.builder.legend.name-mode", "Mode"),
    category,
    description: "",
    defaultValue: schema.LegendDisplayMode.List,
    settings: {
      options: [
        { value: schema.LegendDisplayMode.List, label: i18n.t("grafana-ui.builder.legend.mode-options.label-list", "List") },
        { value: schema.LegendDisplayMode.Table, label: i18n.t("grafana-ui.builder.legend.mode-options.label-table", "Table") }
      ]
    },
    showIf: (c) => c.legend.showLegend
  }).addRadio({
    path: "legend.placement",
    name: i18n.t("grafana-ui.builder.legend.name-placement", "Placement"),
    category,
    description: "",
    defaultValue: "bottom",
    settings: {
      options: [
        { value: "bottom", label: i18n.t("grafana-ui.builder.legend.placement-options.label-bottom", "Bottom") },
        { value: "right", label: i18n.t("grafana-ui.builder.legend.placement-options.label-right", "Right") }
      ]
    },
    showIf: (c) => c.legend.showLegend
  }).addNumberInput({
    path: "legend.width",
    name: i18n.t("grafana-ui.builder.legend.name-width", "Width"),
    category,
    settings: {
      placeholder: "Auto"
    },
    showIf: (c) => c.legend.showLegend && c.legend.placement === "right"
  });
  if (includeLegendCalcs) {
    builder.addCustomEditor({
      id: "legend.calcs",
      path: "legend.calcs",
      name: i18n.t("grafana-ui.builder.legend.name-values", "Values"),
      category,
      description: i18n.t("grafana-ui.builder.legend.description-values", "Select values or calculations to show in legend"),
      editor: data.standardEditorsRegistry.get("stats-picker").editor,
      defaultValue: [],
      settings: {
        allowMultiple: true
      },
      showIf: (currentConfig) => currentConfig.legend.showLegend !== false
    });
  }
}

exports.addLegendOptions = addLegendOptions;
