// @grafana/ui/dist/cjs/options/builder/tooltip.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var i18n = require('@grafana/i18n');
var schema = require('@grafana/schema');

"use strict";
const optsWithHideZeros = {
  tooltip: {
    mode: schema.TooltipDisplayMode.Single,
    sort: schema.SortOrder.None,
    hideZeros: false
  }
};
function addTooltipOptions(builder, singleOnly = false, setProximity = false, defaultOptions) {
  var _a, _b, _c, _d;
  const category = [i18n.t("grafana-ui.builder.tooltip.category", "Tooltip")];
  const modeOptions = singleOnly ? [
    { value: schema.TooltipDisplayMode.Single, label: i18n.t("grafana-ui.builder.tooltip.modeOptions.label-single", "Single") },
    { value: schema.TooltipDisplayMode.None, label: i18n.t("grafana-ui.builder.tooltip.modeOptions.label-hidden", "Hidden") }
  ] : [
    { value: schema.TooltipDisplayMode.Single, label: i18n.t("grafana-ui.builder.tooltip.modeOptions.label-single", "Single") },
    { value: schema.TooltipDisplayMode.Multi, label: i18n.t("grafana-ui.builder.tooltip.modeOptions.label-all", "All") },
    { value: schema.TooltipDisplayMode.None, label: i18n.t("grafana-ui.builder.tooltip.modeOptions.label-hidden", "Hidden") }
  ];
  const sortOptions = [
    { value: schema.SortOrder.None, label: i18n.t("grafana-ui.builder.tooltip.sortOptions.label-none", "None") },
    { value: schema.SortOrder.Ascending, label: i18n.t("grafana-ui.builder.tooltip.sortOptions.label-ascending", "Ascending") },
    { value: schema.SortOrder.Descending, label: i18n.t("grafana-ui.builder.tooltip.sortOptions.label-descending", "Descending") }
  ];
  builder.addRadio({
    path: "tooltip.mode",
    name: i18n.t("grafana-ui.builder.tooltip.name-tooltip-mode", "Tooltip mode"),
    category,
    defaultValue: (_b = (_a = defaultOptions == null ? void 0 : defaultOptions.tooltip) == null ? void 0 : _a.mode) != null ? _b : schema.TooltipDisplayMode.Single,
    settings: {
      options: modeOptions
    }
  }).addRadio({
    path: "tooltip.sort",
    name: i18n.t("grafana-ui.builder.tooltip.name-values-sort-order", "Values sort order"),
    category,
    defaultValue: (_d = (_c = defaultOptions == null ? void 0 : defaultOptions.tooltip) == null ? void 0 : _c.sort) != null ? _d : schema.SortOrder.None,
    showIf: (options) => {
      var _a2;
      return ((_a2 = options.tooltip) == null ? void 0 : _a2.mode) === schema.TooltipDisplayMode.Multi;
    },
    settings: {
      options: sortOptions
    }
  }).addBooleanSwitch({
    path: "tooltip.hideZeros",
    name: i18n.t("grafana-ui.builder.tooltip.name-hide-zeros", "Hide zeros"),
    category,
    defaultValue: false,
    showIf: (options) => {
      var _a2, _b2;
      return ((_a2 = defaultOptions == null ? void 0 : defaultOptions.tooltip) == null ? void 0 : _a2.hideZeros) !== void 0 && ((_b2 = options.tooltip) == null ? void 0 : _b2.mode) === schema.TooltipDisplayMode.Multi;
    }
  });
  if (setProximity) {
    builder.addNumberInput({
      path: "tooltip.hoverProximity",
      name: i18n.t("grafana-ui.builder.tooltip.name-hover-proximity", "Hover proximity"),
      description: i18n.t(
        "grafana-ui.builder.tooltip.description-hover-proximity",
        "How close the cursor must be to a point to trigger the tooltip, in pixels"
      ),
      category,
      settings: {
        integer: true
      },
      showIf: (options) => {
        var _a2;
        return ((_a2 = options.tooltip) == null ? void 0 : _a2.mode) !== schema.TooltipDisplayMode.None;
      }
    });
  }
  builder.addNumberInput({
    path: "tooltip.maxWidth",
    name: i18n.t("grafana-ui.builder.tooltip.name-max-width", "Max width"),
    category,
    settings: {
      integer: true
    },
    showIf: (options) => {
      var _a2;
      return ((_a2 = options.tooltip) == null ? void 0 : _a2.mode) !== schema.TooltipDisplayMode.None;
    }
  }).addNumberInput({
    path: "tooltip.maxHeight",
    name: i18n.t("grafana-ui.builder.tooltip.name-max-height", "Max height"),
    category,
    defaultValue: void 0,
    settings: {
      integer: true
    },
    showIf: (options, data, annotations) => {
      var _a2;
      return ((_a2 = options.tooltip) == null ? void 0 : _a2.mode) === schema.TooltipDisplayMode.Multi || (annotations == null ? void 0 : annotations.some((df) => {
        var _a3, _b2;
        return ((_b2 = (_a3 = df.meta) == null ? void 0 : _a3.custom) == null ? void 0 : _b2.resultType) === "exemplar";
      }));
    }
  });
}

exports.addTooltipOptions = addTooltipOptions;
exports.optsWithHideZeros = optsWithHideZeros;
