// @grafana/runtime/dist/cjs/services/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendSrv = require('./backendSrv.cjs');
var dataSourceSrv = require('./dataSourceSrv.cjs');
var LocationSrv = require('./LocationSrv.cjs');
var EchoSrv = require('./EchoSrv.cjs');
var templateSrv = require('./templateSrv.cjs');
var live = require('./live.cjs');
var LocationService = require('./LocationService.cjs');
var appEvents = require('./appEvents.cjs');
var usePluginComponent = require('./pluginExtensions/usePluginComponent.cjs');
var usePluginComponents = require('./pluginExtensions/usePluginComponents.cjs');
var usePluginLinks = require('./pluginExtensions/usePluginLinks.cjs');
var usePluginFunctions = require('./pluginExtensions/usePluginFunctions.cjs');
var useHelpNavItem = require('./navigation/useHelpNavItem.cjs');
var getObservablePluginLinks = require('./pluginExtensions/getObservablePluginLinks.cjs');
var getObservablePluginComponents = require('./pluginExtensions/getObservablePluginComponents.cjs');
var utils = require('./pluginExtensions/utils.cjs');
var user = require('./user.cjs');
var RuntimeDataSource = require('./RuntimeDataSource.cjs');
var ScopesContext = require('./ScopesContext.cjs');

"use strict";

exports.getBackendSrv = backendSrv.getBackendSrv;
exports.isFetchError = backendSrv.isFetchError;
exports.setBackendSrv = backendSrv.setBackendSrv;
exports.getDataSourceSrv = dataSourceSrv.getDataSourceSrv;
exports.setDataSourceSrv = dataSourceSrv.setDataSourceSrv;
exports.getLocationSrv = LocationSrv.getLocationSrv;
exports.setLocationSrv = LocationSrv.setLocationSrv;
exports.EchoEventType = EchoSrv.EchoEventType;
exports.FakeEchoSrv = EchoSrv.FakeEchoSrv;
exports.getEchoSrv = EchoSrv.getEchoSrv;
exports.registerEchoBackend = EchoSrv.registerEchoBackend;
exports.setEchoSrv = EchoSrv.setEchoSrv;
exports.getTemplateSrv = templateSrv.getTemplateSrv;
exports.setTemplateSrv = templateSrv.setTemplateSrv;
exports.getGrafanaLiveSrv = live.getGrafanaLiveSrv;
exports.setGrafanaLiveSrv = live.setGrafanaLiveSrv;
exports.HistoryWrapper = LocationService.HistoryWrapper;
exports.LocationServiceProvider = LocationService.LocationServiceProvider;
exports.locationSearchToObject = LocationService.locationSearchToObject;
Object.defineProperty(exports, "locationService", {
  enumerable: true,
  get: function () { return LocationService.locationService; }
});
exports.navigationLogger = LocationService.navigationLogger;
exports.setLocationService = LocationService.setLocationService;
exports.useLocationService = LocationService.useLocationService;
exports.CopyPanelEvent = appEvents.CopyPanelEvent;
exports.RefreshEvent = appEvents.RefreshEvent;
exports.ThemeChangedEvent = appEvents.ThemeChangedEvent;
exports.TimeRangeUpdatedEvent = appEvents.TimeRangeUpdatedEvent;
exports.getAppEvents = appEvents.getAppEvents;
exports.setAppEvents = appEvents.setAppEvents;
exports.setPluginComponentHook = usePluginComponent.setPluginComponentHook;
exports.usePluginComponent = usePluginComponent.usePluginComponent;
exports.setPluginComponentsHook = usePluginComponents.setPluginComponentsHook;
exports.usePluginComponents = usePluginComponents.usePluginComponents;
exports.setPluginLinksHook = usePluginLinks.setPluginLinksHook;
exports.usePluginLinks = usePluginLinks.usePluginLinks;
exports.setPluginFunctionsHook = usePluginFunctions.setPluginFunctionsHook;
exports.usePluginFunctions = usePluginFunctions.usePluginFunctions;
exports.setHelpNavItemHook = useHelpNavItem.setHelpNavItemHook;
exports.useHelpNavItem = useHelpNavItem.useHelpNavItem;
exports.getObservablePluginLinks = getObservablePluginLinks.getObservablePluginLinks;
exports.getObservablePluginComponents = getObservablePluginComponents.getObservablePluginComponents;
exports.getLimitedComponentsToRender = utils.getLimitedComponentsToRender;
exports.isPluginExtensionComponent = utils.isPluginExtensionComponent;
exports.isPluginExtensionLink = utils.isPluginExtensionLink;
exports.renderLimitedComponents = utils.renderLimitedComponents;
exports.setCurrentUser = user.setCurrentUser;
exports.RuntimeDataSource = RuntimeDataSource.RuntimeDataSource;
exports.ScopesContext = ScopesContext.ScopesContext;
exports.useScopes = ScopesContext.useScopes;
