// @grafana/runtime/dist/cjs/services/RuntimeDataSource.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var data = require('@grafana/data');

"use strict";
class RuntimeDataSource extends data.DataSourceApi {
  constructor(pluginId, uid) {
    const instanceSettings = {
      name: "RuntimeDataSource-" + pluginId,
      uid,
      type: pluginId,
      readOnly: true,
      jsonData: {},
      access: "direct",
      meta: {
        id: pluginId,
        name: "RuntimeDataSource-" + pluginId,
        type: data.PluginType.datasource,
        info: {
          author: {
            name: ""
          },
          description: "",
          links: [],
          logos: {
            large: "",
            small: ""
          },
          screenshots: [],
          updated: "",
          version: ""
        },
        module: "",
        baseUrl: ""
      }
    };
    super(instanceSettings);
    this.instanceSettings = instanceSettings;
  }
  testDatasource() {
    return Promise.resolve({
      status: "success",
      message: ""
    });
  }
}

exports.RuntimeDataSource = RuntimeDataSource;
