// @grafana/ui/dist/cjs/components/VizTooltip/types.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

"use strict";
var ColorIndicator = /* @__PURE__ */ ((ColorIndicator2) => {
  ColorIndicator2["series"] = "series";
  ColorIndicator2["value"] = "value";
  ColorIndicator2["hexagon"] = "hexagon";
  ColorIndicator2["pie_1_4"] = "pie_1_4";
  ColorIndicator2["pie_2_4"] = "pie_2_4";
  ColorIndicator2["pie_3_4"] = "pie_3_4";
  ColorIndicator2["marker_sm"] = "marker_sm";
  ColorIndicator2["marker_md"] = "marker_md";
  ColorIndicator2["marker_lg"] = "marker_lg";
  return ColorIndicator2;
})(ColorIndicator || {});
var ColorPlacement = /* @__PURE__ */ ((ColorPlacement2) => {
  ColorPlacement2["hidden"] = "hidden";
  ColorPlacement2["first"] = "first";
  ColorPlacement2["leading"] = "leading";
  ColorPlacement2["trailing"] = "trailing";
  return ColorPlacement2;
})(ColorPlacement || {});
const DEFAULT_COLOR_INDICATOR = "series" /* series */;

exports.ColorIndicator = ColorIndicator;
exports.ColorPlacement = ColorPlacement;
exports.DEFAULT_COLOR_INDICATOR = DEFAULT_COLOR_INDICATOR;
