// @grafana/ui/dist/cjs/components/VizTooltip/VizTooltip.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var Portal = require('../Portal/Portal.cjs');
var VizTooltipContainer = require('./VizTooltipContainer.cjs');

"use strict";
const VizTooltip = ({ content, position, offset }) => {
  const styles = ThemeContext.useStyles2(getStyles);
  if (position) {
    return /* @__PURE__ */ jsxRuntime.jsx(Portal.Portal, { className: styles.portal, children: /* @__PURE__ */ jsxRuntime.jsx(VizTooltipContainer.VizTooltipContainer, { position, offset: offset || { x: 0, y: 0 }, children: content }) });
  }
  return null;
};
VizTooltip.displayName = "VizTooltip";
const getStyles = () => {
  return {
    portal: css.css({
      position: "absolute",
      top: 0,
      left: 0,
      pointerEvents: "none",
      width: "100%",
      height: "100%"
    })
  };
};

exports.VizTooltip = VizTooltip;
