// @grafana/ui/dist/cjs/components/RefreshPicker/RefreshPicker.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var dateFns = require('date-fns');
var React = require('react');
var data = require('@grafana/data');
var e2eSelectors = require('@grafana/e2e-selectors');
var i18n = require('@grafana/i18n');
var ButtonGroup = require('../Button/ButtonGroup.cjs');
var ButtonSelect = require('../Dropdown/ButtonSelect.cjs');
var ToolbarButton = require('../ToolbarButton/ToolbarButton.cjs');

"use strict";
const defaultIntervals = ["5s", "10s", "30s", "1m", "5m", "15m", "30m", "1h", "2h", "1d"];
const offOption = {
  label: "Off",
  value: "",
  ariaLabel: "Turn off auto refresh"
};
const liveOption = {
  label: "Live",
  value: "LIVE",
  ariaLabel: "Turn on live streaming"
};
const autoOption = {
  label: "Auto",
  value: "auto",
  ariaLabel: "Select refresh from the query range"
};
const RefreshPickerComponent = React.memo((props) => {
  const {
    intervals,
    onRefresh,
    onIntervalChanged,
    value,
    tooltip,
    isLoading,
    isLive,
    text,
    noIntervalPicker,
    showAutoInterval,
    width,
    primary,
    isOnCanvas
  } = props;
  const currentValue = value || "";
  const options = intervalsToOptions({ intervals, showAutoInterval });
  const option = options.find(({ value: value2 }) => value2 === currentValue);
  const translatedOffOption = translateOption(offOption.value);
  let selectedValue = option || translatedOffOption;
  const handleChangeSelect = (item) => {
    if (onIntervalChanged && item.value != null) {
      onIntervalChanged(item.value);
    }
  };
  const getVariant = () => {
    if (isLive) {
      return "primary";
    }
    if (primary) {
      return "primary";
    }
    return isOnCanvas ? "canvas" : "default";
  };
  const variant = getVariant();
  if (selectedValue.label === translatedOffOption.label) {
    selectedValue = { value: "" };
  }
  const durationAriaLabel = selectedValue.ariaLabel;
  const ariaLabelDurationSelectedMessage = i18n.t(
    "refresh-picker.aria-label.duration-selected",
    "Choose refresh time interval with current interval {{durationAriaLabel}} selected",
    { durationAriaLabel }
  );
  const ariaLabelChooseIntervalMessage = i18n.t(
    "refresh-picker.aria-label.choose-interval",
    "Auto refresh turned off. Choose refresh time interval"
  );
  const ariaLabel = selectedValue.value === "" ? ariaLabelChooseIntervalMessage : ariaLabelDurationSelectedMessage;
  const tooltipIntervalSelected = i18n.t("refresh-picker.tooltip.interval-selected", "Set auto refresh interval");
  const tooltipAutoRefreshOff = i18n.t("refresh-picker.tooltip.turned-off", "Auto refresh off");
  const tooltipAutoRefresh = selectedValue.value === "" ? tooltipAutoRefreshOff : tooltipIntervalSelected;
  return /* @__PURE__ */ jsxRuntime.jsxs(ButtonGroup.ButtonGroup, { className: "refresh-picker", children: [
    /* @__PURE__ */ jsxRuntime.jsx(
      ToolbarButton.ToolbarButton,
      {
        "aria-label": text,
        tooltip,
        onClick: onRefresh,
        variant,
        icon: isLoading ? "spinner" : "sync",
        style: width ? { width } : void 0,
        "data-testid": e2eSelectors.selectors.components.RefreshPicker.runButtonV2,
        children: text
      }
    ),
    !noIntervalPicker && /* @__PURE__ */ jsxRuntime.jsx(
      ButtonSelect.ButtonSelect,
      {
        className: css.css({
          borderTopLeftRadius: "unset",
          borderBottomLeftRadius: "unset"
        }),
        value: selectedValue,
        options,
        onChange: handleChangeSelect,
        variant,
        "data-testid": e2eSelectors.selectors.components.RefreshPicker.intervalButtonV2,
        "aria-label": ariaLabel,
        tooltip: tooltipAutoRefresh
      }
    )
  ] });
});
RefreshPickerComponent.displayName = "RefreshPicker";
const RefreshPicker = Object.assign(RefreshPickerComponent, {
  isLive: (refreshInterval) => refreshInterval === liveOption.value,
  liveOption,
  offOption,
  autoOption
});
const translateOption = (option) => {
  switch (option) {
    case liveOption.value:
      return {
        label: i18n.t("refresh-picker.live-option.label", "Live"),
        value: option,
        ariaLabel: i18n.t("refresh-picker.live-option.aria-label", "Turn on live streaming")
      };
    case offOption.value:
      return {
        label: i18n.t("refresh-picker.off-option.label", "Off"),
        value: option,
        ariaLabel: i18n.t("refresh-picker.off-option.aria-label", "Turn off auto refresh")
      };
    case autoOption.value:
      return {
        label: i18n.t("refresh-picker.auto-option.label", autoOption.label),
        value: option,
        ariaLabel: i18n.t("refresh-picker.auto-option.aria-label", autoOption.ariaLabel)
      };
  }
  return {
    label: option,
    value: option
  };
};
const intervalsToOptions = ({
  intervals = defaultIntervals,
  showAutoInterval = false
} = {}) => {
  const options = intervals.map((interval) => {
    const duration = data.parseDuration(interval);
    const ariaLabel = dateFns.formatDuration(duration);
    return {
      label: interval,
      value: interval,
      ariaLabel
    };
  });
  if (showAutoInterval) {
    options.unshift(translateOption(autoOption.value));
  }
  options.unshift(translateOption(offOption.value));
  return options;
};

exports.RefreshPicker = RefreshPicker;
exports.defaultIntervals = defaultIntervals;
exports.intervalsToOptions = intervalsToOptions;
exports.translateOption = translateOption;
