// @grafana/ui/dist/cjs/components/Tags/Tag.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var Skeleton = require('react-loading-skeleton');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var skeleton = require('../../utils/skeleton.cjs');
var tags = require('../../utils/tags.cjs');
var Icon = require('../Icon/Icon.cjs');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var Skeleton__default = /*#__PURE__*/_interopDefaultCompat(Skeleton);

"use strict";
const TagComponent = React.forwardRef(({ name, onClick, icon, className, colorIndex, ...rest }, ref) => {
  const theme = ThemeContext.useTheme2();
  const styles = getTagStyles(theme, name, colorIndex);
  const onTagClick = (event) => {
    event.preventDefault();
    event.stopPropagation();
    onClick == null ? void 0 : onClick(name, event);
  };
  const classes = css.cx(styles.wrapper, className, { [styles.hover]: onClick !== void 0 });
  return onClick ? /* @__PURE__ */ jsxRuntime.jsxs("button", { ...rest, className: classes, onClick: onTagClick, ref, children: [
    icon && /* @__PURE__ */ jsxRuntime.jsx(Icon.Icon, { name: icon }),
    name
  ] }) : /* @__PURE__ */ jsxRuntime.jsxs("span", { ...rest, className: classes, ref, children: [
    icon && /* @__PURE__ */ jsxRuntime.jsx(Icon.Icon, { name: icon }),
    name
  ] });
});
TagComponent.displayName = "Tag";
const TagSkeleton = ({ rootProps }) => {
  const styles = ThemeContext.useStyles2(getSkeletonStyles);
  return /* @__PURE__ */ jsxRuntime.jsx(Skeleton__default.default, { width: 60, height: 22, containerClassName: styles.container, ...rootProps });
};
const Tag = skeleton.attachSkeleton(TagComponent, TagSkeleton);
const getSkeletonStyles = () => ({
  container: css.css({
    lineHeight: 1
  })
});
const getTagStyles = (theme, name, colorIndex) => {
  let colors;
  if (colorIndex === void 0) {
    colors = tags.getTagColorsFromName(name);
  } else {
    colors = tags.getTagColor(colorIndex);
  }
  return {
    wrapper: css.css({
      appearance: "none",
      borderStyle: "none",
      fontWeight: theme.typography.fontWeightMedium,
      fontSize: theme.typography.size.sm,
      lineHeight: theme.typography.bodySmall.lineHeight,
      verticalAlign: "baseline",
      backgroundColor: colors.color,
      color: theme.v1.palette.gray98,
      whiteSpace: "pre",
      textShadow: "none",
      padding: "3px 6px",
      borderRadius: theme.shape.radius.sm
    }),
    hover: css.css({
      "&:hover": {
        opacity: 0.85,
        cursor: "pointer"
      }
    })
  };
};

exports.Tag = Tag;
