// @grafana/ui/dist/cjs/components/Tags/TagList.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var i18n = require('@grafana/i18n');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var skeleton = require('../../utils/skeleton.cjs');
var Tag = require('./Tag.cjs');

"use strict";
const TagListComponent = React.memo(
  React.forwardRef(
    ({ displayMax, tags, icon, onClick, className, getAriaLabel, getColorIndex }, ref) => {
      const theme = ThemeContext.useTheme2();
      const styles = getStyles(theme, Boolean(displayMax && displayMax > 0));
      const numTags = tags.length;
      const tagsToDisplay = displayMax ? tags.slice(0, displayMax) : tags;
      return /* @__PURE__ */ jsxRuntime.jsxs("ul", { className: css.cx(styles.wrapper, className), "aria-label": i18n.t("grafana-ui.tags.list-label", "Tags"), ref, children: [
        tagsToDisplay.map((tag, i) => /* @__PURE__ */ jsxRuntime.jsx("li", { className: styles.li, children: /* @__PURE__ */ jsxRuntime.jsx(
          Tag.Tag,
          {
            name: tag,
            icon,
            onClick,
            "aria-label": getAriaLabel == null ? void 0 : getAriaLabel(tag, i),
            "data-tag-id": i,
            colorIndex: getColorIndex == null ? void 0 : getColorIndex(tag, i)
          }
        ) }, tag)),
        displayMax && displayMax > 0 && numTags - displayMax > 0 && /* @__PURE__ */ jsxRuntime.jsx("li", { className: styles.li, children: /* @__PURE__ */ jsxRuntime.jsxs("span", { className: styles.moreTagsLabel, children: [
          "+ ",
          numTags - displayMax
        ] }) })
      ] });
    }
  )
);
TagListComponent.displayName = "TagList";
const TagListSkeleton = ({ rootProps }) => {
  const styles = ThemeContext.useStyles2(getSkeletonStyles);
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: styles.container, ...rootProps, children: [
    /* @__PURE__ */ jsxRuntime.jsx(Tag.Tag.Skeleton, {}),
    /* @__PURE__ */ jsxRuntime.jsx(Tag.Tag.Skeleton, {})
  ] });
};
const TagList = skeleton.attachSkeleton(TagListComponent, TagListSkeleton);
const getSkeletonStyles = (theme) => ({
  container: css.css({
    display: "flex",
    gap: theme.spacing(1)
  })
});
const getStyles = (theme, isTruncated) => {
  return {
    wrapper: css.css({
      position: "relative",
      alignItems: isTruncated ? "center" : "unset",
      display: "flex",
      flex: "1 1 auto",
      flexWrap: "wrap",
      flexShrink: isTruncated ? 0 : 1,
      justifyContent: "flex-end",
      gap: "6px"
    }),
    moreTagsLabel: css.css({
      color: theme.colors.text.secondary,
      fontSize: theme.typography.size.sm
    }),
    li: css.css({
      listStyle: "none"
    })
  };
};

exports.TagList = TagList;
