// @grafana/ui/dist/cjs/components/Table/TableNG/Cells/AutoCell.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var data = require('@grafana/data');
var MaybeWrapWithLink = require('../components/MaybeWrapWithLink.cjs');
var constants = require('../constants.cjs');
var styles = require('../styles.cjs');

"use strict";
function AutoCell({ value, field, rowIdx }) {
  const displayValue = field.display(value);
  const formattedValue = data.formattedValueToString(displayValue);
  return /* @__PURE__ */ jsxRuntime.jsx(MaybeWrapWithLink.MaybeWrapWithLink, { field, rowIdx, children: formattedValue });
}
const getStyles = (_theme, { textWrap, shouldOverflow, maxHeight }) => css.css({
  ...textWrap && { whiteSpace: "pre-line" },
  ...shouldOverflow && {
    [styles.getActiveCellSelector(Boolean(maxHeight))]: {
      whiteSpace: "pre-line"
    }
  },
  ...maxHeight != null && textWrap && {
    height: "auto",
    overflowY: "hidden",
    display: "-webkit-box",
    WebkitBoxOrient: "vertical",
    WebkitLineClamp: Math.floor(maxHeight / constants.TABLE.LINE_HEIGHT),
    [styles.getActiveCellSelector(true)]: {
      display: "flex",
      WebkitLineClamp: "none",
      WebkitBoxOrient: "unset",
      overflowY: "auto",
      height: "fit-content"
    }
  }
});
const getJsonCellStyles = (_theme, { textWrap, shouldOverflow, maxHeight }) => css.css({
  fontFamily: "monospace",
  ...textWrap && { whiteSpace: "pre" },
  ...shouldOverflow && {
    [styles.getActiveCellSelector(Boolean(maxHeight))]: {
      whiteSpace: "pre"
    }
  }
});

exports.AutoCell = AutoCell;
exports.getJsonCellStyles = getJsonCellStyles;
exports.getStyles = getStyles;
