// @grafana/ui/dist/cjs/components/Table/TableNG/Cells/DataLinksCell.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var utils = require('../utils.cjs');

"use strict";
const DataLinksCell = ({ field, rowIdx }) => {
  const links = utils.getCellLinks(field, rowIdx);
  if (!(links == null ? void 0 : links.length)) {
    return null;
  }
  return links.map((link, idx) => /* @__PURE__ */ jsxRuntime.jsx("a", { onClick: link.onClick, href: link.href, target: link.target, children: link.title }, idx));
};
const getStyles = (theme, { textWrap, textAlign }) => css.css({
  ...textWrap && {
    flexDirection: "column",
    justifyContent: "center",
    alignItems: `${utils.getJustifyContent(textAlign)} !important`
    // we can't guarantee order, and alignItems is set on a sibling class.
  },
  "> a": {
    flexWrap: "nowrap",
    ...!textWrap && {
      paddingInline: theme.spacing(0.5),
      borderRight: `2px solid ${theme.colors.border.medium}`,
      "&:first-child": {
        paddingInlineStart: 0
      },
      "&:last-child": {
        paddingInlineEnd: 0,
        borderRight: "none"
      }
    }
  }
});

exports.DataLinksCell = DataLinksCell;
exports.getStyles = getStyles;
