// @grafana/ui/dist/esm/components/Table/TableNG/Cells/MarkdownCell.mjs
import { jsx } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import { renderMarkdown } from '@grafana/data';
import { MaybeWrapWithLink } from '../components/MaybeWrapWithLink.mjs';
import { getActiveCellSelector } from '../styles.mjs';

"use strict";
function MarkdownCell({ field, rowIdx, disableSanitizeHtml }) {
  const rawValue = field.values[rowIdx];
  if (rawValue == null) {
    return null;
  }
  const renderValue = field.display(rawValue);
  return /* @__PURE__ */ jsx(MaybeWrapWithLink, { field, rowIdx, children: /* @__PURE__ */ jsx(
    "div",
    {
      className: "markdown-container",
      dangerouslySetInnerHTML: {
        __html: renderMarkdown(renderValue.text, { noSanitize: disableSanitizeHtml }).trim()
      }
    }
  ) });
}
const getStyles = (theme, { maxHeight }) => css({
  [`&, ${getActiveCellSelector(Boolean(maxHeight))}`]: {
    whiteSpace: "normal"
  },
  ".markdown-container": {
    width: "100%",
    // for elements like `p`, `h*`, etc. which have an inherent margin,
    // we want to remove the bottom margin for the last one in the container.
    "> *:last-child": {
      marginBottom: 0
    }
  },
  "ol, ul": {
    paddingLeft: theme.spacing(2.5)
  },
  p: {
    whiteSpace: "pre-line"
  }
});

export { MarkdownCell, getStyles };
