// @grafana/ui/dist/cjs/components/Forms/RadioButtonGroup/RadioButtonGroup.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var lodash = require('lodash');
var React = require('react');
var data = require('@grafana/data');
var ThemeContext = require('../../../themes/ThemeContext.cjs');
var Icon = require('../../Icon/Icon.cjs');
var RadioButton = require('./RadioButton.cjs');

"use strict";
function RadioButtonGroup({
  options,
  value,
  onChange,
  onClick,
  disabled,
  disabledOptions,
  size = "md",
  id,
  className,
  fullWidth = false,
  autoFocus = false,
  "aria-label": ariaLabel,
  invalid = false
}) {
  const handleOnChange = React.useCallback(
    (option) => {
      return () => {
        if (onChange) {
          onChange(option.value);
        }
      };
    },
    [onChange]
  );
  const handleOnClick = React.useCallback(
    (option) => {
      return () => {
        if (onClick) {
          onClick(option.value);
        }
      };
    },
    [onClick]
  );
  const internalId = id != null ? id : lodash.uniqueId("radiogroup-");
  const groupName = React.useRef(internalId);
  const styles = ThemeContext.useStyles2(getStyles);
  const activeButtonRef = React.useRef(null);
  React.useEffect(() => {
    if (autoFocus && activeButtonRef.current) {
      activeButtonRef.current.focus();
    }
  }, [autoFocus]);
  return /* @__PURE__ */ jsxRuntime.jsx(
    "div",
    {
      role: "radiogroup",
      "aria-label": ariaLabel,
      className: css.cx(styles.radioGroup, fullWidth && styles.fullWidth, invalid && styles.invalid, className),
      children: options.map((opt, i) => {
        const isItemDisabled = disabledOptions && opt.value && disabledOptions.includes(opt.value);
        const icon = opt.icon ? data.toIconName(opt.icon) : void 0;
        const hasNonIconPart = Boolean(opt.imgUrl || opt.label || opt.component);
        return /* @__PURE__ */ jsxRuntime.jsxs(
          RadioButton.RadioButton,
          {
            size,
            disabled: isItemDisabled || disabled,
            active: value === opt.value,
            "aria-label": opt.ariaLabel,
            onChange: handleOnChange(opt),
            onClick: handleOnClick(opt),
            id: `option-${opt.value}-${internalId}`,
            name: groupName.current,
            description: opt.description,
            fullWidth,
            ref: value === opt.value ? activeButtonRef : void 0,
            children: [
              icon && /* @__PURE__ */ jsxRuntime.jsx(Icon.Icon, { name: icon, className: css.cx(hasNonIconPart && styles.icon) }),
              opt.imgUrl && /* @__PURE__ */ jsxRuntime.jsx("img", { src: opt.imgUrl, alt: opt.label, className: styles.img }),
              opt.label,
              " ",
              opt.component ? /* @__PURE__ */ jsxRuntime.jsx(opt.component, {}) : null
            ]
          },
          `o.label-${i}`
        );
      })
    }
  );
}
RadioButtonGroup.displayName = "RadioButtonGroup";
const getStyles = (theme) => {
  return {
    radioGroup: css.css({
      backgroundColor: theme.colors.background.primary,
      display: "inline-flex",
      flexDirection: "row",
      flexWrap: "nowrap",
      border: `1px solid ${theme.components.input.borderColor}`,
      borderRadius: theme.shape.radius.default,
      padding: RadioButton.RADIO_GROUP_PADDING,
      "&:hover": {
        borderColor: theme.components.input.borderHover
      }
    }),
    fullWidth: css.css({
      display: "flex",
      flexGrow: 1
    }),
    icon: css.css({
      marginRight: "6px"
    }),
    img: css.css({
      width: theme.spacing(2),
      height: theme.spacing(2),
      marginRight: theme.spacing(1)
    }),
    invalid: css.css({
      border: `1px solid ${theme.colors.error.border}`
    })
  };
};

exports.RadioButtonGroup = RadioButtonGroup;
