// @grafana/ui/dist/esm/components/Table/TableNG/components/RowExpander.mjs
import { jsx } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import { selectors } from '@grafana/e2e-selectors';
import { t } from '@grafana/i18n';
import { useStyles2 } from '../../../../themes/ThemeContext.mjs';
import { Icon } from '../../../Icon/Icon.mjs';

"use strict";
function RowExpander({ onCellExpand, isExpanded, rowId }) {
  const styles = useStyles2(getStyles);
  function handleKeyDown(e) {
    if (e.key === " " || e.key === "Enter") {
      e.preventDefault();
      onCellExpand(e);
    }
  }
  const label = isExpanded ? t("grafana-ui.row-expander-ng.aria-label-collapse", "Collapse row") : t("grafana-ui.row-expander.aria-label-expand", "Expand row");
  return /* @__PURE__ */ jsx(
    "div",
    {
      role: "button",
      tabIndex: 0,
      className: styles.expanderCell,
      onClick: onCellExpand,
      onKeyDown: handleKeyDown,
      "aria-label": label,
      "data-testid": selectors.components.Panels.Visualization.TableNG.RowExpander,
      "aria-expanded": isExpanded,
      "aria-controls": rowId,
      children: /* @__PURE__ */ jsx(Icon, { name: isExpanded ? "angle-down" : "angle-right", size: "lg", "aria-hidden": "true" })
    }
  );
}
const getStyles = (_theme) => ({
  expanderCell: css({
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    cursor: "pointer"
  })
});

export { RowExpander };
