// @grafana/ui/dist/esm/components/Table/TableNG/components/SummaryCell.mjs
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { css } from '@emotion/css';
import clsx from 'clsx';
import { useMemo } from 'react';
import { ReducerID, fieldReducers } from '@grafana/data';
import { selectors } from '@grafana/e2e-selectors';
import { t } from '@grafana/i18n';
import { useStyles2, useTheme2 } from '../../../../themes/ThemeContext.mjs';
import { useReducerEntries } from '../hooks.mjs';
import { getDefaultCellStyles } from '../styles.mjs';
import { getDisplayName, getJustifyContent } from '../utils.mjs';

"use strict";
const getReducerName = (reducerId) => {
  var _a;
  if (reducerId === ReducerID.countAll) {
    return t("grafana-ui.table.footer.reducer.count", "Count");
  }
  return ((_a = fieldReducers.get(reducerId)) == null ? void 0 : _a.name) || reducerId;
};
const SummaryCell = ({
  rows,
  footers,
  field,
  colIdx,
  hideLabel = false,
  rowLabel = false,
  textAlign
}) => {
  const styles = useStyles2(getStyles, textAlign, hideLabel);
  const theme = useTheme2();
  const defaultFooterCellStyles = getDefaultCellStyles(theme, {
    textAlign: "left",
    // alignment is set in footerItem
    shouldOverflow: true,
    textWrap: false
  });
  const displayName = getDisplayName(field);
  const reducerResultsEntries = useReducerEntries(field, rows, displayName, colIdx);
  const cellClass = clsx(styles.footerCell, defaultFooterCellStyles);
  const firstFooterReducers = useMemo(() => {
    var _a, _b;
    for (const footer of footers) {
      if ((_b = (_a = footer == null ? void 0 : footer.reducers) == null ? void 0 : _a.length) != null ? _b : 0 > 0) {
        return footer.reducers;
      }
    }
    return;
  }, [footers]);
  const renderRowLabel = rowLabel && reducerResultsEntries.length === 0 && Boolean(firstFooterReducers);
  const SummaryCellItem = ({ children }) => /* @__PURE__ */ jsx("div", { className: styles.footerItem, children });
  const SummaryCellLabel = ({ children }) => /* @__PURE__ */ jsx(
    "div",
    {
      "data-testid": selectors.components.Panels.Visualization.TableNG.Footer.ReducerLabel,
      className: styles.footerItemLabel,
      children
    }
  );
  const SummaryCellValue = ({ children }) => /* @__PURE__ */ jsx(
    "div",
    {
      "data-testid": selectors.components.Panels.Visualization.TableNG.Footer.Value,
      className: styles.footerItemValue,
      children
    }
  );
  return /* @__PURE__ */ jsxs(
    "div",
    {
      className: cellClass,
      "data-testid": reducerResultsEntries.length === 0 && !renderRowLabel ? "summary-cell-empty" : void 0,
      children: [
        reducerResultsEntries.map(([reducerId, reducerResult]) => {
          if (reducerResult === null) {
            return /* @__PURE__ */ jsx(SummaryCellItem, { children: rowLabel ? /* @__PURE__ */ jsx(SummaryCellLabel, { children: getReducerName(reducerId) }) : /* @__PURE__ */ jsx(Fragment, { children: "\xA0" }) }, reducerId);
          }
          return /* @__PURE__ */ jsxs(SummaryCellItem, { children: [
            !hideLabel && /* @__PURE__ */ jsx(SummaryCellLabel, { children: getReducerName(reducerId) }),
            /* @__PURE__ */ jsx(SummaryCellValue, { children: reducerResult })
          ] }, reducerId);
        }),
        renderRowLabel && firstFooterReducers.map((reducerId) => /* @__PURE__ */ jsx(SummaryCellItem, { children: /* @__PURE__ */ jsx(SummaryCellLabel, { children: getReducerName(reducerId) }) }, reducerId))
      ]
    }
  );
};
const getStyles = (theme, textAlign, hideLabel) => ({
  footerCell: css({
    flexDirection: "column",
    minHeight: "100%",
    width: "100%"
  }),
  footerItem: css({
    display: "flex",
    flexDirection: "row",
    justifyContent: hideLabel ? getJustifyContent(textAlign) : "space-between",
    alignItems: "flex-start",
    width: "100%",
    gap: theme.spacing(0.5)
  }),
  footerItemLabel: css({
    flexShrink: 0,
    overflow: "hidden",
    textOverflow: "ellipsis",
    whiteSpace: "nowrap",
    color: theme.colors.text.secondary,
    fontSize: theme.typography.bodySmall.fontSize,
    fontWeight: theme.typography.fontWeightLight,
    textTransform: "uppercase",
    lineHeight: "22px"
  }),
  footerItemValue: css({
    overflow: "hidden",
    textOverflow: "ellipsis",
    fontWeight: theme.typography.fontWeightMedium
  })
});

export { SummaryCell, getStyles };
