// @grafana/ui/dist/cjs/components/VizLegend/VizLegendList.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var ThemeContext = require('../../themes/ThemeContext.cjs');
var InlineList = require('../List/InlineList.cjs');
var List = require('../List/List.cjs');
var VizLegendListItem = require('./VizLegendListItem.cjs');

"use strict";
const VizLegendList = ({
  items,
  itemRenderer,
  onLabelMouseOver,
  onLabelMouseOut,
  onLabelClick,
  placement,
  className,
  readonly
}) => {
  const styles = ThemeContext.useStyles2(getStyles);
  if (!itemRenderer) {
    itemRenderer = (item) => /* @__PURE__ */ jsxRuntime.jsx(
      VizLegendListItem.VizLegendListItem,
      {
        item,
        onLabelClick,
        onLabelMouseOver,
        onLabelMouseOut,
        readonly
      }
    );
  }
  const getItemKey = (item) => `${item.getItemKey ? item.getItemKey() : item.label}`;
  switch (placement) {
    case "right": {
      const renderItem = (item, index) => {
        return /* @__PURE__ */ jsxRuntime.jsx("span", { className: styles.itemRight, children: itemRenderer(item, index) });
      };
      return /* @__PURE__ */ jsxRuntime.jsx("div", { className: css.cx(styles.rightWrapper, className), children: /* @__PURE__ */ jsxRuntime.jsx(List.List, { items, renderItem, getItemKey }) });
    }
    case "bottom":
    default: {
      const leftItems = items.filter((item) => item.yAxis === 1);
      const rightItems = items.filter((item) => item.yAxis !== 1);
      const renderItem = (item, index) => {
        return /* @__PURE__ */ jsxRuntime.jsx("span", { className: styles.itemBottom, children: itemRenderer(item, index) });
      };
      return /* @__PURE__ */ jsxRuntime.jsxs("div", { className: css.cx(styles.bottomWrapper, className), children: [
        leftItems.length > 0 && /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles.section, children: /* @__PURE__ */ jsxRuntime.jsx(InlineList.InlineList, { items: leftItems, renderItem, getItemKey }) }),
        rightItems.length > 0 && /* @__PURE__ */ jsxRuntime.jsx("div", { className: css.cx(styles.section, styles.sectionRight), children: /* @__PURE__ */ jsxRuntime.jsx(InlineList.InlineList, { items: rightItems, renderItem, getItemKey }) })
      ] });
    }
  }
};
VizLegendList.displayName = "VizLegendList";
const getStyles = (theme) => {
  const itemStyles = css.css({
    paddingRight: "10px",
    display: "flex",
    fontSize: theme.typography.bodySmall.fontSize,
    whiteSpace: "nowrap"
  });
  return {
    itemBottom: itemStyles,
    itemRight: css.cx(
      itemStyles,
      css.css({
        marginBottom: theme.spacing(0.5)
      })
    ),
    rightWrapper: css.css({
      padding: theme.spacing(0.5)
    }),
    bottomWrapper: css.css({
      display: "flex",
      flexWrap: "wrap",
      justifyContent: "space-between",
      width: "100%",
      padding: theme.spacing(0.5),
      gap: "15px 25px"
    }),
    section: css.css({
      display: "flex"
    }),
    sectionRight: css.css({
      justifyContent: "flex-end",
      flexGrow: 1,
      flexBasis: "50%"
    })
  };
};

exports.VizLegendList = VizLegendList;
