// @grafana/ui/dist/cjs/components/DateTimePickers/TimeZonePicker/TimeZoneGroup.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var ThemeContext = require('../../../themes/ThemeContext.cjs');

"use strict";
const TimeZoneGroup = (props) => {
  const { children, label } = props;
  const styles = ThemeContext.useStyles2(getStyles);
  if (!label) {
    return /* @__PURE__ */ jsxRuntime.jsx("div", { children });
  }
  return /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
    /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles.header, children: /* @__PURE__ */ jsxRuntime.jsx("span", { className: styles.label, children: label }) }),
    children
  ] });
};
const getStyles = (theme) => {
  return {
    header: css.css({
      padding: "7px 10px",
      width: "100%",
      borderTop: `1px solid ${theme.colors.border.weak}`,
      textTransform: "capitalize"
    }),
    label: css.css({
      fontSize: theme.typography.size.sm,
      color: theme.colors.text.secondary,
      fontWeight: theme.typography.fontWeightMedium
    })
  };
};

exports.TimeZoneGroup = TimeZoneGroup;
