// @grafana/ui/dist/cjs/components/DateTimePickers/TimeZonePicker/TimeZoneDescription.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var css = require('@emotion/css');
var React = require('react');
var ThemeContext = require('../../../themes/ThemeContext.cjs');

"use strict";
const TimeZoneDescription = ({ info }) => {
  const styles = ThemeContext.useStyles2(getStyles);
  const description = useDescription(info);
  if (!info) {
    return null;
  }
  return /* @__PURE__ */ jsxRuntime.jsx("div", { className: styles.description, children: description });
};
const useDescription = (info) => {
  return React.useMemo(() => {
    const parts = [];
    if (!info) {
      return "";
    }
    if (info.name === "Europe/Simferopol") {
      return "Ukraine, EEST";
    }
    if (info.countries.length > 0) {
      const country = info.countries[0];
      parts.push(country.name);
    }
    if (info.abbreviation) {
      parts.push(info.abbreviation);
    }
    return parts.join(", ");
  }, [info]);
};
const getStyles = (theme) => {
  return {
    description: css.css({
      fontWeight: "normal",
      fontSize: theme.typography.size.sm,
      color: theme.colors.text.secondary,
      whiteSpace: "normal",
      textOverflow: "ellipsis"
    })
  };
};

exports.TimeZoneDescription = TimeZoneDescription;
