// @grafana/ui/dist/cjs/components/ConfirmButton/DeleteButton.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var i18n = require('@grafana/i18n');
var Button = require('../Button/Button.cjs');
var ConfirmButton = require('./ConfirmButton.cjs');

"use strict";
const DeleteButton = ({ size, disabled, onConfirm, "aria-label": ariaLabel, closeOnConfirm }) => {
  return /* @__PURE__ */ jsxRuntime.jsx(
    ConfirmButton.ConfirmButton,
    {
      confirmText: i18n.t("grafana-ui.confirm-button.confirmText-delete", "Delete"),
      confirmVariant: "destructive",
      size: size || "md",
      disabled,
      onConfirm,
      closeOnConfirm,
      children: /* @__PURE__ */ jsxRuntime.jsx(
        Button.Button,
        {
          "aria-label": ariaLabel != null ? ariaLabel : i18n.t("grafana-ui.confirm-button.aria-label-delete", "Delete"),
          variant: "destructive",
          icon: "times",
          size: size || "sm"
        }
      )
    }
  );
};

exports.DeleteButton = DeleteButton;
